---
title: "Experimental overview"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
---


```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```


```{r libraries, warning=FALSE, message=FALSE}


# qtl mapping + mediation
library(intermediate) # "simecek/intermediate"
library(qtl2) 

# # plotting
library(ggpubr)
library(igraph)
library(ggraph)
library(pheatmap)
library(cowplot)
library(GGally)
library(corrplot)

# annotations + general genomic things
#library(biomaRt)
library(GenomicRanges)

# data processing
library(Hmisc) # rcorr
library(gprofiler2)
library(sva)
library(WebGestaltR)

library(readxl)
library(tidyverse)
select <- dplyr::select # I am adding this explicitly
rename <- dplyr::rename # I am adding this explicitly

# setting path
library(here)

```

```{r functions}

# Making downloadable data tables
# https://www.r-bloggers.com/vignette-downloadable-tables-in-rmarkdown-with-the-dt-package/
create_dt <- function(x){
  DT::datatable(x,
                extensions = 'Buttons',
                rownames = FALSE, 
                filter="top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel'),
                               pageLength = 5, 
                               scrollX= TRUE
                               ))
  
}

```

```{r load_data}

# Note: I am keeping v91 ids + annotations for these two data sets. I am not going to add the protein data to these so I think it is okay. 

# The mediations with ESC/NPC transcript were done for all NPC eQTL lod >6 using annotations from v84 ensembl ids and Duy's script that adds LOD at the peak, causality test. I am not using the causality test to filter anything but I do use lod drop z-threshold (z< -4) and positional threshold (diff < 10Mb) for filtering.

# I am re-writing the overlap script for ESC/NPC eQTL without the ESC pQTL and will add allele effect correlations to it too . 

all.genes_v91 <- read_tsv( file = here("_data","/ENSMUSGid_to_symbol_v91.txt")) %>% 
  rename( ensembl_gene_id = `Gene stable ID`,
          gene_start = `Gene start (bp)`, 
          gene_end = `Gene end (bp)`, 
          gene_chr = `Chromosome/scaffold name`,
          gene_biotype = `Gene type`, 
          mgi_symbol = `MGI symbol`) %>% 
  mutate( midpoint = (gene_start+gene_end)/2)

# ESC expression
load(here("_data/DO_mESC_paired_eQTL_forMapping.RData"))
raw.expr.esc_rna <- esc.raw.expr
exprZ.esc_rna <- esc.exprZ
kinship_loco.esc_rna <- esc.kinship_loco
probs.esc_rna <- esc.probs
covar.esc_rna <- esc.covar
covarTidy.esc_rna <- covarTidy
exprComBat.esc_rna <- esc.expr.ComBat
expr.esc_rna <- expm1(exprComBat.esc_rna) # re-transforming since the data was log(x+1) before combat
expr.esc_rna[expr.esc_rna < 0] <- 0
expr.esc_rna <- t(expr.esc_rna)
rm(esc.expr, esc.exprZ, esc.kinship_loco, esc.probs, esc.expr.ComBat, esc.raw.expr, covarTidy, exprComBat.esc_rna, esc.covar, esc.covarTidy)

# NPC expression
load(here("_data/DO_mNPC_paired_eQTL_forMapping.RData"))
raw.expr.npc_rna <- npc.raw.expr
exprZ.npc_rna <- npc.exprZ
kinship_loco.npc_rna <- npc.kinship_loco
probs.npc_rna <- npc.probs
covar.npc_rna <- npc.covar
covarTidy.npc_rna <- npc.covarTidy
exprComBat.npc_rna <- npc.expr.ComBat
expr.npc_rna <- expm1(exprComBat.npc_rna) # re-transforming since the data was log(x+1) before combat
expr.npc_rna[expr.npc_rna < 0] <- 0
expr.npc_rna <- t(expr.npc_rna)
rm(npc.expr, npc.exprZ, npc.kinship_loco, npc.probs, npc.expr.ComBat, npc.raw.expr, exprComBat.npc_rna, npc.covar, npc.covarTidy)

# ESC eQTL map + effects
load(here("_data/ESC_eQTL_effects.RData"))
peaks.esc_rna <- peaks
effects_blup.esc_rna <- effects_blup
effects_std.esc_rna <- effects_std
rm(peaks, effects_blup, effects_std)

# Adding effects to eQTL peaks
peaks.esc_rna.blup <- cbind(peaks.esc_rna, effects_blup.esc_rna) %>%
  dplyr::rename("ensembl_gene_id" = "phenotype")
colnames(peaks.esc_rna.blup) <- c(
  colnames(peaks.esc_rna.blup)[1:2],
  paste0(colnames(peaks.esc_rna.blup)[3:dim(peaks.esc_rna.blup)[2]], ".esc_rna")
)

# adding annotations to eQTL peaks with effects
peaks.esc_rna.wEffs <- peaks.esc_rna.blup %>% 
  left_join( all.genes_v91) %>% 
  mutate(midpoint = (gene_start + gene_end) / 2) %>% 
  mutate( same_chrom =  (peak_chr == gene_chr),
          diff = abs(midpoint - interp_bp_peak.esc_rna)) %>% 
  mutate( local.esc_rna = ifelse( same_chrom & 
                            diff < 10e06, TRUE, FALSE
    ))%>% 
  select(-same_chrom, -diff)


# NPC eQTL map + effects
load(here("_data/NPC_eQTL_effects.RData"))
peaks.npc_rna <- peaks
effects_blup.npc_rna <- effects_blup
effects_std.npc_rna <- effects_std
rm(peaks, effects_blup, effects_std)

# Adding effects to eQTL peaks
peaks.npc_rna.blup <- cbind(peaks.npc_rna, effects_blup.npc_rna) %>%
  dplyr::rename("ensembl_gene_id" = "phenotype")
colnames(peaks.npc_rna.blup) <- c(
  colnames(peaks.npc_rna.blup)[1:2],
  paste0(colnames(peaks.npc_rna.blup)[3:dim(peaks.npc_rna.blup)[2]], ".npc_rna")
)

# adding annotations to eQTL peaks with effects
peaks.npc_rna.wEffs <- peaks.npc_rna.blup %>% 
  left_join( all.genes_v91) %>% 
  mutate(midpoint = (gene_start + gene_end) / 2) %>% 
  mutate( same_chrom =  (peak_chr == gene_chr),
          diff = abs(midpoint - interp_bp_peak.npc_rna)) %>% 
  mutate( local.npc_rna = ifelse( same_chrom & 
                            diff < 10e06, TRUE, FALSE
    )) %>% 
  select(-same_chrom, -diff)


# get all genes that contain the union of both data sets + esc/npc gene lists.
all.genes <- filter(all.genes_v91, ensembl_gene_id %in% c(colnames(exprZ.esc_rna), colnames(exprZ.npc_rna)) )
esc.genes <- filter( all.genes, ensembl_gene_id %in% colnames(exprZ.esc_rna)) # note that we are missing annotations for 61 ids, ensembl say they are deprecated.
npc.genes <- filter( all.genes, ensembl_gene_id %in% colnames(exprZ.npc_rna)) # note that we are missing annotations for 36 ids, ensembl says they are deprecated.


## shared data
# get the set of shared genes
shared.genes <- intersect(colnames(expr.esc_rna),colnames(expr.npc_rna))

# get the set of shared samples
shared.samples <- intersect(rownames(expr.esc_rna)[!grepl("repB",rownames(expr.esc_rna))],
                            rownames(expr.npc_rna)[!grepl("repB",rownames(expr.npc_rna))])

# let's subset the expression matrices for shared genes + samples 
shared.expr.npc_rna <- expr.npc_rna[shared.samples,shared.genes ]
shared.expr.esc_rna  <- expr.esc_rna[shared.samples,shared.genes]

# let's subset the probs matrices
shared.probs.esc_rna <- probs.esc_rna[ ind=shared.samples]
shared.probs.npc_rna <- probs.npc_rna[ind=shared.samples]

## merged peaks
# ESC/NPC eQTL overlap
load(here("_data","peaks_comparison_10Mb_ESC_NPC.RData")) # peaks.esc.npc.rna

# there are some significant NPC eQTL that should be shared but is not. See below, might be worth fixing those.
# peaks.esc.npc.rna %>% 
#   filter( lod.esc_rna > 5 & lod.npc_rna >7.5 & match != "shared" & abs(interp_bp_peak.esc_rna-interp_bp_peak.npc_rna) < 10e6)

## mediation results
load(here("_data","DO_mNPC_eQTL_NPC_RNA_mediation_lod6_merged.RData")) ## eqtl_npc_rna_meds
load(here("_data","DO_mNPC_eQTL_ESC_RNA_mediation_lod6_merged.RData")) ## eqtl_esc_rna_meds

## adding cc_variants for association mapping
query_variants <- create_variant_query_func(here("_data/cc_variants.sqlite"))
query_genes <- create_gene_query_func("_data/mouse_genes_mgi.sqlite")


# prep some stuff for plotting:
uchr <- c(as.character(1:19), "X")
cl <- dplyr::select(map_dat2, chr, pos_bp) %>% group_by(chr) %>%
  summarize(len=max(pos_bp))
clp <- with(cl, setNames(len, chr))
chrom_lens <- setNames(as.numeric(clp[uchr]), uchr)
chrom_lens_offset <- cumsum(chrom_lens) - chrom_lens
chrom_lens_midpt <- chrom_lens_offset + chrom_lens/2

# qtl colors
qtl.colors <- c( esc_rna = "#009E73", 
                 npc_rna = "#D55E00",
                 shared = "#7570B3")
# founder colors
founder_colors <- c(AJ = "#F0E442", B6 = "#555555", `129` = "#E69F00", NOD = "#0072B2",
   NZO = "#56B4E9", CAST = "#009E73", PWK = "#D55E00", WSB = "#CC79A7")

# lifr genotypes
# get_LIFR_genotypes
# using Dan's code to get LIFR genotypes for the full list of animals
probs <- probs.npc_rna
markers <- tibble(name = dimnames(probs[[15]])[[3]]) %>%
  mutate(name2 = name) %>%
  separate(name2, into = c("chrom", "pos"), sep = "_", convert = TRUE)

# LIFr SNP is chr15:7116944 (rs50454566)
mm <- filter(markers, chrom == "15", pos > 7090000, pos < 7130000) # 3 markers
probs2 <- probs$`15`[, , mm$name]

closest_geno <- function(p, tol = 0.01) {
  if (sum(abs(p - c(1, 0))) < tol) {
    return("A")
  }
  if (sum(abs(p - c(0, 1))) < tol) {
    return("B")
  }
  if (sum(abs(p - c(0.5, 0.5))) < tol) {
    return("H")
  }
  return(NA)
}
call_geno <- function(mat) {
  # mat is nsamp*8 (haps)
  # A = A_J
  # B = B6
  # C = 129
  # D = NOD
  # E = NZO
  # F = CAST
  # G = PWK
  # H = WSB
  # I want to divide NOD + CAST + PWK + WSB
  # vs. the other four
  grp <- c("A", "A", "A", "B", "A", "B", "B", "B")
  collapsed <- apply(mat, 1, function(x) tapply(x, grp, sum))
  apply(collapsed, 2, closest_geno)
}

probs3 <- apply(probs2, 3, call_geno)
# assert_that(noNA(probs3))
one <- probs3[, 1] # marker left of Lifr
two <- probs3[, 2] # closest marker to Lifr
three <- probs3[, 3] # marker right of Lifr
# "PB360.49" has an ancestry switch between markers 1 & 2!


# Get samples in group A (inbred strains) and group B (wild-derived + NOD)
inbred <- rownames(probs[[1]])[one == "A" & two == "A" & three == "A"]
wildder <- rownames(probs[[1]])[one == "B" & two == "B" & three == "B"]
hets <- rownames(probs[[1]])[one == "H" & two == "H" & three == "H"]
# cat(inbred, sep="\n", file="lifr_genotype_inbred.txt")
# cat(wildder, sep="\n", file="lifr_genotype_wildder.txt")
# cat(hets, sep="\n", file="lifr_genotype_het.txt")
#
data_frame(
  lifr_geno = factor(c(rep("Ref", length(inbred)), c(rep("Alt", length(wildder))), c(rep("Het", length(hets))))),
  lifr = factor(c(rep(0, length(inbred)), c(rep(1, length(wildder))), c(rep(2, length(hets))))),
  rowname = c((inbred), (wildder), (hets))
) %>%
  mutate(rowname = ifelse(is.na(rowname), "PB360.49_repA", rowname)) -> covar.lifr

covarTidy.npc_rna %>% 
  left_join( covar.lifr, by = c("sampleid"="rowname")) -> covarTidy.npc_rna


```

Here, I am contrasting global gene expression and eQTL mapping results between embryonic stem cells (ESCs) and neural progenitor cells (NPCs) derived from many Diversity Outbred (DO) mice. We analyzed RNA-seq data from a total of `r nrow(expr.esc_rna)` ESC and `r nrow(expr.npc_rna)` NPC lines; 127 lines had RNA-seq data for both ESCs and NPCs. After filtering out genes with low expression and genes not expressed in at least half of the samples, we detect `r formatC(ncol(expr.esc_rna),big.mark = ",")` in ESCs and `r formatC(ncol(expr.npc_rna),big.mark = ",")` in NPCs, with `r formatC(length(shared.genes),big.mark = ",")` genes detected in both cell types. The filtered data was then normalized to the upper quartile value (each read count is divided by the 75th percentile of the read counts across samples) to allow for comparison of expression measures within and across cell-types. Both data-sets were generated in batches, and we used ComBat as implemented in the R package sva to correct for batch effects while controlling for the sexes of lines in each batch. Both ESC and NPC expression values were transformed to rank normal scores prior to eQTL mapping.


In this first tab, I focus on the NPC transcriptome investigating the variation in gene expression in DO mNPCs using a variety of methods including: principal component analysis (PCA), gene set variation analysis (GSVA) and weighted correlation network analysis (WGCNA). Then, I looked at the co-variation across DO mNPCs and mESCs by looking at the agreement between transcriptomes within genetically identical samples and between genes across cell states. Further, there are genes uniquely expressed in each cell state which I functionally characterized using overrepresentation analysis (ORA). I used Wilcoxon rank sum test to identify differentially expressed genes in mNPCs in contrast to mESCs and functionally characterized up and down regulated genes using ORA.

In the second tab, I am presenting the genetic mapping and mediation results with DO mNPC transcripts. First, I calculate the heritability of each transcript in DO mNPCs and present the DO mNPC eQTL map. Next, I integrate the DO mNPC eQTL map with the mESC eQTL map and compare the allel effects of overlapping eQTL. Mediation analysis results are summarized in two notebooks where the first one is summarizing the full mediation results focusing on local and distant eQTL with best mediators being ESC or NPC transcripts. Second notebook describes the DO mNPC eQTL hotspots, their characterization and mediation. 



















