---
title: ""
output:
  html_document:
    toc: false
    df_print: paged
    code_folding: hide
---

```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```

# Variation in NPC transcriptome {.tabset .tabset-fade .tabset-pills}

There are `r formatC(ncol(expr.npc_rna), big.mark = ",")` many genes measured across `r nrow(expr.npc_rna)` samples.

```{r npc_stats, warning=FALSE, message=FALSE}

# npc_rna gene stats
var.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, var = V1) %>%
  arrange(desc(var))

n.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, n = V1) %>%
  arrange(desc(n))

mean.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, mean = V1) %>%
  arrange(desc(mean))

var.npc_rna <- inner_join(var.npc_rna, mean.npc_rna) %>%
  inner_join(., n.npc_rna) %>%
  left_join(., select(all.genes, ensembl_gene_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.npc_rna = 100 * sd / (mean)) %>%
  rename(mean.npc_rna = mean, sd.npc_rna = sd, n.npc_rna=n, var.npc_rna=var)

# npc stats per sample
var_npc_per_sample <- expr.npc_rna %>% 
  t() %>% 
  as_tibble( rownames = "ensembl_gene_id") %>%
  summarise( across( where(is.numeric),  var , na.rm=T )) %>% 
  pivot_longer( 1:ncol(.), names_to = "sampleid", values_to ="var_sample") %>% 
  left_join( select(covarTidy.npc_rna, sampleid, sex))
  

mean_npc_per_sample <- expr.npc_rna %>%
  t() %>% 
  as_tibble( rownames = "ensembl_gene_id") %>%
  summarise( across( where(is.numeric),  mean , na.rm=T )) %>% 
  pivot_longer( 1:ncol(.), names_to = "sampleid", values_to ="mean_sample") %>% 
  left_join( select(covarTidy.npc_rna, sampleid, sex))

```

## Overview

Below are the distributions of mean and variation of transcripts in DO mNPC lines.

```{r npc_stat_plots_gene, fig.width=12, fig.height=4, warning=FALSE, message=FALSE}

# add mean + variance + CV plots per gene
var.npc_rna %>%
  ggplot() +
  aes(x = mean.npc_rna) +
  geom_histogram(bins = 200) +
  xlab("Mean") +
  theme_pubclean(base_size = 14)+
  scale_x_log10() -> p.mean.hist

var.npc_rna %>%
  ggplot() +
  aes(x = var.npc_rna) +
  geom_histogram(bins = 200 ) +
  xlab("Variance") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()  -> p.var.hist

var.npc_rna %>%
  ggplot() +
  aes(x = cv.npc_rna) +
  geom_histogram(bins = 200 ) +
  xlab("% Coefficient of variation") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()  -> p.cv.hist

ggscatter(var.npc_rna,
  x = "mean.npc_rna", y = "var.npc_rna", size = 3, alpha = 0.6,
  add = "reg.line", # Add regression line
  conf.int = TRUE, # Add confidence interval
  add.params = list(color = "blue", fill = "lightgray"), show.legend.text = FALSE,
  yscale = "log10", xscale = "log10"
) +
  stat_cor(method = "pearson", label.x = -1, label.y = 10) + # Add correlation coefficient
  xlab("Mean") +
  ylab("Variance") +
  theme_pubclean(base_size = 14) +
  rremove("legend") -> p.var.mean

ggarrange( p.mean.hist, p.var.hist, p.var.mean, nrow = 1)

```

Here are the distributions of mean and variation of DO mNPC lines.

```{r npc_stats_plots_sample, fig.width=16, fig.height=4, warning=FALSE, message=FALSE}
# add mean + variance + CV plots per sample
var_npc_per_sample %>% 
  ggplot() +
  aes(x = var_sample, col = sex, fill =sex) +
  geom_histogram(bins = 100 ) +
  xlab("Variance") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  color_palette("npg")+
  fill_palette("npg") -> p.var_sample.hist

mean_npc_per_sample %>% 
  ggplot() +
  aes(x = mean_sample, col = sex, fill =sex) +
  geom_histogram(bins = 100 ) +
  xlab("Mean") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  color_palette("npg")+
  fill_palette("npg") -> p.mean_sample.hist

full_join(var_npc_per_sample, mean_npc_per_sample) %>% 
  mutate(cv_sample = 100*(sqrt(var_sample))/mean_sample) %>%
  ggscatter(
    .,
  x = "mean_sample", y = "var_sample", size = 3, alpha = 0.6,
  add = "reg.line", # Add regression line
  conf.int = TRUE, # Add confidence interval
  col = "sex",
  add.params = list(color = "blue", fill = "lightgray"),
  show.legend.text = FALSE
  #yscale = "log10", xscale = "log10"
) +
  #stat_cor(method = "pearson", label.x = 10, label.y = 21) + # Add correlation coefficient
  xlab("Mean per sample") +
  ylab("Variance per sample") +
  theme_pubclean(base_size = 14) +
  color_palette("npg")+
  fill_palette("npg")+
  facet_wrap(~sex)+
  theme(
  strip.background = element_blank(),
  strip.text.x = element_blank()
  )-> p.var.mean_sample

ggarrange( p.mean_sample.hist, p.var_sample.hist,  p.var.mean_sample, nrow = 1, widths = c(0.5,0.5,1))

```

## Principal component analysis

```{r npc_pca, warning=FALSE, message=FALSE}

pca.npc <- prcomp(expr.npc_rna, scale.=T, center=T)

```

Let's look at the first five PCs and how they group the data for the NPC transcriptome. No separation by sex is observed in the NPC data!

```{r pca_plot_pc1to5, fig.width=10, fig.height=8, warning=FALSE, message=FALSE}

pca.npc$x%>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(covarTidy.npc_rna, by = c("rowname" = "sampleid")) %>%
  mutate(sex = ifelse(sex == "F", "Female", "Male")) %>%
  GGally::ggpairs(.,
    columns = 2:6, progress = FALSE, ggplot2::aes(color = sex),
    upper = list(continuous = "density", combo = "box_no_facet"),
    lower = list(continuous = "points", combo = "dot_no_facet")
  ) + 
  theme_pubclean(base_size = 14) +
  color_palette("npg")+
  fill_palette("npg")+
  theme(legend.position = "top")

```

```{r npc_scree_plot, warning=FALSE, message=FALSE}

# plot showing variance explained
var_explained <- (pca.npc$sdev^2 / sum(pca.npc$sdev^2))
tibble(var =  var_explained[1:10]* 100, PC = paste0("PC", seq(1:10))) %>%
  arrange(desc(var)) %>%
  mutate(label = factor(PC, levels = PC)) %>%
  ggplot(aes(x = label, y = var)) +
  geom_col() +
  xlab("Principal Component") +
  ylab("% Variation explained") +
  theme_pubclean(base_size = 14)+
  ylim(0,15)


```

```{r pc1_2_plot, fig.height=5, fig.width=6}


pca.npc$x%>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(covarTidy.npc_rna, by = c("rowname" = "sampleid")) %>%
  mutate(sex = ifelse(sex == "F", "Female", "Male")) %>% 
  ggplot()+
  aes(
    x = PC1, 
    y = PC2, 
    col = sex
  )+
  geom_point(size = 4, alpha = 0.7)+
  theme_pubclean(base_size = 20)+
  color_palette("npg")+
  xlab(paste0("PC1 (",100*round(var_explained[1],2),"%)"))+
  ylab(paste0("PC2 (",100*round(var_explained[2],2),"%)")) -> pca_plot

pca_plot

```

Are PC1 drivers over-represented in any biological processes or pathways? Looking at the top 5% contributing genes of PC1.

```{r pc1_ora}

pc1_drivers <- pca.npc$rotation[,"PC1", drop = FALSE] %>% 
  as_tibble( rownames = "ensembl_gene_id") %>% 
  left_join( npc.genes %>%  
               select(ensembl_gene_id, mgi_symbol)
             ) %>% 
  filter( abs(PC1) >= quantile(abs(PC1), 0.95))

g.pc1 <- gost(
  query = pc1_drivers$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.pc1$result <- g.pc1$result %>% filter(term_size < 660)

```

```{r ora_pc1_results}

g.pc1$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

Looking at the genes identified as part of:

-   Spindle checkpoint and related processes

-   Pluripotency regulation and related processes

```{r ora_pc1_genes}

# any category with spindle, segregation
spindle_genes <- g.pc1$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "(?i)spindle") |
            str_detect(term_name, "(?i)segregation")|
            str_detect(term_name, "(?i)chromatid") |
            str_detect(term_name, "(?i)kinetochore") 
           #str_detect(term_name, "(?i)checkpoint") 
            #str_detect(intersection,"Mad2l1")
          ) %>% 
  filter( !source %in% c("GO:CC","GO:MF")) %>% 
  select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes)

# any category with pluripotency, lifr, blastocyst, embryo
pluri_genes <- g.pc1$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "leukemia inhibitory factor") |
            str_detect(term_name, "blastocyst")|
            str_detect(term_name, "embryo")) %>% 
    select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes)

```


```{r, fig.width=12, fig.height=4}


var.npc_rna %>% 
  ggscatter(., 
            x = "mean.npc_rna", 
            y = "var.npc_rna", 
            size = 3, 
            alpha = 0.5,
            col="gray",
            yscale = "log10",
            xscale = "log10",
            show.legend.text = FALSE
            ) +
  xlab("Mean transcript abundance") +
  ylab("Variance in transcript abundance") +
  ggtitle("Pluripotency related genes")+
  theme_pubclean(base_size = 18) + 
  rremove("legend") +
  geom_point(
    data =   filter( var.npc_rna, ensembl_gene_id %in% pluri_genes$ensembl_gene_id) ,
    col = "blue", alpha = 0.6, size = 3)  -> plot_pluri_genes


var.npc_rna %>% 
  ggscatter(., 
            x = "mean.npc_rna", 
            y = "var.npc_rna", 
            size = 3, 
            alpha = 0.5,
            col="gray",
            yscale = "log10",
            xscale = "log10",
            show.legend.text = FALSE
            ) +
  xlab("Mean transcript abundance") +
  ylab("Variance in transcript abundance") +
  ggtitle("SAC related genes")+
  theme_pubclean(base_size = 18) + 
  rremove("legend") +
  geom_point(
    data =   filter( var.npc_rna, ensembl_gene_id %in% spindle_genes$ensembl_gene_id) ,
    col = "blue", alpha = 0.6, size = 3) -> plot_spindle_genes

ggarrange( plot_pluri_genes, plot_spindle_genes, nrow =1 
        )

```


```{r pluri_genes_pc1_values, fig.width=12, fig.height=6}

pca.npc$rotation %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(all.genes, by = c("rowname" = "ensembl_gene_id")) %>%
  filter( mgi_symbol %in% pluri_genes$mgi_symbol) %>% 
  select(mgi_symbol, PC1) %>% 
  full_join(pluri_genes) -> pluri_genes_pc1

g<-graph_from_data_frame(pluri_genes_pc1[,c("mgi_symbol","term_name")], directed=FALSE)
V(g)$lfc <- c((pluri_genes_pc1 %>% 
  select(mgi_symbol, PC1) %>% 
  distinct())$PC1, rep(NA, length(unique(pluri_genes_pc1$term_name)))) # last two are the category names

ggraph(g)+
  geom_edge_link(col="gray", alpha=0.1)+
  geom_node_point(aes(col=lfc),size=5)+
  geom_node_text(aes(label = name), repel = TRUE )+
  scale_color_viridis( limits = c(-0.025,0.025))+
  theme_classic()+
  theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank())+
  labs(col="PC1 loadings")


```


```{r spindle_genes_pc1_values, fig.width=14, fig.height=12}

pca.npc$rotation %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(all.genes, by = c("rowname" = "ensembl_gene_id")) %>%
  filter( mgi_symbol %in% spindle_genes$mgi_symbol) %>% 
  select(mgi_symbol, PC1) %>% 
  full_join(spindle_genes) -> spindle_genes_pc1

g<-graph_from_data_frame(spindle_genes_pc1[,c("mgi_symbol","term_name")], directed=FALSE)
V(g)$lfc <- c((spindle_genes_pc1 %>% 
  select(mgi_symbol, PC1) %>% 
  distinct())$PC1, rep(NA, length(unique(spindle_genes_pc1$term_name)))) # last two are the category names

ggraph(g)+
  geom_edge_link(col="gray", alpha=0.1)+
  geom_node_point(aes(col=lfc),size=5)+
  geom_node_text(aes(label = name), repel = TRUE )+
  scale_color_viridis( limits = c(-0.025,0.025))+
  theme_classic()+
  theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank())+
  labs(col="PC1 loadings")


```

Highligthing two ontologies in particular for the poster and perhaps the paper(?):

```{r pc1_ora_higlight, fig.width=14, fig.height=4}

highlights <- c("response to leukemia inhibitory factor",
                "Mitotic Spindle Checkpoint"
                )

g.pc1$result %>% 
  filter( p_value <0.01) %>% 
  filter( term_name %in% highlights) %>% 
  select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes) %>% 
  left_join(
    pca.npc$rotation %>%
      as_tibble( rownames = "ensembl_gene_id")
  )-> highlight_genes_pc1

g<-graph_from_data_frame(highlight_genes_pc1[,c("mgi_symbol","term_name")], directed=F)
V(g)$lfc <- c((highlight_genes_pc1 %>% 
  select(mgi_symbol, PC1) %>% 
  distinct())$PC1, rep(NA, length(unique(highlight_genes_pc1$term_name)))) # last two are the category names

highlights_network_plot <- ggraph(g,layout = "stress")+
  geom_edge_link(col="gray", alpha=0.1)+
  geom_node_point(aes(col=lfc),size=5)+
  geom_node_text(aes(label = name), repel = TRUE )+
  scale_color_viridis( limits = c(-0.025,0.025))+
  theme_classic()+
  theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank())+
  labs(col="PC1 loadings")

figure_1b <- ggarrange(pca_plot, highlights_network_plot, widths = c(0.6, 1.2))
figure_1b

# ggsave(figure_1b, filename = here("figures","figure_1b.svg"),
#                                                 width = 14, height = 4, units = "in", dpi = 300)

```


## Most and least variable genes in NPCs

```{r most_least_var}

high.var.genes <- var.npc_rna %>% 
  filter( cv.npc_rna >= quantile(cv.npc_rna, 0.95))

low.var.genes <- var.npc_rna %>% 
  filter( cv.npc_rna <= quantile(cv.npc_rna, 0.05))

var.npc_rna <- var.npc_rna %>%
  mutate(type = ifelse(ensembl_gene_id %in% high.var.genes$ensembl_gene_id, "high", "none")) %>%
  mutate(type = ifelse(ensembl_gene_id %in% low.var.genes$ensembl_gene_id, "low", type))

g.high.var <- gost(
  query = high.var.genes$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.high.var$result <- g.high.var$result %>% filter(term_size < 660)

g.low.var <- gost(
  query = low.var.genes$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.low.var$result <- g.low.var$result %>% filter(term_size < 660)

```

ORA results with most variable (highest .05% CV) genes in mNPCs (FDR \<0.01):

```{r ora_most_var}

g.high.var$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

<br>

<br>

ORA results with least variable (lowest .05% CV) genes in mNPCs (FDR \<0.01):

```{r ora_least_var}

spindle_genes_low_var <- g.low.var$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "(?i)spindle") |
            str_detect(term_name, "(?i)segregation")|
            str_detect(term_name, "(?i)chromatid") |
            str_detect(term_name, "(?i)kinetochore") |
            str_detect(intersection,"Mad2l1")) %>% 
  select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes)

g.low.var$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

## Sex effects on gene expression

```{r sex_eff, cache = TRUE}

# doing an anova first and then tukey's 
# reporting ones with significance on both as genes with sex effects

# updating the code to use anova followed by tukey's hsd:
expr.npc_rna %>%
  t() %>% 
  as_tibble(rownames = "ensembl_gene_id") %>%
  pivot_longer( cols = rownames(expr.npc_rna),
                values_to = "gene_expr",
                names_to = "sampleid") %>% 
  left_join(., select(covarTidy.npc_rna, sampleid, sex)) %>% 
  group_by(ensembl_gene_id) %>% 
  rstatix::anova_test( gene_expr ~ sex) %>% 
  rstatix::adjust_pvalue( method = "BH") %>% 
  rstatix::add_significance("p.adj") %>% 
  as_tibble() -> npc_rna_sex_aov

# passing the full data to tukey's then filtering
expr.npc_rna %>%
  t() %>% 
  as_tibble(rownames = "ensembl_gene_id") %>% 
  pivot_longer( cols = rownames(expr.npc_rna),
                values_to = "gene_expr",
                names_to = "sampleid") %>% 
  left_join(., select(covarTidy.npc_rna, sampleid, sex)) %>% 
  group_by(ensembl_gene_id) %>% 
  rstatix::tukey_hsd(gene_expr ~ sex) %>% 
  filter( ensembl_gene_id %in% (filter(npc_rna_sex_aov, p.adj.signif != "ns"))$ensembl_gene_id & # filtering to only include genes with signif aov + tukeys
            p.adj.signif != "ns") -> npc_rna_sex_tukeys


# get the medians for later
expr.npc_rna %>%
  t() %>% 
  as_tibble(rownames = "ensembl_gene_id") %>%
  pivot_longer( cols = rownames(expr.npc_rna),
                values_to = "gene_expr",
                names_to = "sampleid") %>% 
  left_join(., select(covarTidy.npc_rna, sampleid, sex)) %>% 
  group_by(ensembl_gene_id,sex) %>% 
  summarize( med = median(gene_expr, na.rm =T)) %>% 
  pivot_wider( id_cols = "ensembl_gene_id",
               names_from = "sex",
               values_from = "med")-> npc_rna_sex_med

```

```{r sex_effects_table}

npc_rna_sex_tukeys %>%
  left_join( npc.genes) %>% 
  left_join( npc_rna_sex_med) %>% 
  arrange(p.adj) %>%
  mutate_if( is.numeric, round, 2) %>%
  select(
    `Gene ID` = ensembl_gene_id,
    `MGI Symbol`= mgi_symbol, 
    `Gene location (chr)` = gene_chr,
    `Female median`=`F`,
    `Male median`= M
   ) %>%
  create_dt()

```

Overrepresented pathways that show sex effects:

```{r ora_sex_effs}

sex_eff_genes <- npc_rna_sex_tukeys %>%
  left_join( npc.genes) %>% 
  filter(p.adj < 0.05) 

npc_sex_ora <- gost(query = sex_eff_genes$mgi_symbol, 
                 organism = "mmusculus",
                 significant = TRUE, 
                 domain_scope = "custom",
                 custom_bg = npc.genes$mgi_symbol,
                 correction_method = "fdr",
                 evcodes = TRUE)
npc_sex_ora$result <- npc_sex_ora$result %>% filter(term_size <500)

```

```{r}

# any category with spindle, segregation
spindle_genes_sex_eff <- npc_sex_ora$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "(?i)spindle") |
            str_detect(term_name, "(?i)segregation")|
            str_detect(term_name, "(?i)chromatid") |
            str_detect(term_name, "(?i)kinetochore") |
            str_detect(intersection,"Mad2l1")) %>% 
  select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes)

npc_sex_ora$result %>% 
    select( term_name, source, FDR = p_value, term_size, intersection_size,intersection) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

## Gene set variation analysis

```{r npc_rna_gsva, cache=TRUE, warning=FALSE, message=FALSE}

# reading in the GO + mgi downloaded from: http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt
go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt") %>% 
  mutate( genes = str_split(genes, ",")) %>% 
  unnest() # separete the symbols, note the overlap: length(intersect(unique(go_terms$genes), npc.genes$mgi_symbol) ) = 11806


slim_go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/map2MGIslim.txt") %>% 
  select(-term) %>% 
  mutate( ONT = case_when( aspect == "P" ~  "BP",
                     aspect == "F" ~ "MF",
                     aspect == "C" ~ "CC"
                     )
          ) %>% 
  select(-aspect)

genesbygo <- split(go_terms$genes, go_terms$GO_id)

go_terms_annot <- go_terms %>%  
  select(GO_id) %>% 
  distinct() %>% 
  left_join( slim_go_terms %>%  select( GO_id, ONT) %>% distinct())

goannot_wdef <- AnnotationDbi::select(GO.db, keys= unique(go_terms$GO_id), columns=c("GOID","DEFINITION","ONTOLOGY","TERM")) %>%
  left_join( slim_go_terms, by=c("GOID"="GO_id")) %>% 
  mutate( ONTOLOGY = ONT) %>% 
  select(-ONT)

go_bp <- goannot_wdef %>% filter( ONTOLOGY == "BP") %>% 
  select(GOID) %>%  distinct()

# expr - change gene id's to symbols for GSVA with GO
expr.npc_rna_upd <- expr.npc_rna[, npc.genes$ensembl_gene_id]
colnames(expr.npc_rna_upd) <- npc.genes$mgi_symbol


# I can't seem to run GSVA anymore! Keep erroring out. 
# I was able to run it separately in a stand alone R session.
# loading the results here instead of running.
load(here("_data","Results_from_GSVA_2022-09-23.RData"))
# gsva_rna <- gsva(  expr = t(expr.npc_rna_upd),
#                    gset.idx.list = genesbygo,
#                     method ="gsva",
#                     kcdf = "none",
#                     min.sz = 5, 
#                     max.sz = 1000,
#                     mx.diff = TRUE
#                    )


# following up on rna results
gsva_rna %>% 
  as_tibble(rownames = "Category") %>% 
  filter( Category %in% go_bp$GOID) %>% #filtering for BP
  pivot_longer( cols = rownames(expr.npc_rna_upd),
                values_to = "Enrichment_Score",
                names_to = "sampleid") %>% 
  # add sexes + lifr genotypes
  left_join( covarTidy.npc_rna) -> gsva_rna_results

gsva_rna_results %>% 
  group_by( Category) %>% 
  rstatix::anova_test( Enrichment_Score ~ sex+lifr_geno+sex*lifr_geno) %>% 
  rstatix::adjust_pvalue( method = "BH") %>%
  rstatix::add_significance("p.adj") %>% 
  ungroup() -> gsva_rna_aov

gsva_rna_results %>% 
  group_by(Category) %>% 
  rstatix::tukey_hsd( Enrichment_Score ~ sex+lifr_geno+sex:lifr_geno) %>% 
  ungroup() %>% 
  as_tibble() %>% 
  left_join( goannot_wdef, by = c("Category" = "GOID")) -> gsva_rna_tukey

gsva_rna_aov %>% 
  as_tibble() %>% 
  filter( p.adj.signif != "ns" ) -> signif_eff_terms_rna

gsva_rna_tukey %>% 
  inner_join( ., select( signif_eff_terms_rna, Category, term = Effect)) -> signif_results_tukey_rna

```

Below are the list of GO Biological processes that show significant differences by sex or genotype at the *Lifr* locus, or their interaction (adjusted p-value \< 0.01).

```{r gsva_results_table}

signif_results_tukey_rna %>%
  select(Effect= term, Category, TERM,group1, group2, estimate,p.adj) %>% 
  filter( p.adj < 0.01) %>% 
  distinct() %>% 
 mutate( "Adjusted p-value" = formatC(p.adj, digits=2, format ="e"),
          estimate= round(estimate,2)) %>% 
  select(-p.adj) %>% 
  arrange(estimate) %>% 
  create_dt()

```

### QTL mapping with GSVA scores

```{r gsva_qtl_scans}

gsva_rna_results %>%
  # filter( Category %in% (signif_results_tukey_rna %>% filter( term %in% c("sex","lifr_geno"),
  #                                                         p.adj.signif != "ns" ))$Category ) %>%
  select(Category, sampleid, Enrichment_Score) %>%
  pivot_wider( id_cols = sampleid , names_from = Category, values_from = Enrichment_Score) %>%
  column_to_rownames("sampleid") %>%
  as.matrix() -> gsva_results_signif_mat

# rankZ
gsva_results_signif_mat_rankZ <- apply(gsva_results_signif_mat, 2,rankZ )

# qtl mapping
# gsva_qtl <- scan1( genoprobs = probs.npc_rna,
#                    pheno = gsva_results_signif_mat_rankZ,
#                    kinship = kinship_loco.npc_rna,
#                    addcovar = covar.npc_rna)
# save( gsva_qtl, file = here("_data","GSVA_npc_qtl_scans.RData"))

load(here("_data","GSVA_npc_qtl_scans.RData"))
gsva_qtl_peaks <- find_peaks( gsva_qtl, threshold = 7, gmap)
# add interp_peak_bp, before, after
gsva_qtl_peaks <- gsva_qtl_peaks %>% 
  left_join( ., goannot_wdef %>% select(lodcolumn=GOID,TERM) %>% distinct()) %>% 
  mutate( TERM = ifelse( is.na(TERM), lodcolumn, TERM)) %>% 
  mutate(phenotype=lodcolumn) %>%
  mutate( peak_chr = chr,
          peak_cM = pos) %>%
  interp_bp(.) #add bp location for peaks

```


```{r}

gsva_qtl_peaks %>% 
  select( GOID = lodcolumn, TERM, peak_chr = chr, Lod = lod, peak_cM, peak_bp = interp_bp_peak) %>% 
  filter( GOID %in% (signif_results_tukey_rna %>% filter( term %in% c("sex","lifr_geno"),
                                                           p.adj.signif != "ns" ))$Category ) %>%
  mutate_if(is.numeric, round, 2) %>% 
  create_dt()

```

