---
title: ""
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
---

```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```


# Variation in NPC transcriptome {.tabset .tabset-fade .tabset-pills}

There are `r formatC(ncol(expr.npc_rna), big.mark = ",")` many genes measured across `r nrow(expr.npc_rna)` samples. 


```{r npc_stats, warning=FALSE, message=FALSE}

# npc_rna gene stats
var.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, var = V1) %>%
  arrange(desc(var))

n.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, n = V1) %>%
  arrange(desc(n))

mean.npc_rna <- expr.npc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, mean = V1) %>%
  arrange(desc(mean))

var.npc_rna <- inner_join(var.npc_rna, mean.npc_rna) %>%
  inner_join(., n.npc_rna) %>%
  left_join(., select(all.genes, ensembl_gene_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.npc_rna = 100 * sd / (mean)) %>%
  rename(mean.npc_rna = mean, sd.npc_rna = sd, n.npc_rna=n, var.npc_rna=var)

# npc stats per sample
var_npc_per_sample <- expr.npc_rna %>% 
  t() %>% 
  as_tibble( rownames = "ensembl_gene_id") %>%
  summarise( across( where(is.numeric),  var , na.rm=T )) %>% 
  pivot_longer( 1:ncol(.), names_to = "sampleid", values_to ="var_sample") %>% 
  left_join( select(covarTidy.npc_rna, sampleid, sex))
  

mean_npc_per_sample <- expr.npc_rna %>%
  t() %>% 
  as_tibble( rownames = "ensembl_gene_id") %>%
  summarise( across( where(is.numeric),  mean , na.rm=T )) %>% 
  pivot_longer( 1:ncol(.), names_to = "sampleid", values_to ="mean_sample") %>% 
  left_join( select(covarTidy.npc_rna, sampleid, sex))

```


## Overview

Below are the distributions of mean and variation of transcripts in DO mNPC lines. 

```{r npc_stat_plots_gene, fig.width=12, fig.height=4, warning=FALSE, message=FALSE}

# add mean + variance + CV plots per gene
var.npc_rna %>%
  ggplot() +
  aes(x = mean.npc_rna) +
  geom_histogram(bins = 200) +
  xlab("Mean") +
  theme_pubclean(base_size = 14)+
  scale_x_log10() -> p.mean.hist

var.npc_rna %>%
  ggplot() +
  aes(x = var.npc_rna) +
  geom_histogram(bins = 200 ) +
  xlab("Variance") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()  -> p.var.hist

var.npc_rna %>%
  ggplot() +
  aes(x = cv.npc_rna) +
  geom_histogram(bins = 200 ) +
  xlab("% Coefficient of variation") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()  -> p.cv.hist

ggscatter(var.npc_rna,
  x = "mean.npc_rna", y = "var.npc_rna", size = 3, alpha = 0.6,
  add = "reg.line", # Add regression line
  conf.int = TRUE, # Add confidence interval
  add.params = list(color = "blue", fill = "lightgray"), show.legend.text = FALSE,
  yscale = "log10", xscale = "log10"
) +
  stat_cor(method = "pearson", label.x = -1, label.y = 10) + # Add correlation coefficient
  xlab("Mean") +
  ylab("Variance") +
  theme_pubclean(base_size = 14) +
  rremove("legend") -> p.var.mean

ggarrange( p.mean.hist, p.var.hist, p.var.mean, nrow = 1)

```


Here are the distributions of mean and variation of DO mNPC lines.

```{r npc_stats_plots_sample, fig.width=16, fig.height=4, warning=FALSE, message=FALSE}
# add mean + variance + CV plots per sample
var_npc_per_sample %>% 
  ggplot() +
  aes(x = var_sample, col = sex, fill =sex) +
  geom_histogram(bins = 100 ) +
  xlab("Variance") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  color_palette("npg")+
  fill_palette("npg") -> p.var_sample.hist

mean_npc_per_sample %>% 
  ggplot() +
  aes(x = mean_sample, col = sex, fill =sex) +
  geom_histogram(bins = 100 ) +
  xlab("Mean") +
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  color_palette("npg")+
  fill_palette("npg") -> p.mean_sample.hist

full_join(var_npc_per_sample, mean_npc_per_sample) %>% 
  mutate(cv_sample = 100*(sqrt(var_sample))/mean_sample) %>%
  ggscatter(
    .,
  x = "mean_sample", y = "var_sample", size = 3, alpha = 0.6,
  add = "reg.line", # Add regression line
  conf.int = TRUE, # Add confidence interval
  col = "sex",
  add.params = list(color = "blue", fill = "lightgray"),
  show.legend.text = FALSE
  #yscale = "log10", xscale = "log10"
) +
  #stat_cor(method = "pearson", label.x = 10, label.y = 21) + # Add correlation coefficient
  xlab("Mean per sample") +
  ylab("Variance per sample") +
  theme_pubclean(base_size = 14) +
  color_palette("npg")+
  fill_palette("npg")+
  facet_wrap(~sex)+
  theme(
  strip.background = element_blank(),
  strip.text.x = element_blank()
  )-> p.var.mean_sample

ggarrange( p.var_sample.hist, p.mean_sample.hist, p.var.mean_sample, nrow = 1, widths = c(0.5,0.5,1))

```

## Principal component analysis

```{r npc_pca, warning=FALSE, message=FALSE}

pca.npc <- prcomp(expr.npc_rna, scale.=T, center=T)

```

Let's look at the first five PCs and how they group the data for the NPC transcriptome. I am coloring the points by sex: **red = Male** and **black = Female**. No separation by sex is observed in the NPC data! 

```{r pca_plot_pc1to5, fig.width=10, fig.height=8, warning=FALSE, message=FALSE}

pca.npc$x%>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(covarTidy.npc_rna, by = c("rowname" = "sampleid")) %>%
  mutate(sex = ifelse(sex == "F", "Female", "Male")) %>%
  GGally::ggpairs(.,
    columns = 2:6, progress = FALSE, ggplot2::aes(color = sex),
    upper = list(continuous = "density", combo = "box_no_facet"),
    lower = list(continuous = "points", combo = "dot_no_facet")
  ) + theme_pubclean(base_size = 14) +
  color_palette("npg")+
  fill_palette("npg")

```

```{r npc_scree_plot, warning=FALSE, message=FALSE}

# plot showing variance explained
var_explained <- (pca.npc$sdev^2 / sum(pca.npc$sdev^2))
tibble(var =  var_explained[1:10]* 100, PC = paste0("PC", seq(1:10))) %>%
  arrange(desc(var)) %>%
  mutate(label = factor(PC, levels = PC)) %>%
  ggplot(aes(x = label, y = var)) +
  geom_col() +
  xlab("Principal Component") +
  ylab("% Variation explained") +
  theme_pubclean(base_size = 14)+
  ylim(0,15)


```

```{r pc1_2_plot, fig.height=5, fig.width=6}


pca.npc$x%>%
  as.data.frame() %>%
  rownames_to_column() %>%
  left_join(covarTidy.npc_rna, by = c("rowname" = "sampleid")) %>%
  mutate(sex = ifelse(sex == "F", "Female", "Male")) %>% 
  ggplot()+
  aes(
    x = PC1, 
    y = PC2, 
    col = sex
  )+
  geom_point(size = 4, alpha = 0.7)+
  theme_pubclean(base_size = 20)+
  color_palette("npg")+
  xlab(paste0("PC1 (",100*round(var_explained[1],2),"%)"))+
  ylab(paste0("PC2 (",100*round(var_explained[2],2),"%)"))

```

Are PC1 drivers over-represented in any biological processes or pathways? Looking at the top 5% contributing genes of PC1.

```{r pc1_ora}

pc1_drivers <- pca.npc$rotation[,"PC1", drop = FALSE] %>% 
  as_tibble( rownames = "ensembl_gene_id") %>% 
  left_join( npc.genes %>%  
               select(ensembl_gene_id, mgi_symbol)
             ) %>% 
  filter( abs(PC1) >= quantile(abs(PC1), 0.95))

g.pc1 <- gost(
  query = pc1_drivers$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.pc1$result <- g.pc1$result %>% filter(term_size < 660)

```


```{r ora_pc1_results}

g.pc1$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

## Most and least variable genes in NPCs

```{r most_least_var}

high.var.genes <- var.npc_rna %>% 
  filter( cv.npc_rna >= quantile(cv.npc_rna, 0.95))

low.var.genes <- var.npc_rna %>% 
  filter( cv.npc_rna <= quantile(cv.npc_rna, 0.05))

var.npc_rna <- var.npc_rna %>%
  mutate(type = ifelse(ensembl_gene_id %in% high.var.genes$ensembl_gene_id, "high", "none")) %>%
  mutate(type = ifelse(ensembl_gene_id %in% low.var.genes$ensembl_gene_id, "low", type))

g.high.var <- gost(
  query = high.var.genes$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.high.var$result <- g.high.var$result %>% filter(term_size < 660)

g.low.var <- gost(
  query = low.var.genes$mgi_symbol,
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.low.var$result <- g.low.var$result %>% filter(term_size < 660)

```

ORA results with most variable (highest .05% CV) genes in mNPCs (FDR <0.01):

```{r ora_most_var}

g.high.var$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```

<br>
<br>

ORA results with least variable (lowest .05% CV) genes in mNPCs (FDR <0.01):

```{r ora_least_var}

g.low.var$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```


## Gene set variation analysis

```{r npc_rna_gsva, cache=TRUE, warning=FALSE, message=FALSE}

# reading in the GO + mgi downloaded from: http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt
go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt") %>% 
  mutate( genes = str_split(genes, ",")) %>% 
  unnest() # separete the symbols, note the overlap: length(intersect(unique(go_terms$genes), npc.genes$mgi_symbol) ) = 11806


slim_go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/map2MGIslim.txt") %>% 
  select(-term) %>% 
  mutate( ONT = case_when( aspect == "P" ~  "BP",
                     aspect == "F" ~ "MF",
                     aspect == "C" ~ "CC"
                     )
          ) %>% 
  select(-aspect)

genesbygo <- split(go_terms$genes, go_terms$GO_id)

go_terms_annot <- go_terms %>%  
  select(GO_id) %>% 
  distinct() %>% 
  left_join( slim_go_terms %>%  select( GO_id, ONT) %>% distinct())

goannot_wdef <- AnnotationDbi::select(GO.db, keys= unique(go_terms$GO_id), columns=c("GOID","DEFINITION","ONTOLOGY","TERM")) %>%
  left_join( slim_go_terms, by=c("GOID"="GO_id")) %>% 
  mutate( ONTOLOGY = ONT) %>% 
  select(-ONT)

go_bp <- goannot_wdef %>% filter( ONTOLOGY == "BP") %>% 
  select(GOID) %>%  distinct()

# expr - change gene id's to symbols for GSVA with GO
expr.npc_rna_upd <- expr.npc_rna[, npc.genes$ensembl_gene_id]
colnames(expr.npc_rna_upd) <- npc.genes$mgi_symbol


# I can't seem to run GSVA anymore! Keep erroring out. 
# I was able to run it separately in a stand alone R session.
# loading the results here instead of running.
load(here("_data","Results_from_GSVA_2022-09-23.RData"))
# gsva_rna <- gsva(  expr = t(expr.npc_rna_upd),
#                    gset.idx.list = genesbygo,
#                     method ="gsva",
#                     kcdf = "none",
#                     min.sz = 5, 
#                     max.sz = 1000,
#                     mx.diff = TRUE
#                    )


# following up on rna results
gsva_rna %>% 
  as_tibble(rownames = "Category") %>% 
  filter( Category %in% go_bp$GOID) %>% #filtering for BP
  pivot_longer( cols = rownames(expr.npc_rna_upd),
                values_to = "Enrichment_Score",
                names_to = "sampleid") %>% 
  # add sexes + lifr genotypes
  left_join( covarTidy.npc_rna) -> gsva_rna_results

gsva_rna_results %>% 
  group_by( Category) %>% 
  rstatix::anova_test( Enrichment_Score ~ sex+lifr_geno+sex*lifr_geno) %>% 
  rstatix::adjust_pvalue( method = "BH") %>%
  rstatix::add_significance("p.adj") %>% 
  ungroup() -> gsva_rna_aov

gsva_rna_results %>% 
  group_by(Category) %>% 
  rstatix::tukey_hsd( Enrichment_Score ~ sex+lifr_geno+sex:lifr_geno) %>% 
  ungroup() %>% 
  as_tibble() %>% 
  left_join( goannot_wdef, by = c("Category" = "GOID")) -> gsva_rna_tukey

gsva_rna_aov %>% 
  as_tibble() %>% 
  filter( p.adj.signif != "ns" ) -> signif_eff_terms_rna

gsva_rna_tukey %>% 
  inner_join( ., select( signif_eff_terms_rna, Category, term = Effect)) -> signif_results_tukey_rna

```

Below are the list of GO Biological processes that show significant differences by sex or genotype at the *Lifr* locus, or their interaction (adjusted p-value < 0.01). 

```{r gsva_results_table}

signif_results_tukey_rna %>%
  select(Effect= term, Category, TERM,group1, group2, estimate,p.adj) %>% 
  filter( p.adj < 0.01) %>% 
  distinct() %>% 
 mutate( "Adjusted p-value" = formatC(p.adj, digits=2, format ="e"),
          estimate= round(estimate,2)) %>% 
  select(-p.adj) %>% 
  arrange(estimate) %>% 
  create_dt()

```




