---
title: ""
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
---
    
```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```

# Covariation in ESC and NPC transcriptome {.tabset .tabset-fade .tabset-pills}

## Overlap between transcriptomes 

```{r,fig.width=8, fig.height=4, warning=FALSE, message=FALSE}


sample.overlap <- euler(c( "ESC" = nrow(covarTidy.esc_rna)-length(shared.samples), 
                           "NPC"=nrow(covarTidy.npc_rna)-length(shared.samples), 
                           "ESC&NPC"=length(shared.samples)) ,shape="ellipse")

gene.overlap <-  euler(c( "ESC" = nrow(esc.genes)-length(shared.genes), 
                           "NPC"= nrow(npc.genes)-length(shared.genes), 
                           "ESC&NPC"=length(shared.genes)) ,shape="ellipse")

p1 <- plot(sample.overlap,quantities = TRUE, main='Sample overlap',
           col=c(qtl.colors[["esc_rna"]],qtl.colors[["npc_rna"]],qtl.colors[["shared"]]), 
           fill=c(qtl.colors[["esc_rna"]],qtl.colors[["npc_rna"]],qtl.colors[["shared"]]),
           alpha=0.6)
p2 <- plot(gene.overlap,quantities = TRUE, main="Gene overlap",
           col=c(qtl.colors[["esc_rna"]],qtl.colors[["npc_rna"]],qtl.colors[["shared"]]), 
           fill=c(qtl.colors[["esc_rna"]],qtl.colors[["npc_rna"]],qtl.colors[["shared"]]),
           alpha=0.6)
ggarrange(p1,NULL,p2,nrow = 1, ncol=3 , widths = c(1, 0.1,0.9))  

```

```{r esc_rna_gene_stats, warning=FALSE, message=FALSE}

# esc_rna gene stats
var.esc_rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ var(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, var = V1) %>%
  arrange(desc(var))

n.esc_rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ sum(!is.na(.)))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, n = V1) %>%
  arrange(desc(n))

mean.esc_rna <- expr.esc_rna %>%
  as_tibble(.) %>%
  summarise_all(list(~ mean(., na.rm = T))) %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  rename(ensembl_gene_id = rowname, mean = V1) %>%
  arrange(desc(mean))

var.esc_rna <- inner_join(var.esc_rna, mean.esc_rna) %>%
  inner_join(., n.esc_rna) %>%
  left_join(., select(all.genes, ensembl_gene_id, mgi_symbol, gene_chr)) %>%
  mutate(sd = sqrt(var)) %>%
  mutate(cv.esc_rna = 100 * sd / (mean)) %>%
  rename(mean.esc_rna = mean, sd.esc_rna = sd, n.esc_rna=n, var.esc_rna=var)

```

```{r stat_plots_esc_npc, fig.height=4, fig.width=9, warning=FALSE, message=FALSE}

# mean esc vs npc
var.esc_rna %>% 
  inner_join( var.npc_rna) %>% 
  ggplot()+
  aes( x = mean.esc_rna, 
       y = mean.npc_rna  )+
  geom_point()+
  geom_smooth( method= "lm")+
  stat_cor()+
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  scale_y_log10()+
  xlab("Mean transcript abundance in ESCs")+
  ylab("Mean transcript abundance in NPCs") -> p.mean

# variance esc vs npc
var.esc_rna %>% 
  inner_join( var.npc_rna) %>% 
  ggplot()+
  aes( x = var.esc_rna, 
       y = var.npc_rna  )+
  geom_point()+
  geom_smooth( method= "lm")+
  stat_cor()+
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  scale_y_log10()+
  xlab("Variance in transcript abundance in ESCs")+
  ylab("Variance transcript abundance in NPCs") -> p.var


# CV esc vs npc
var.esc_rna %>% 
  inner_join( var.npc_rna) %>% 
  ggplot()+
  aes( x = cv.esc_rna, 
       y = cv.npc_rna  )+
  geom_point()+
  geom_smooth( method= "lm")+
  stat_cor()+
  theme_pubclean(base_size = 14)+
  scale_x_log10()+
  scale_y_log10()+
  xlab("%CV in transcript abundance in ESCs")+
  ylab("%CV transcript abundance in NPCs") -> p.cv

ggarrange(p.mean, p.cv,  nrow = 1)

```


## Correlation between transcriptomes 


```{r sample_corr, warning=FALSE, message=FALSE}

shared.expr.esc_rna1 <- t(exprZ.esc_rna[shared.samples, shared.genes])
shared.expr.npc_rna1 <- t(exprZ.npc_rna[shared.samples, shared.genes])
colnames(shared.expr.esc_rna1) <- paste0(shared.samples,".esc_rna")
colnames(shared.expr.npc_rna1) <- paste0(shared.samples,".npc_rna")
rankZ_cor_sp <- rcorr(shared.expr.esc_rna1,
                    shared.expr.npc_rna1,
                    type = "spearman")

shared.expr.esc_rna2 <- t(expr.esc_rna[shared.samples, shared.genes])
shared.expr.npc_rna2 <- t(expr.npc_rna[shared.samples, shared.genes])
colnames(shared.expr.esc_rna2) <- paste0(shared.samples,".esc_rna")
colnames(shared.expr.npc_rna2) <- paste0(shared.samples,".npc_rna")
expr_cor_sp <- rcorr(shared.expr.esc_rna2,
                    shared.expr.npc_rna2,
                    type = "pearson")

```

Pearson correlation between ESC and NPC transcriptomes using upper quartile normalized, batch corrected transcript abundance values.

```{r sample_cor_hist, fig.width=6, fig.height=4, warning=FALSE, message=FALSE}

sample_cor_df <- as_tibble( rankZ_cor_sp$r[colnames(shared.expr.esc_rna1), colnames(shared.expr.npc_rna1)],
                            rownames = "esc") %>%
  pivot_longer( colnames(shared.expr.npc_rna1), names_to = "npc", values_to = "cor_rankz") %>%
  inner_join( (as_tibble( rankZ_cor_sp$P[colnames(shared.expr.esc_rna1), colnames(shared.expr.npc_rna1)],
                            rownames = "esc") %>%
      pivot_longer( colnames(shared.expr.npc_rna1), names_to = "npc", values_to = "p_val_rankz") ) ) %>% 
  inner_join(
    as_tibble(
      expr_cor_sp$r[colnames(shared.expr.esc_rna2), colnames(shared.expr.npc_rna2)],
                            rownames = "esc") %>% 
        pivot_longer( colnames(shared.expr.npc_rna2), names_to = "npc", values_to = "cor_expr")
    ) %>% 
    inner_join( (as_tibble( expr_cor_sp$P[colnames(shared.expr.esc_rna2), colnames(shared.expr.npc_rna2)],
                            rownames = "esc") %>%
      pivot_longer( colnames(shared.expr.npc_rna2), names_to = "npc", values_to = "p_val_expr") ) ) 
  


sample_cor_df %>% 
  mutate( sampleid_esc = gsub(".esc_rna","",esc),
          sampleid_npc = gsub(".npc_rna","", npc)) %>%
  filter( sampleid_esc == sampleid_npc) %>% 
  mutate( sampleid = sampleid_npc) %>% 
  #left_join(covarTidy.npc_rna %>%  select(sampleid, sex)) %>% 
  ggplot()+
  aes( x = cor_expr)+
  #geom_point()
  geom_histogram(binwidth = 0.01, alpha = 0.8)+
  theme_pubclean(base_size = 18)+
  color_palette("npg")+
  fill_palette("npg")+
  xlim(0,1)+
  xlab("Correlation")+
  ylab("")


```


## Correlation between genes

Add here - gene based correlations, ORA with significantly positively correlated + negatively correlated + no correlation genes

```{r gene_cor, eval = FALSE}

shared.expr.esc_rna3 <- shared.expr.esc_rna
colnames(shared.expr.esc_rna3) <- paste0(colnames(shared.expr.esc_rna3),"_esc")
shared.expr.npc_rna3 <- shared.expr.npc_rna
colnames(shared.expr.npc_rna3) <- paste0(colnames(shared.expr.npc_rna3),"_npc")

gene.cor <- rcorr(shared.expr.npc_rna3, shared.expr.esc_rna3, type="pearson")

gene_cor_df <- tibble( cor =diag(gene.cor$r[colnames(shared.expr.npc_rna3), colnames(shared.expr.esc_rna3)]),
                       p_val = diag(gene.cor$P[colnames(shared.expr.npc_rna3), colnames(shared.expr.esc_rna3)]),
                       n = diag(gene.cor$n[colnames(shared.expr.npc_rna3), colnames(shared.expr.esc_rna3)]),
                       ensembl_gene_id = gsub("_npc","",colnames(shared.expr.npc_rna3))
                       ) %>%
  left_join(., all.genes) %>%
  mutate(p_adj = p.adjust(p_val, method = "BH"))

save(gene_cor_df, file = here("_data","gene_cor_df.RData"))

```

```{r gene_cor_plot, warning=FALSE, message=FALSE, fig.height=4, fig.width=6}
load(here("_data","gene_cor_df.RData"))

gene_cor_df %>% 
  mutate(p_val_rank = ifelse(p_adj < 0.05, "p < 0.05", "ns") ) %>%
  ggplot() +
  aes(x = cor) +
  geom_histogram(aes(fill = p_val_rank), show.legend = T, bins = 300, alpha = 0.6) +
  xlab("Correlation") +
  ylab("")+
  scale_fill_viridis_d(direction = -1, limits = c("ns", "p < 0.05")) +
  labs(fill = "Significance") +
  #ggtitle("Transcript vs Protein abundance for genes") +
  xlim(-0.5,1)+
  theme_pubclean(base_size = 18)+
  theme( legend.text = element_text(size = 14),
         legend.title = element_text(size = 16))

```


```{r gene_cor_table, warning=FALSE, message=FALSE}

gene_cor_df %>%
  mutate(p_val_rank = ifelse(p_adj < 5e-10, "p < 5e-10", ifelse(p_adj < 0.005, "p < 0.005", ifelse(p_adj < 0.05, "p < 0.05", "ns")))) %>%
  select(mgi_symbol, ensembl_gene_id, gene_chr, gene_start, gene_end, cor, p_adj, p_val_rank) %>%
  mutate(p_adj = formatC(p_adj, digits = 2, format = "e"), cor = formatC(cor, digits = 2, format = "g")) %>%
  create_dt()

```


```{r gene_cor_ora, warning=FALSE, message=FALSE}

neg_cor <- gene_cor_df %>%
  filter( cor < 0, p_adj < 0.05)

pos_cor <- gene_cor_df %>%
  filter( cor > 0, p_adj < 0.05) %>%
  arrange( desc(cor) )

no_cor <- gene_cor_df %>%
  filter( abs(cor) < 0.05) 

ora_neg_cor <- gost( query = neg_cor$mgi_symbol,
                     organism = "mmusculus",
                     domain_scope = "custom",
                     custom_bg = gene_cor_df$mgi_symbol,
                     evcodes = TRUE,
  correction_method = "fdr"
                     )
ora_neg_cor$result <- filter( ora_neg_cor$result, term_size < 600)
ora_pos_cor <- gost( query = pos_cor$mgi_symbol,
                     organism = "mmusculus",
                     domain_scope = "custom",
                     custom_bg = gene_cor_df$mgi_symbol,
                     evcodes = TRUE,
  correction_method = "fdr"
                     )
ora_pos_cor$result <- filter( ora_pos_cor$result, term_size < 600)

ora_no_cor <- gost( query = no_cor$mgi_symbol,
                     organism = "mmusculus",
                     domain_scope = "custom",
                     custom_bg = gene_cor_df$mgi_symbol,
                     evcodes = TRUE,
  correction_method = "fdr"
                     )
ora_no_cor$result <- filter( ora_no_cor$result, term_size < 600)

ora_neg_cor$result %>%
  mutate( group = "Negative correlation") %>%
  rbind( mutate( ora_pos_cor$result, group = "Positive correlation")) %>%
  rbind( mutate( ora_no_cor$result , group = "No correlation")) -> ora_all_corr

ora_all_corr %>%
  select(group,source, term_name, intersection_size, term_size , FDR = p_value) %>%
  #mutate( p_value = p.adjust(p_value, method = "BH") ) %>% # already corrected
  filter( FDR < 0.05) %>%
  mutate_if(is.numeric, formatC, digits =2, format ="fg") %>%
  create_dt()

```





## NPC specific genes

Here are the overrepresentation results using genes only expressed in NPC lines using all of the genes identified in ESC and NPC lines as the custom background.

```{r npc_specific_genes_ora, cache = TRUE, warning=FALSE, message=FALSE}

# get npc specific genes
all.genes %>% 
  filter( ensembl_gene_id %in% npc.genes$ensembl_gene_id & !ensembl_gene_id %in% esc.genes$ensembl_gene_id) -> npc.specific.genes

npc_specific_ora <- gost(query = npc.specific.genes$mgi_symbol, 
                 organism = "mmusculus",
                 significant = TRUE, 
                 domain_scope = "custom",
                 custom_bg = all.genes$mgi_symbol,
                 correction_method = "fdr",
                 evcodes = TRUE)
npc_specific_ora$result <- npc_specific_ora$result %>% filter(term_size <500)


```

```{r npc_specific_ora_plot, warning=FALSE, message=FALSE}

gostplot(npc_specific_ora)

```


```{r npc_specific_ora_table, warning=FALSE, message=FALSE}

npc_specific_ora$result %>% 
  select( term_name, source,FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```



## ESC specific genes

Here are the overrepresentation results using genes only expressed in ESC lines using all of the genes identified in ESC and NPC lines as the custom background.


```{r esc_specific_genes_ora, cache=TRUE, warning=FALSE, message=FALSE}

# get esc specific genes
all.genes %>% 
  filter( !ensembl_gene_id %in% npc.genes$ensembl_gene_id & ensembl_gene_id %in% esc.genes$ensembl_gene_id) -> npc.specific.genes-> esc.specific.genes

esc_specific_ora <- gost(query = esc.specific.genes$mgi_symbol, 
                 organism = "mmusculus",
                 significant = TRUE, 
                 domain_scope = "custom",
                 custom_bg = all.genes$mgi_symbol,
                 correction_method = "fdr",
                 evcodes = TRUE)
esc_specific_ora$result <- esc_specific_ora$result %>% filter(term_size <500)


```

```{r esc_specific_ora_gostplot, warning=FALSE, message=FALSE}

gostplot(esc_specific_ora, capped = FALSE)

```


```{r esc_specific_ora_table, warning=FALSE, message=FALSE}

esc_specific_ora$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```


## Gene Set Variation Analysis of combined ESC-NPC Data set

```{r esc_npc_rna_gsva, cache=TRUE, warning=FALSE, message=FALSE}


# reading in the GO + mgi downloaded from: http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt
go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/MGIgenes_by_GOid.txt") %>% 
  mutate( genes = str_split(genes, ",")) %>% 
  unnest() # separete the symbols, note the overlap: length(intersect(unique(go_terms$genes), all.prots$mgi_symbol) ) = 6757


slim_go_terms <- read_tsv( "http://www.informatics.jax.org/gotools/data/input/map2MGIslim.txt") %>% 
  select(-term) %>% 
  mutate( ONT = case_when( aspect == "P" ~  "BP",
                     aspect == "F" ~ "MF",
                     aspect == "C" ~ "CC"
                     )
          ) %>% 
  select(-aspect)

genesbygo <- split(go_terms$genes, go_terms$GO_id)

go_terms_annot <- go_terms %>%  
  select(GO_id) %>% 
  distinct() %>% 
  left_join( slim_go_terms %>%  select( GO_id, ONT) %>% distinct())

goannot_wdef <- AnnotationDbi::select(GO.db, keys= unique(go_terms$GO_id), columns=c("GOID","DEFINITION","ONTOLOGY","TERM")) %>%
  left_join( slim_go_terms, by=c("GOID"="GO_id")) %>% 
  mutate( ONTOLOGY = ONT) %>% 
  select(-ONT)

go_bp <- goannot_wdef %>% filter( ONTOLOGY == "BP") %>% 
  select(GOID) %>%  distinct()

# expr - change protein id's to symbols for GSVA with GO
shared.genes2 <- all.genes %>% 
  filter(ensembl_gene_id %in% shared.genes)
shared.expr.npc_rna_upd <- shared.expr.npc_rna[, shared.genes2$ensembl_gene_id]
colnames(shared.expr.npc_rna_upd) <- shared.genes2$mgi_symbol
rownames(shared.expr.npc_rna_upd) <- paste0( rownames(shared.expr.npc_rna_upd),"_npc")
shared.expr.esc_rna_upd <- shared.expr.esc_rna[,shared.genes2$ensembl_gene_id]
colnames(shared.expr.esc_rna_upd) <- shared.genes2$mgi_symbol
rownames(shared.expr.esc_rna_upd) <- paste0( rownames(shared.expr.esc_rna_upd),"_esc")

shared.expr.merged <- rbind(shared.expr.npc_rna_upd, 
                            shared.expr.esc_rna_upd)

# I can't seem to run GSVA anymore! Keep erroring out. 
# I was able to run it separately in a stand alone R session.
# loading the results here instead of running.
load(here("_data","Results_from_GSVA_2022-09-23.RData"))
# gsva_rna_esc_npc <- gsva(  expr = t(shared.expr.merged),
#                     genesbygo,
#                     method ="gsva",
#                     kcdf = "none",
#                     min.sz = 5, 
#                     max.sz = 1000,
#                     mx.diff = TRUE)

# following up on rna results
gsva_rna_esc_npc %>% 
  as_tibble(rownames = "Category") %>% 
  #filter( Category %in% go_bp$GOID) %>% #filtering for BP
  pivot_longer( cols = rownames(shared.expr.merged),
                values_to = "Enrichment_Score",
                names_to = "sample") %>% 
  separate( sample, into = c("temp1","temp2", "type"), sep = "_") %>% 
  unite( sampleid, c(temp1, temp2), sep = "_") %>% 
  # add sexes + lifr genotypes
  left_join( covarTidy.npc_rna) -> gsva_rna_esc_npc_results

gsva_rna_esc_npc_results %>% 
  group_by( Category) %>% 
  rstatix::anova_test( Enrichment_Score ~ type+sex+type*sex) %>% 
  rstatix::adjust_pvalue( method = "BH") %>%
  rstatix::add_significance("p.adj") %>% 
  ungroup() -> gsva_rna_esc_npc_aov

gsva_rna_esc_npc_results %>% 
  group_by(Category) %>% 
  rstatix::tukey_hsd( Enrichment_Score ~ type+sex+type*sex) %>% 
  ungroup() %>% 
  as_tibble() %>% 
  left_join( goannot_wdef, by = c("Category" = "GOID")) -> gsva_rna_esc_npc_tukey

gsva_rna_esc_npc_aov %>% 
  as_tibble() %>% 
  filter( p.adj.signif != "ns" ) -> signif_eff_terms_rna_esc_npc

gsva_rna_esc_npc_tukey %>% 
  inner_join( ., select( signif_eff_terms_rna_esc_npc, Category, term = Effect)) -> signif_results_tukey_rna_esc_npc

```

Below are the list of GO Biological processes that show significant differences by cell type or sex(adjusted p-value <0.01).

```{r gsva_results_table}

signif_results_tukey_rna_esc_npc %>%
  filter( Category %in% go_bp$GOID) %>%
  filter( term != "type:sex") %>% 
  select(Effect= term, Category, TERM, group1, group2, estimate,p.adj) %>% 
  filter( p.adj < 0.01) %>% 
  distinct() %>% 
 mutate( "Adjusted p-value" = formatC(p.adj, digits=2, format ="e"),
          estimate= round(estimate,2)) %>% 
  select(-p.adj) %>% 
  arrange(estimate) %>% 
  create_dt()

```

Examples:

```{r gsva_examples, fig.width=8, fig.height=4}

# ribosome biogenesis GO:0042254
# translation GO:0006412
# GO:0010564	regulation of cell cycle process
# GO:0006094	gluconeogenesis
# GO:0001825	blastocyst formation

# GO:0030182	neuron differentiation
# GO:0048699	generation of neurons
# GO:0048666	neuron development
# GO:0021954	central nervous system neuron development
# GO:0045995	regulation of embryonic development
# GO:0097150	neuronal stem cell population maintenance
# GO:0007417	central nervous system development

gsva_rna_esc_npc_results %>%
  filter( Category ==  "GO:0007417") %>%
  left_join( select(goannot_wdef, Category = GOID, TERM)) %>% 
  ggplot()+
  aes( x = type,
       y = Enrichment_Score,
       col = type)+
  geom_boxplot(width =0.2, size = 1.1)+
  #geom_jitter()+
  #geom_beeswarm(aes(col = sex))+
  theme_pubclean(base_size = 16)+
  stat_pvalue_manual( filter(signif_results_tukey_rna_esc_npc,Category  == "GO:0007417", term == "type"),
                      label = "{p.adj.signif}",
                      y.position = 0.85)+
  color_palette("jco")+
  ylab("Enrichment Score")+
  ggtitle("Central nervous system development")+
  xlab("")+
  ylim(-1,1)+
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 16),
        axis.title = element_text(size =16),
        title = element_text(size =10)) -> plot_nerv

gsva_rna_esc_npc_results %>%
  filter( Category ==  "GO:0006094") %>%
  left_join( select(goannot_wdef, Category = GOID, TERM)) %>% 
  ggplot()+
  aes( x = type,
       y = Enrichment_Score,
       col = type)+
  geom_boxplot(width =0.2, size = 1.1)+
  #geom_jitter()+
  #geom_beeswarm(aes(col = sex))+
  theme_pubclean(base_size = 16)+
  stat_pvalue_manual( filter(signif_results_tukey_rna_esc_npc,Category  == "GO:0006094", term == "type"),
                      label = "{p.adj.signif}",
                      y.position = 0.85)+
  color_palette("jco")+
  ylab("Enrichment Score")+
  ggtitle("Gluconeogenesis")+
  xlab("")+
  ylim(-1,1)+
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 16),
        axis.title = element_text(size =16),
        title = element_text(size =10)) -> plot_glucogenesis

gsva_rna_esc_npc_results %>%
  filter( Category ==  "GO:0006412") %>%
  left_join( select(goannot_wdef, Category = GOID, TERM)) %>% 
  ggplot()+
  aes( x = type,
       y = Enrichment_Score,
       col = type)+
  geom_boxplot(width =0.2, size = 1.1)+
  #geom_jitter()+
  #geom_beeswarm(aes(col = sex))+
  theme_pubclean(base_size = 16)+
  stat_pvalue_manual( filter(signif_results_tukey_rna_esc_npc,Category  == "GO:0006412", term == "type"),
                      label = "{p.adj.signif}",
                      y.position = 0.85)+
  color_palette("jco")+
  ylab("Enrichment Score")+
  ggtitle("Translation")+
  xlab("")+
  ylim(-1,1)+
  theme(axis.text.x = element_text(size = 20),
        axis.text.y = element_text(size = 16),
        axis.title = element_text(size =16),
        title = element_text(size =10)) -> plot_translation

ggarrange( plot_nerv, plot_translation, common.legend =  TRUE, legend = "none", nrow = 1)

```




