---
title: ""
output:
  html_document:
    toc: false
    df_print: paged
    code_folding: hide
---

```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```

```{r calc_herit}

# calculate heritability
kinship.npc_rna <- calc_kinship( probs.npc_rna)
npc_herit <-  est_herit(
  pheno = exprZ.npc_rna,
  kinship = kinship.npc_rna, 
  addcovar = covar.npc_rna
)

kinship.esc_rna <- calc_kinship( probs.esc_rna)
esc_herit <- est_herit(
  pheno = exprZ.esc_rna,
  kinship = kinship.esc_rna,
  addcovar = covar.esc_rna
)

```

# Genetic architecture of the NPC transcriptome {.tabset .tabset-fade .tabset-pills}

Out of `r formatC(ncol(exprZ.npc_rna), big.mark =",")` transcripts, `r npc_herit %>% as_tibble(rownames= "ensembl_gene_id") %>% filter(value >0) %>% nrow()` has non-zero heritability and the median heritability is `r round((npc_herit %>% as_tibble(rownames= "ensembl_gene_id") %>% summarize( mean = median(value)))$mean,2)`. We mapped `r formatC((peaks.npc_rna %>% filter(lod >7.5) %>% nrow()),big.mark=",")` eQTL peaks from `r formatC((peaks.npc_rna %>% filter(lod >7.5) %>% select(phenotype) %>% distinct() %>% nrow()),big.mark=",")` unique genes in NPCs with a lod score above 7.5 where `r formatC((peaks.npc_rna.wEffs %>% filter(lod.npc_rna >7.5, local.npc_rna ==T) %>% nrow()),big.mark=",")` are local and `r formatC((peaks.npc_rna.wEffs %>% filter(lod.npc_rna >7.5, local.npc_rna ==F) %>% nrow()),big.mark=",")` are distant. Distant pQTL form hotspots on chromosomes 1, 10 and 11.

## Heritability of transcript abundance in DO mNPCs

```{r heritability_npc_rna, fig.width=6, fig.height=4, message=FALSE, warning=FALSE}

# plot heritability histogram
npc_herit %>%
  as_tibble(rownames= "ensembl_gene_id") %>%
  left_join( all.genes) %>%
  ggplot()+
  aes(
    x = value
    )+
  geom_histogram( alpha = 0.6, binwidth  = 0.01)+
  theme_pubclean(base_size = 14)+
  ylab("") +
  xlab("Heritability of transcript abundance in DO mNPCs")

```

## DO mNPC eQTL map

Using +/- 10Mb for classifying local/distant eQTL.

```{r npc_eqtl_map_prep, warning=FALSE, message=FALSE}

# eQTL map for NPC rna
# prep the objects
#map_dat2 <- map_dat2 %>% mutate(pos_bp = as.numeric(pos_bp), pos_cM = as.numeric(pos))
map_dat2$chromF <- factor(map_dat2$chrom, levels = c(as.character(1:19), "X"))
chrom_markers <- select(map_dat2, chromF, n) %>%
  rename(chrom = chromF) %>%
  group_by(chrom) %>%
  summarize(start = min(n), end = max(n)) %>%
  GenomicRanges::GRanges()
windows <- unlist(GenomicRanges::slidingWindows(chrom_markers, width = 50, step = 10))
markers_bynum <- select(map_dat2, chrom, n) %>%
  dplyr::rename(start = n) %>%
  mutate(end = start) %>%
  GenomicRanges::GRanges()

distant_esc_rna <- filter(peaks.esc_rna.wEffs, lod.esc_rna > 7.5, !is.na(local.esc_rna) & !(local.esc_rna)) %>%
  select(peak_chr, interp_bp_peak.esc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.esc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()
distant_esc_rna_lod6 <- filter(peaks.esc_rna.wEffs, lod.esc_rna > 6, !is.na(local.esc_rna) & !(local.esc_rna)) %>%
  select(peak_chr, interp_bp_peak.esc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.esc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()

distant_npc_rna <- filter(peaks.npc_rna.wEffs, lod.npc_rna > 7.5, !is.na(local.npc_rna) & !(local.npc_rna)) %>%
  select(peak_chr, interp_bp_peak.npc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.npc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()
distant_npc_rna_lod6 <- filter(peaks.npc_rna.wEffs, lod.npc_rna > 6, !is.na(local.npc_rna) & !(local.npc_rna)) %>%
  select(peak_chr, interp_bp_peak.npc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.npc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()

markers <- select(map_dat2, chrom, pos_bp) %>%
  dplyr::rename(start = pos_bp) %>%
  mutate(end = start) %>%
  GenomicRanges::GRanges() # length 69,005
esc_hotspot <- GenomicRanges::nearest(distant_esc_rna, markers)
esc_hotspot_lod6 <- GenomicRanges::nearest(distant_esc_rna_lod6, markers)
npc_hotspot <- GenomicRanges::nearest(distant_npc_rna, markers)
npc_hotspot_lod6 <- GenomicRanges::nearest(distant_npc_rna_lod6, markers)
# assert_that(noNA(x), noNA(y))
# assert_that(noNA(x))
windows$distant_esc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[esc_hotspot])
windows$distant_esc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[esc_hotspot_lod6])
windows$distant_npc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[npc_hotspot])
windows$distant_npc_rna_lod6 <- GenomicRanges::countOverlaps(windows, markers_bynum[npc_hotspot_lod6])
window_counts <- tibble(
  chrom = as.character(GenomicRanges::seqnames(windows)),
  start = GenomicRanges::start(windows),
  end = GenomicRanges::end(windows),
  distant_esc_rna = windows$distant_esc_rna, 
  distant_npc_rna = windows$distant_npc_rna,
  distant_npc_rna_lod6 = windows$distant_npc_rna_lod6
)

# plotting
map_dat2 <- map_dat2 %>% mutate( pos_cM = as.numeric(pos))
mm <- match(window_counts$start, map_dat2$n)
m2 <- match(window_counts$end, map_dat2$n)
window_counts$pos_cM_start <- map_dat2$pos_cM[mm]
window_counts$pos_bp_start <- map_dat2$pos_bp[mm]
window_counts$pos_cM_end <- map_dat2$pos_cM[m2]
window_counts$pos_bp_end <- map_dat2$pos_bp[m2]
window_counts <- window_counts %>%
  mutate(midpoint = (pos_cM_end + pos_cM_start) / 2, 4)

x <- select(window_counts, chrom, starts_with("pos_bp"), starts_with("distant")) %>%
  filter(
    #distant_esc_rna > quantile(distant_esc_rna,0.995) | 
    #distant_npc_rna_lod6 > quantile(distant_npc_rna_lod6, 0.995) | 
           distant_npc_rna > quantile(distant_npc_rna,0.995) )
bands.esc.npc.rna <- x %>%
  rename(start = pos_bp_start, end = pos_bp_end) %>%
  GenomicRanges::GRanges() %>%
  GenomicRanges::reduce()
# reduce collapses overlapping windows into one big window. Works perfectly here.
bands.esc.npc.rna$distant_esc_rna <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_esc_rna)
bands.esc.npc.rna$distant_esc_rna_lod6<- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_esc_rna_lod6)
bands.esc.npc.rna$distant_npc_rna <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_npc_rna)
bands.esc.npc.rna$distant_npc_rna_lod6 <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_npc_rna_lod6)

```

```{r npc_eqtl_map, fig.height=11, fig.width=8, warning=FALSE, message=FALSE}

peaks.npc_rna.wEffs$cumsum_bp_peak <- peaks.npc_rna.wEffs$interp_bp_peak + chrom_lens_offset[peaks.npc_rna.wEffs$peak_chr]
peaks.npc_rna.wEffs$cumsum_bp_gene <- peaks.npc_rna.wEffs$midpoint + chrom_lens_offset[peaks.npc_rna.wEffs$gene_chr]

chroms <-names(chrom_lens)
chrom_segments <- tibble( start = 0, 
                          end = chrom_lens,
                          chr = chroms,
                          type = as.character(rep(c(0,1),10)))
chrom_segments$start <- chrom_segments$start+ chrom_lens_offset[chrom_segments$chr]
chrom_segments$end <- chrom_segments$end+ chrom_lens_offset[chrom_segments$chr]

ggplot()+
  geom_rect( data = chrom_segments, aes( xmin =start, xmax = end, ymin = 0, ymax = max(end), fill = type), 
             inherit.aes = FALSE, alpha = 0.2, show.legend = FALSE)+
  scale_fill_manual(values = c("dark gray","white"))+
  geom_point(data = peaks.npc_rna.wEffs %>% 
            filter( lod.npc_rna > 7.5), 
            aes( x = cumsum_bp_peak, y = cumsum_bp_gene),
            size = 2, 
            col =qtl.colors[["npc_rna"]],
             inherit.aes = FALSE )+
  theme_pubclean(base_size = 16)+
  scale_x_discrete( name = "eQTL peak",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.02))+
  scale_y_discrete( name = "Gene midpoint",limits = chrom_lens_midpt, labels = names(chrom_lens), expand = expansion( mult = 0.02))+
  theme( axis.text = element_text(size = 10),
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_blank()) -> npc_eqtl_plot



bands.esc.npc.rna %>% 
  as_tibble() %>% 
  select( chrom = seqnames, start, end, distant_npc_rna) %>% 
  mutate( hotspot_midpoint = (start+end)/2 ) %>% 
  # adding all the marker locations to match axes
  rbind( (map_dat2 %>% 
              select( chrom, start = pos_bp, end =pos_bp) %>% 
              mutate( distant_npc_rna = 0,
                      hotspot_midpoint = start))) %>% 
  mutate( chrom = factor(chrom, levels = c(seq(1:19),"X")) ) -> npc_eqtl_counts

# adding all the markers with 0 hotspot values to match the axes
npc_eqtl_counts$midpoint_offset <- npc_eqtl_counts$hotspot_midpoint + chrom_lens_offset[npc_eqtl_counts$chrom]

npc_eqtl_counts %>% 
  ggplot()+
  aes( x = midpoint_offset, 
       y = distant_npc_rna)+
  geom_bar( stat = "identity", width = 100, col =qtl.colors[["npc_rna"]], fill= qtl.colors[["npc_rna"]] )+
  theme_pubclean(base_size = 16)+
  scale_x_continuous( name = "Chr",
                      breaks = chrom_lens_midpt, 
                      labels = names(chrom_lens), expand = expansion(mult = .02) )+
  xlab("")+
  ylab("# of distant eQTL")+
  theme( axis.text = element_text(size = 10)) -> trans_band_plot


npc_eqtl_map_plot <- ggarrange( npc_eqtl_plot, trans_band_plot, nrow = 2, heights = c(0.7, 0.3))

# ggsave(npc_eqtl_map_plot, filename = here("figures","figure3a.svg"), width = 8, height = 11, dpi = 300, units = "in")
npc_eqtl_map_plot

```


# Comparison of eQTL maps of DO mESC and mNPC {.tabset .tabset-fade .tabset-pills}

## Heritability of transcript abundance ESC vs NPC lines

```{r herit_comp, fig.height=4, fig.width=5}

npc_herit %>%
  as_tibble(rownames= "ensembl_gene_id") %>%
  rename( "npc_h" = "value") %>% 
  inner_join(
    esc_herit %>% 
      as_tibble( rownames = "ensembl_gene_id") %>% 
      rename( "esc_h" = "value")
  ) %>% 
  left_join( all.genes) %>% 
  ggplot()+
  aes(
    x= esc_h,
    y = npc_h
  )+
  geom_point(alpha = 0.5)+
  geom_smooth( method= "lm")+
  stat_cor()+
  xlab("ESC heritability")+
  ylab("NPC heritability")+
  theme_pubclean(base_size = 16)




```

## DO mESC and mNPC eQTL map: Shared genes and all eQTL

Using +/- 10Mb for looking at overlap. 

```{r esc_npc_eqtl_map, fig.height=8, fig.width=8, warning=FALSE, message=FALSE}


ggplot()+
  geom_rect( data = chrom_segments, aes( xmin =start, xmax = end, ymin = 0, ymax = max(end), fill = type), 
             inherit.aes = FALSE, alpha = 0.2, show.legend = FALSE)+
  scale_fill_manual(values = c("dark gray","white"))+
  geom_point(data = peaks.npc_rna.wesc.overlap %>% 
               filter( lod.npc_rna > 7.5, ensembl_gene_id %in% shared.genes, match =="npc_rna"), 
            aes( x = cumsum_bp_peak.npc_rna, y = cumsum_bp_gene,col="NPC"),
            size = 2, 
            #col = qtl.colors[["npc_rna"]],
             inherit.aes = FALSE )+
  geom_point(data = peaks.esc_rna.wnpc.overlap %>% 
               filter( lod.esc_rna > 7.5, ensembl_gene_id %in% shared.genes, match =="esc_rna"), 
            aes( x = cumsum_bp_peak.esc_rna, y = cumsum_bp_gene, col ="ESC"),
            size = 2, 
            #col = qtl.colors[["esc_rna"]],
             inherit.aes = FALSE )+
  geom_point(data = peaks.npc_rna.wesc.overlap %>% 
               filter( lod.npc_rna > 7.5, ensembl_gene_id %in% shared.genes, match =="shared"), 
            aes( x = cumsum_bp_peak.npc_rna, y = cumsum_bp_gene,col="Shared"),
            size = 2, 
            #col = qtl.colors[["shared"]],
             inherit.aes = FALSE )+
  theme_pubclean(base_size = 16)+
  scale_x_discrete( name = "eQTL peak",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.02))+
  scale_y_discrete( name = "Gene midpoint",limits = chrom_lens_midpt, labels = names(chrom_lens), expand = expansion( mult = 0.02))+
  theme( axis.text = element_text(size = 10),
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_blank())+
  scale_color_manual( name = "eQTL",values = c("NPC"=qtl.colors[["npc_rna"]], "ESC"=qtl.colors[["esc_rna"]], "Shared"=qtl.colors[["shared"]]), labels = c("NPC","ESC","Shared"))

```


## DO mNPC eQTL map integrated: Shared genes and only significant NPC eQTL 

```{r esc_npc_qtl_map_updated, fig.width=8,fig.height=11, warning=FALSE, message=FALSE}

npc_eqtl_shared_plot <- ggplot()+
  geom_rect( data = chrom_segments, aes( xmin =start, xmax = end, ymin = 0, ymax = max(end), fill = type), 
             inherit.aes = FALSE, alpha = 0.2, show.legend = FALSE)+
  scale_fill_manual(values = c("dark gray","white"))+
  geom_point(data = peaks.npc_rna.wesc.overlap %>% 
               filter( lod.npc_rna > 7.5, ensembl_gene_id %in% shared.genes, match =="npc_rna"), 
            aes( x = cumsum_bp_peak.npc_rna, y = cumsum_bp_gene,col="NPC"),
            size = 2, 
            #col = qtl.colors[["npc_rna"]],
             inherit.aes = FALSE )+
  geom_point(data = peaks.npc_rna.wesc.overlap %>% 
               filter( lod.npc_rna > 7.5, ensembl_gene_id %in% shared.genes, match =="shared"), 
            aes( x = cumsum_bp_peak.npc_rna, y = cumsum_bp_gene,col="Shared"),
            size = 2, 
            #col = qtl.colors[["shared"]],
             inherit.aes = FALSE )+
  theme_pubclean(base_size = 16)+
  scale_x_discrete( name = "eQTL peak",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.02))+
  scale_y_discrete( name = "Gene midpoint",limits = chrom_lens_midpt, labels = names(chrom_lens), expand = expansion( mult = 0.02))+
  theme( axis.text = element_text(size = 10),
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_blank())+
  scale_color_manual( name = "eQTL",values = c("NPC"=qtl.colors[["npc_rna"]],  "Shared"=qtl.colors[["shared"]]), labels = c("NPC","Shared"))



npc_eqtl_shared_map_plot <- ggarrange( npc_eqtl_shared_plot, trans_band_plot, nrow = 2, heights = c(0.7, 0.3))

#ggsave(npc_eqtl_shared_map_plot, filename = here("figures","npc_eqtl_shared_map.svg"), width = 8, height = 11, dpi = 300, units = "in")

npc_eqtl_shared_map_plot
```

## Overlapping eQTL

```{r shared_qtl_effects, warning=FALSE, message=FALSE}

# calculate allele effect correlations for overlapping qtl
peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >= 7.5,  match =="shared") -> shared_eqtl
peaks.esc_rna.wEffs %>% 
  inner_join( ., shared_eqtl %>% 
                select( ensembl_gene_id, peak_chr, lod.npc_rna)) %>% 
  select(
    ensembl_gene_id, peak_chr, lod.npc_rna, paste0(LETTERS[1:8],".esc_rna")
  ) %>% 
  full_join(
    peaks.npc_rna.wEffs %>% 
      inner_join( ., shared_eqtl %>% 
                select( ensembl_gene_id, peak_chr, lod.npc_rna, lod.esc_rna)) %>% 
      select(
        ensembl_gene_id, peak_chr, lod.npc_rna, paste0(LETTERS[1:8],".npc_rna")
      )
  ) %>%
  left_join( shared_eqtl) %>% 
  mutate( qtl_id = 1:n()) -> shared_eqtl_weffs

shared_eqtl_weffs %>% 
  select( qtl_id, paste0(LETTERS[1:8],".npc_rna")) %>% 
  column_to_rownames("qtl_id") %>% 
  t() -> shared_eqtl_npc_effs
colnames(shared_eqtl_npc_effs) <- paste0(colnames(shared_eqtl_npc_effs),"_npc")
shared_eqtl_weffs %>% 
  select( qtl_id, paste0(LETTERS[1:8],".esc_rna")) %>% 
  column_to_rownames("qtl_id") %>% 
  t() -> shared_eqtl_esc_effs
colnames(shared_eqtl_esc_effs) <- paste0(colnames(shared_eqtl_esc_effs),"_esc")

shared_eqtl_effs_cor <- rcorr( shared_eqtl_npc_effs, shared_eqtl_esc_effs)
shared_eqtl_effs_cor_df <- tibble(
  cor = diag( shared_eqtl_effs_cor$r[colnames(shared_eqtl_npc_effs),colnames(shared_eqtl_esc_effs)]),
  p_val = diag( shared_eqtl_effs_cor$P[colnames(shared_eqtl_npc_effs),colnames(shared_eqtl_esc_effs)]),
  n = diag( shared_eqtl_effs_cor$n[colnames(shared_eqtl_npc_effs),colnames(shared_eqtl_esc_effs)]),
  qtl_id = as.numeric(gsub("_npc","",colnames(shared_eqtl_npc_effs)))
) %>% 
  left_join( shared_eqtl_weffs)



```


```{r shared_eqtl_table}

# table of shared eqtl with lod scores + allele effect correlations
shared_eqtl_effs_cor_df %>% 
  left_join( select(peaks.npc_rna.wEffs, ensembl_gene_id, peak_chr, lod.npc_rna, local.npc_rna)
  ) %>% 
  filter( !is.na(local.npc_rna)) %>% 
  mutate( local = ifelse( local.npc_rna == TRUE, "Local", "Distant")) %>% 
   mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  mutate(`Significance` =  ifelse(p_adj < 0.1, "FDR < 0.1", "ns")) %>% 
  left_join( npc.genes %>% 
               select(ensembl_gene_id, mgi_symbol)) %>% 
  select(ensembl_gene_id, mgi_symbol, peak_chr, lod.esc_rna, lod.npc_rna, cor, Significance,local.npc_rna) %>%
  mutate_if(is.numeric, round ,2) %>% 
  create_dt()

```


```{r shared_lod_scores, fig.height=6, fig.width= 10}

# plot esc vs npc lod scores for shared eQTL

shared_eqtl_effs_cor_df %>% 
  left_join( select(peaks.npc_rna.wEffs, ensembl_gene_id, peak_chr, lod.npc_rna, local.npc_rna)
  ) %>% 
  filter( !is.na(local.npc_rna)) %>% 
  mutate( local = ifelse( local.npc_rna == TRUE, "Local", "Distant")) %>%
   mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  mutate(`Significance` =  ifelse(p_adj < 0.1, "FDR < 0.1", "ns")) %>% 
  ggplot()+
  aes( x = lod.esc_rna,
       y = lod.npc_rna,
       col = Significance)+
  geom_point(size =3, alpha = 0.7)+
  theme_pubclean(base_size = 18)+
  xlab( "ESC LOD")+
  ylab( "NPC LOD")+
  xlim(0, 80)+
  ylim(0,80)+
  #scale_color_gradient2()+
  scale_color_manual(values = c("dark red","dark gray")) +
  labs( col = "Significance")+
  facet_wrap(~local)


```



```{r shared_qtl_eff_plot, fig.height=5, fig.width=7,warning=FALSE, message=FALSE}

# plot all + local + distant
shared_eqtl_effs_cor_df %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  mutate(`Significance` =  ifelse(p_adj < 0.1, "FDR < 0.1", "ns")) %>% #count( p_adj <0.1)
  ggplot() +
  aes(x = cor, col = Significance, fill = Significance) +
  geom_histogram(binwidth = 0.01) +
  theme_pubclean(base_size = 18) +
  #facet_wrap(~local, scales = "free") +
  scale_color_manual(values = c(qtl.colors[["shared"]],"dark gray")) +
  scale_fill_manual(values = c(qtl.colors[["shared"]],"dark gray")) +
  xlab("Haplotype effects correlation")+
  ylab("Count")+
  xlim(-1,1)+
  theme(legend.position = "top")
  
  

```


## Overlapping eQTL with opposite allele effects

```{r shared_qtl_neg_corr}

# table
shared_eqtl_effs_cor_df %>% 
  left_join( select(peaks.npc_rna.wEffs, ensembl_gene_id, peak_chr, lod.npc_rna, local.npc_rna)
  ) %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  filter( cor < 0, p_adj < 0.1) %>% 
  left_join( npc.genes %>% 
               select(ensembl_gene_id, mgi_symbol)) %>% 
  select(ensembl_gene_id, mgi_symbol, peak_chr, lod.esc_rna, lod.npc_rna, cor, local.npc_rna) %>%
  mutate_if(is.numeric, round ,2) %>% 
  create_dt()
  


```

Only Stmn3 eQTL is distant (same chromosome but 11Mb away from midpoint).

```{r shared_qtl_neg_cor_plot, fig.height=12, fig.width=16}


shared_eqtl_effs_cor_df %>%
  left_join( all.genes %>% select(ensembl_gene_id, mgi_symbol)) %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  filter( cor < 0, p_adj < 0.1) %>% 
  select(mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>%
  gather(founder, ESC_eff, `A.esc_rna`:`H.esc_rna`) %>%
  mutate(founder = gsub(".esc_rna", "", founder)) -> temp1

shared_eqtl_effs_cor_df %>%
  left_join( all.genes %>% select(ensembl_gene_id, mgi_symbol)) %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  filter( cor < 0, p_adj < 0.1) %>% 
  select(mgi_symbol, paste0(LETTERS[1:8], ".npc_rna")) %>%
  gather(founder, NPC_eff, `A.npc_rna`:`H.npc_rna`) %>%
  mutate(founder = gsub(".npc_rna", "", founder)) -> temp2

inner_join(temp1, temp2) %>%
  gather(QTL, eff, `ESC_eff`:`NPC_eff`) %>%
  ggplot() +
  aes(x = founder, y = eff, col = QTL, group = interaction(QTL, mgi_symbol)) +
  geom_point() +
  geom_line() +
  facet_wrap(~mgi_symbol, ncol = 5, scales = "free") +
  theme_pubclean(base_size = 14) +
  geom_hline(yintercept = 0) +
  ggtitle("Overlapping local QTL with opposing allele effects")


```

## Overlapping eQTL lacking agreement in allele effects (abs(cor)<0.1, adjusted p >0.1)

```{r shared_qtl_no_corr}

# table? plot? with overlapping eqtl with no correlation btw allele effects
shared_eqtl_effs_cor_df %>% 
  left_join( select(peaks.npc_rna.wEffs, ensembl_gene_id, peak_chr, lod.npc_rna, local.npc_rna)
  ) %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  filter( abs(cor) < 0.1, p_adj > 0.1) %>% 
    left_join( npc.genes %>% 
               select(ensembl_gene_id, mgi_symbol)) %>% 
  select(ensembl_gene_id, mgi_symbol, peak_chr, lod.esc_rna, lod.npc_rna, cor, local.npc_rna) %>%
  mutate_if(is.numeric, round ,2) %>% 
  create_dt()

```

## eQTL uniquely observed in NPCs


```{r npc_only_eqtl_table}

shared_eqtl_effs_cor_df %>% 
  mutate( p_adj = p.adjust(p_val, method = "BH")) %>% 
  filter( p_adj > 0.1) %>% 
  select(ensembl_gene_id, lod.npc_rna, peak_chr) -> shared.wo.corr

peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >= 7.5, match =="npc_rna", local.npc_rna ==T, ensembl_gene_id %in% shared.genes) %>% 
  rbind(
    peaks.npc_rna.wesc.overlap %>% 
      inner_join( shared.wo.corr) %>% 
      filter( local.npc_rna)
  ) -> unique_npc_local_eqtl


peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >= 7.5, match =="npc_rna", local.npc_rna ==F, ensembl_gene_id %in% shared.genes) %>% 
    rbind(
    peaks.npc_rna.wesc.overlap %>% 
      inner_join( shared.wo.corr) %>% 
      filter( !local.npc_rna)
  ) -> unique_npc_distant_eqtl


unique_npc_local_eqtl %>% 
  rbind(unique_npc_distant_eqtl) %>% 
  select( ensembl_gene_id, mgi_symbol, gene_chr, peak_chr, lod.npc_rna, lod.esc_rna,local.npc_rna ) %>% 
  mutate_if(is.numeric, round, 2) %>% 
  create_dt()


```

```{r, eval = FALSE}

# Looking at the best mediator for the unique NPC eQTL.
eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    unique_npc_local_eqtl %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 2) %>% 
  filter( target.symbol == mediator.symbol) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, target.chr, target.lod,mediation.lod, mediator.lod,  type)


```

```{r npc_only_eqtl_ora}

genes.w.eqtl <- peaks.npc_rna.wEffs %>% 
  filter(lod.npc_rna > 7.5) %>% 
  select(ensembl_gene_id, local.npc_rna) %>% 
  distinct() %>% 
  left_join(npc.genes) %>% 
  filter( ensembl_gene_id %in% shared.genes)

g.unique.local <- gost(
  query = unique(unique_npc_local_eqtl$mgi_symbol),
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = unique((genes.w.eqtl %>% filter(local.npc_rna))$mgi_symbol), #npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.unique.local$result <- g.unique.local$result %>% filter(term_size < 660)

g.unique.dist <- gost(
  query = unique(unique_npc_distant_eqtl$mgi_symbol),
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg =  unique((genes.w.eqtl %>% filter(!local.npc_rna))$mgi_symbol), #npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.unique.dist$result <- g.unique.dist$result %>% filter(term_size < 660)


```


### ORA results for genes with unique NPC local eQTL

Note that I am running ORA for the genes with unique NPC eQTL using a custom background of genes with local eQTL instead of the full NPC gene list.

```{r npc_unique_local_ora}

g.unique.local$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```


### ORA results for genes with unique NPC distant eQTL

Note that I am running ORA for the genes with unique NPC eQTL using a custom background of genes with distant eQTL instead of the full NPC gene list.

```{r npc_unique_dist_ora}

g.unique.dist$result %>% 
  select( term_name, source, FDR = p_value, term_size, intersection_size) %>% 
  filter( FDR <0.01) %>% 
  mutate_if( is.numeric, formatC, digits =2) %>% 
  create_dt()

```


There are more than 400 TFs overrepresented with around 150 of them expressed in NPCs. Only 23 have eQTLs in NPCs as listed below.

```{r tf_motifs_dist_eqtl}

unique_npc_distant_eqtl_TFs <- g.unique.dist$result %>% 
  filter( p_value <0.01) %>% 
  filter(
    source == "TF"
  ) %>% 
  mutate( 
    TF = str_replace(str_replace(str_replace(str_to_upper(str_split_i(str_split_i(term_name, ":",2), ";",1)), pattern = "-", replacement = ""), pattern = ".",replacement = ""), pattern = "," , replacement = "")
    ) %>% 
  mutate( TF = str_to_title(TF)) %>% 
  select(TF, intersection, intersection_size, term_size, term_name, term_id, p_value)


peaks.npc_rna.wesc.overlap %>% 
  filter(mgi_symbol %in% unique_npc_distant_eqtl_TFs$TF,
         lod.npc_rna >=7.5 | lod.esc_rna >= 7.5) %>% 
  select( ensembl_gene_id, mgi_symbol, peak_chr, gene_chr,match, lod.npc_rna, lod.esc_rna, interp_bp_peak.npc_rna, interp_bp_peak.esc_rna, local.npc_rna, local.esc_rna) %>% 
  mutate_if( is.numeric, round, 2) %>% 
  create_dt()
   

```


Out of 757 genes with unique distant eQTL in NPCs, 637 of them intersect with TF motif gene sets in total. The range for the total number of genes identified in relation to a TF varies between 34 and 626, and the number of TFs a gene has been identified in relation to ranges between 134 and 241 suggesting significant overlap across the TF motif gene sets.

```{r tf_motifs_dist_eqtl_genes, results='hide'}

unique_npc_distant_eqtl_TF_genes <- unique_npc_distant_eqtl_TFs %>% 
  select(TF,term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes) 

unique_npc_distant_eqtl_TF_genes %>% 
  select(TF, mgi_symbol) %>% 
  group_by(TF)  %>% 
  mutate( `# of Genes` = n_distinct(mgi_symbol)) %>% 
  select( -mgi_symbol) %>% distinct() %>% arrange(`# of Genes`) 

unique_npc_distant_eqtl_TF_genes %>% 
  select(TF, Gene = mgi_symbol) %>% 
  group_by(Gene) %>% 
  mutate( #TFs = paste0(TF, collapse = ", "), 
          `# of TFs` = n_distinct(TF)) %>% 
  select( -TF) %>% distinct() %>% arrange(`# of TFs`)

# Looking to see if these TFs are best mediators for any of the distant eQTL

eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    unique_npc_distant_eqtl %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 2) %>% 
  filter( target.symbol != mediator.symbol) %>% 
  filter( mediator.symbol %in% unique_npc_distant_eqtl_TF_genes$TF) %>% 
  arrange( mediation.lod) %>% 
  select( target.symbol, target.chr, target.lod,mediator.symbol, mediation.lod, mediator.lod,  type) %>% 
  create_dt()

# nothing really worth following up here!


```



## eQTL uniquely observed in ESCs

Nothing! 

```{r esc_only_eqtl_ora}

peaks.npc_rna.wesc.overlap %>% 
  filter( lod.esc_rna > 7.5,
          match =="esc_rna", 
          local.esc_rna ==T, 
          ensembl_gene_id %in% shared.genes) %>% 
  rbind(
    peaks.npc_rna.wesc.overlap %>% 
      inner_join( shared.wo.corr) %>% 
      filter( local.esc_rna)
  ) -> unique_esc_local_eqtl


peaks.npc_rna.wesc.overlap %>% 
  filter( lod.esc_rna > 7.5, 
          match =="esc_rna", 
          local.esc_rna ==F, 
          ensembl_gene_id %in% shared.genes) %>% 
    rbind(
    peaks.npc_rna.wesc.overlap %>% 
      inner_join( shared.wo.corr) %>% 
      filter( !local.esc_rna)
  ) -> unique_esc_distant_eqtl

genes.w.eqtl_esc <- peaks.esc_rna.wEffs %>% 
  filter(lod.esc_rna > 7.5) %>% 
  select(ensembl_gene_id, local.esc_rna) %>% 
  distinct() %>% 
  left_join(esc.genes) %>% 
  filter( ensembl_gene_id %in% shared.genes)

g.unique.local_esc <- gost(
  query = unique(unique_esc_local_eqtl$mgi_symbol),
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg = unique((genes.w.eqtl_esc %>% filter(local.esc_rna))$mgi_symbol), #npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
#g.unique.local_esc$result <- g.unique.local_esc$result %>% filter(term_size < 660)

g.unique.dist_esc <- gost(
  query = unique(unique_esc_distant_eqtl$mgi_symbol),
  organism = "mmusculus",
  domain_scope = "custom",
  custom_bg =  unique((genes.w.eqtl_esc %>% filter(!local.esc_rna))$mgi_symbol), #npc.genes$mgi_symbol,
  evcodes = TRUE,
  correction_method = "fdr"
)
g.unique.dist_esc$result <- g.unique.dist_esc$result %>% filter(term_size < 660)
```




## Examples {.tabset .tabset-fade .tabset-pills}

Example of shared eQTL with matching allele effects: Csnk2a1

Example of shared eQTL with opposite allele effects: Amph

Example of unique eQTL: Gbx2 (l), Lats1(d)

### Csnk2a1 (local)

```{r Csnk2a1_eqtl}

Csnk2a1_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Csnk2a1", local.npc_rna ==T)

# LOD plot
Csnk2a1_esc_scan <- scan1( pheno = exprZ.esc_rna[,Csnk2a1_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.esc_rna, 
                    kinship = kinship_loco.esc_rna,
                    addcovar = covar.esc_rna)

Csnk2a1_npc_scan <- scan1( pheno = exprZ.npc_rna[,Csnk2a1_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)


Csnk2a1_esc_scan %>% 
  as.data.frame( ) %>% 
  rename( esc_rna = Csnk2a1_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Csnk2a1_esc_scan)[[1]]) %>% 
  left_join(map_dat2) %>% 
  cbind(
    Csnk2a1_npc_scan %>% as.data.frame() %>% rename( npc_rna = Csnk2a1_eqtl$ensembl_gene_id)
  ) -> Csnk2a1_qtl_scans

Csnk2a1_qtl_scans %>% 
  filter( chr == Csnk2a1_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna","esc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod,
      col = qtl_type
      )+
    geom_line( size = 1.5)+
    theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                       labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  xlab(paste0("Chr ",Csnk2a1_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = Csnk2a1_eqtl$gene_start/1e06, xend = Csnk2a1_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Csnk2a1_eqtl$midpoint/1e06, y = -0.8, label ="Csnk2a1", size =6, fontface = "italic")-> Csnk2a1_lod_plot



# Effects plot
Csnk2a1_esc_eff <- peaks.esc_rna.wEffs %>% 
  inner_join( Csnk2a1_eqtl %>% select(ensembl_gene_id, peak_chr, lod.esc_rna))

Csnk2a1_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Csnk2a1_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Csnk2a1_effs <- Csnk2a1_esc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>% 
  cbind( Csnk2a1_npc_eff %>% select(paste0(LETTERS[1:8], ".npc_rna")) ) 

Csnk2a1_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna"),
          paste0(LETTERS[1:8], ".esc_rna")), 
          names_to = c("effect"),
          values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Csnk2a1_haplotype_plot

```


```{r Csnk2a1_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Csnk2a1_lod_plot, Csnk2a1_haplotype_plot, widths = c(1, 0.5))

```


### Dpysl2 (local)

```{r Dpysl2_eqtl}

Dpysl2_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Dpysl2", local.npc_rna ==T)

# LOD plot
Dpysl2_esc_scan <- scan1( pheno = exprZ.esc_rna[,Dpysl2_eqtl$ensembl_gene_id,drop=FALSE], 
                           genoprobs = probs.esc_rna, 
                           kinship = kinship_loco.esc_rna,
                           addcovar = covar.esc_rna)

Dpysl2_npc_scan <- scan1( pheno = exprZ.npc_rna[,Dpysl2_eqtl$ensembl_gene_id,drop=FALSE], 
                           genoprobs = probs.npc_rna, 
                           kinship = kinship_loco.npc_rna,
                           addcovar = covar.npc_rna)


Dpysl2_esc_scan %>% 
  as.data.frame( ) %>% 
  rename( esc_rna = Dpysl2_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Dpysl2_esc_scan)[[1]]) %>% 
  left_join(map_dat2) %>% 
  cbind(
    Dpysl2_npc_scan %>% as.data.frame() %>% rename( npc_rna = Dpysl2_eqtl$ensembl_gene_id)
  ) -> Dpysl2_qtl_scans

Dpysl2_qtl_scans %>% 
  filter( chr == Dpysl2_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna","esc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( 
    x= pos_bp/1e06,
    y = lod,
    col = qtl_type
  )+
  geom_line( size = 1.5)+
  theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  xlab(paste0("Chr ",Dpysl2_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = Dpysl2_eqtl$gene_start/1e06, xend = Dpysl2_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Dpysl2_eqtl$midpoint/1e06, y = -0.8, label ="Dpysl2", size =6, fontface = "italic")-> Dpysl2_lod_plot



# Effects plot
Dpysl2_esc_eff <- peaks.esc_rna.wEffs %>% 
  inner_join( Dpysl2_eqtl %>% select(ensembl_gene_id, peak_chr, lod.esc_rna))

Dpysl2_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Dpysl2_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Dpysl2_effs <- Dpysl2_esc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>% 
  cbind( Dpysl2_npc_eff %>% select(paste0(LETTERS[1:8], ".npc_rna")) ) 

Dpysl2_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna"),
                          paste0(LETTERS[1:8], ".esc_rna")), 
                names_to = c("effect"),
                values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Dpysl2_haplotype_plot

```

```{r Dpysl2_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Dpysl2_lod_plot, Dpysl2_haplotype_plot, widths = c(1, 0.5))

```

### Amph (local)

```{r Amph_eqtl}

Amph_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Amph", local.npc_rna == T)

# LOD plot
Amph_esc_scan <- scan1( pheno = exprZ.esc_rna[,Amph_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.esc_rna, 
                    kinship = kinship_loco.esc_rna,
                    addcovar = covar.esc_rna)

Amph_npc_scan <- scan1( pheno = exprZ.npc_rna[,Amph_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)


Amph_esc_scan %>% 
  as.data.frame( ) %>% 
  rename( esc_rna = Amph_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Amph_esc_scan)[[1]]) %>% 
  left_join(map_dat2) %>% 
  cbind(
    Amph_npc_scan %>% as.data.frame() %>% rename( npc_rna = Amph_eqtl$ensembl_gene_id)
  ) -> Amph_qtl_scans

Amph_qtl_scans %>% 
  filter( chr == Amph_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna","esc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod,
      col = qtl_type
      )+
    geom_line( size = 1.5)+
    theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                       labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  xlab(paste0("Chr ",Amph_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = Amph_eqtl$gene_start/1e06, xend = Amph_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Amph_eqtl$midpoint/1e06, y = -0.8, label ="Amph", size =6, fontface = "italic")-> Amph_lod_plot



# Effects plot
Amph_esc_eff <- peaks.esc_rna.wEffs %>% 
  inner_join( Amph_eqtl %>% select(ensembl_gene_id, peak_chr, lod.esc_rna))

Amph_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Amph_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Amph_effs <- Amph_esc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>% 
  cbind( Amph_npc_eff %>% select(paste0(LETTERS[1:8], ".npc_rna")) ) 

Amph_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna"),
          paste0(LETTERS[1:8], ".esc_rna")), 
          names_to = c("effect"),
          values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Amph_haplotype_plot

```


```{r Amph_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Amph_lod_plot, Amph_haplotype_plot, widths = c(1, 0.5))

```

### Gbx2 (local)

```{r Gbx2_eqtl}

Gbx2_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Gbx2", local.npc_rna ==T)

# LOD plot
Gbx2_npc_scan <- scan1( pheno = exprZ.npc_rna[,Gbx2_eqtl$ensembl_gene_id,drop=FALSE], 
                           genoprobs = probs.npc_rna, 
                           kinship = kinship_loco.npc_rna,
                           addcovar = covar.npc_rna)


Gbx2_npc_scan %>% 
  as.data.frame( ) %>% 
  rename( npc_rna = Gbx2_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Gbx2_npc_scan)[[1]]) %>% 
  left_join(map_dat2) -> Gbx2_qtl_scans

Gbx2_qtl_scans %>% 
  filter( chr == Gbx2_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  #mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( 
    x= pos_bp/1e06,
    y = lod,
    col = qtl_type
  )+
  geom_line( size = 1.5)+
  theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]]), 
                      labels = c("npc_rna"="NPC"))+
  xlab(paste0("Chr ",Gbx2_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = Gbx2_eqtl$gene_start/1e06, xend = Gbx2_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Gbx2_eqtl$midpoint/1e06, y = -0.8, label ="Gbx2", size =6, fontface = "italic")-> Gbx2_lod_plot



# Effects plot
Gbx2_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Gbx2_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Gbx2_effs <- Gbx2_npc_eff %>% 
  select(paste0(LETTERS[1:8], ".npc_rna"))  

Gbx2_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna")), 
                names_to = c("effect"),
                values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]]), 
                      labels = c("npc_rna"="NPC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Gbx2_haplotype_plot

```

```{r Gbx2_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Gbx2_lod_plot, Gbx2_haplotype_plot, widths = c(1, 0.5))

```

### Lats1 (distant)

```{r Lats1_eqtl}

Lats1_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Lats1", local.npc_rna ==F)

# LOD plot

Lats1_npc_scan <- scan1( pheno = exprZ.npc_rna[,Lats1_eqtl$ensembl_gene_id,drop=FALSE], 
                           genoprobs = probs.npc_rna, 
                           kinship = kinship_loco.npc_rna,
                           addcovar = covar.npc_rna)

Lats1_npc_scan %>% 
  as.data.frame( ) %>% 
  mutate( marker = dimnames(Lats1_npc_scan)[[1]]) %>% 
  rename( npc_rna = Lats1_eqtl$ensembl_gene_id) %>% 
  left_join(map_dat2)  -> Lats1_qtl_scans

Lats1_qtl_scans %>% 
  filter( chr == Lats1_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna"))) %>% 
  ggplot()+
  aes( 
    x= pos_bp/1e06,
    y = lod,
    col = qtl_type
  )+
  geom_line( size = 1.5)+
  theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]]), 
                      labels = c("npc_rna"="NPC"))+
  xlab(paste0("Chr ",Lats1_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  #geom_segment( x = Lats1_eqtl$gene_start/1e06, xend = Lats1_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Lats1_eqtl$midpoint/1e06, y = -0.8, label ="Lats1", size =6, fontface = "italic")-> Lats1_lod_plot



# Effects plot
Lats1_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Lats1_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Lats1_effs <- Lats1_npc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".npc_rna"))

Lats1_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna")), 
                names_to = c("effect"),
                values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]]), 
                      labels = c("npc_rna"="NPC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Lats1_haplotype_plot

```

```{r Lats1_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Lats1_lod_plot, Lats1_haplotype_plot, widths = c(1, 0.5))

```

### Dyrk1a

```{r Dyrk1a_eqtl}

Dyrk1a_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.esc_rna >7.5, mgi_symbol =="Dyrk1a")

# LOD plot
Dyrk1a_esc_scan <- scan1( pheno = exprZ.esc_rna[,Dyrk1a_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.esc_rna, 
                    kinship = kinship_loco.esc_rna,
                    addcovar = covar.esc_rna)

Dyrk1a_npc_scan <- scan1( pheno = exprZ.npc_rna[,Dyrk1a_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)


Dyrk1a_esc_scan %>% 
  as.data.frame( ) %>% 
  rename( esc_rna = Dyrk1a_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Dyrk1a_esc_scan)[[1]]) %>% 
  left_join(map_dat2) %>% 
  cbind(
    Dyrk1a_npc_scan %>% as.data.frame() %>% rename( npc_rna = Dyrk1a_eqtl$ensembl_gene_id)
  ) -> Dyrk1a_qtl_scans

Dyrk1a_qtl_scans %>% 
  filter( chr == Dyrk1a_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna","esc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod,
      col = qtl_type
      )+
    geom_line( size = 1.5)+
    theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                       labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  xlab(paste0("Chr ",Dyrk1a_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = Dyrk1a_eqtl$gene_start/1e06, xend = Dyrk1a_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Dyrk1a_eqtl$midpoint/1e06, y = -0.8, label ="Dyrk1a", size =6, fontface = "italic")-> Dyrk1a_lod_plot



# Effects plot
Dyrk1a_esc_eff <- peaks.esc_rna.wEffs %>% 
  inner_join( Dyrk1a_eqtl %>% select(ensembl_gene_id, peak_chr, lod.esc_rna))

Dyrk1a_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( Dyrk1a_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
Dyrk1a_effs <- Dyrk1a_esc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>% 
  cbind( Dyrk1a_npc_eff %>% select(paste0(LETTERS[1:8], ".npc_rna")) ) 

Dyrk1a_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna"),
          paste0(LETTERS[1:8], ".esc_rna")), 
          names_to = c("effect"),
          values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> Dyrk1a_haplotype_plot

```


```{r Dyrk1a_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( Dyrk1a_lod_plot, Dyrk1a_haplotype_plot, widths = c(1, 0.5))

```

```{r dyrk1a_npc_plot, fig.height=6, fig.width=8}


Dyrk1a_npc_eff <- scan1blup( pheno = exprZ.npc_rna[,Dyrk1a_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna[,"16"], 
                    kinship = kinship_loco.npc_rna[["16"]],
                    addcovar = covar.npc_rna)

plot_coefCC( Dyrk1a_npc_eff, pmap, scan1_output =  Dyrk1a_npc_scan, main = "Dyrk1a eQTL scan in DO mNPCs", legend = "topleft")

```


```{r dyrk1a_esc_plot, fig.height=6, fig.width=8}


Dyrk1a_esc_eff <- scan1blup( pheno = exprZ.esc_rna[,Dyrk1a_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.esc_rna[,"16"], 
                    kinship = kinship_loco.esc_rna[["16"]],
                    addcovar = covar.esc_rna)

plot_coefCC( Dyrk1a_esc_eff, pmap, scan1_output =  Dyrk1a_esc_scan, main = "Dyrk1a eQTL scan in DO mESCs", legend = "topleft")


```

```{r dyrk1a_full_genome_scans, fig.height=4, fig.width=12}

plot( Dyrk1a_npc_scan, pmap, main = "Dyrk1a eQTL scan in DO mNPCs", ylim= c(0,20))
abline(h = 7.5, col = "dark red", lty = 2)

plot( Dyrk1a_esc_scan, pmap, main = "Dyrk1a eQTL scan in DO mESCs", ylim= c(0,20))
abline(h = 7.5, col = "dark red", lty = 2)

```








