---
title: ""
output:
  html_document:
    toc: false
    df_print: paged
    code_folding: hide
---


```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

# adding the fixed function from bmediatr to use later
# Conversion function taken from qtl2convert
map_df_to_list <- function (map, chr_column = "chr", pos_column = "cM", marker_column = "marker", 
          Xchr = c("x", "X")) {
  
  if (is.null(marker_column)) {
    marker_column <- "qtl2tmp_marker"
    map[, marker_column] <- rownames(map)
  }
  if (!(marker_column %in% colnames(map))) 
    stop("Column \"", marker_column, "\" not found.")
  if (!(chr_column %in% colnames(map))) 
    stop("Column \"", chr_column, "\" not found.")
  if (!(pos_column %in% colnames(map))) 
    stop("Column \"", pos_column, "\" not found.")
  marker <- map[, marker_column]
  chr <- map[, chr_column]
  uchr <- unique(chr)
  pos <- map[, pos_column]
  result <- split(as.numeric(pos), factor(chr, levels = uchr))
  marker <- split(marker, factor(chr, levels = uchr))
  for (i in seq(along = result)) names(result[[i]]) <- marker[[i]]
  is_x_chr <- rep(FALSE, length(result))
  names(is_x_chr) <- names(result)
  if (!is.null(Xchr)) {
    Xchr_used <- Xchr %in% names(is_x_chr)
    if (any(Xchr_used)) {
      Xchr <- Xchr[Xchr_used]
      is_x_chr[Xchr] <- TRUE
    }
  }
  attr(result, "is_x_chr") <- is_x_chr
  result
}

#' Posterior odds genome plot function
#'
#' This function takes the posterior odds results from bmediatR() and plots the genome-wide scan.
#'
#' @param bmediatR_object Output from bmediatR(). 
#' @param model_type DEFAULT: "mediation". Specifies which model(s)'s posterior probabilities are to be included in the numerator of the posterior odds and then displayed for
#' for genome-wide mediators. 
#' @param med_annot Annotation data for -omic mediators.
#' @param include_chr DEFAULT: c(1:19, "X"). Chromosomes to include in plot.
#' @param expland_lim_factor DEFAULT: 0.025. Scale to increase plot limits by.
#' @param label_thresh DEFAULT: NULL. Label mediators that surpass label_thresh. Default does not add labels.
#' @param label_thresh_greater_than DEFAULT: TRUE. If TRUE, passing mediators have log odds greater than the threshold.
#' If FALSE, passing mediators have log odds less than the threshold.  
#' @param label_only_chr DEFAULT: NULL. Only label mediators that pass label_thresh on the specified chromosome.
#' @param qtl_dat DEFAULT: NULL. QTL data that includes position of QTL and outcome. Adds ticks to the figure.
#' @export
#' @examples plot_posterior_odds()
plot_posterior_odds <- function(bmediatR_object, 
                                model_type = c("mediation", "partial", "complete", "colocal"),
                                med_annot, 
                                med_var = "protein.id",
                                include_chr = c(1:19, "X"), 
                                expand_lim_factor = 0.025, 
                                label_thresh = NULL, 
                                label_thresh_greater_than = TRUE,
                                label_only_chr = NULL,
                                bgcol = "white", altcol = "gray", altbgcol = "white", 
                                hlines_col = "gray80", col = "black", cex = 0.75,
                                qtl_dat = NULL,
                                outcome_symbol = NULL,
                                ymax = NULL,
                                ymin = NULL,
                                ...) {
  
  model_type <- model_type[1]
  
  post_odds <- matrix(bmediatR_object[["ln_post_odds"]][,model_type], ncol = 1)
  rownames(post_odds) <- rownames(bmediatR_object[["ln_post_odds"]])
  class(post_odds) <- "scan1"
  
  med_map_df <- med_annot %>%
    dplyr::select(tidyselect::all_of(med_var), symbol, chr, middle) %>%
    dplyr::filter(chr %in% include_chr) %>%
    dplyr::mutate(chr = factor(chr, levels = c(1:19, "X"))) %>%
    as.data.frame %>% 
    dplyr::arrange(chr)
  if (!is.null(qtl_dat)) {
    ## Add QTL to map for plotting
    med_map_df <- dplyr::bind_rows(med_map_df,
                                   qtl_dat %>%
                                     dplyr::mutate((!!as.symbol(med_var)) := "QTL",
                                                   symbol = "QTL") %>%
                                     dplyr::rename(middle = pos) %>%
                                     dplyr::select(tidyselect::all_of(med_var), symbol, chr, middle))
  }
  med_map <- map_df_to_list(map = med_map_df, marker_column = med_var, pos_column = "middle")
  
  gap <- sum(qtl2::chr_lengths(med_map))/100
  
  lim_shift <- (max(post_odds[,1]) - min(post_odds[,1])) * expand_lim_factor
  
  if (is.null(ymax)) { ymax <- max(post_odds[,1]) + lim_shift }
  if (is.null(ymin)) { ymin <- min(post_odds[,1]) - lim_shift }
  
  qtl2:::plot.scan1(post_odds, map = med_map, ylab = "Log posterior odds", type = "p", pch = 20, 
                    ylim = c(ymin, ymax),
                    bgcol = bgcol, altcol = altcol, altbgcol = altbgcol, hlines_col = hlines_col, col = col, 
                    cex = cex, gap = gap,
                    ...)
  
  xpos <- qtl2:::map_to_xpos(map = med_map, gap = gap)
  
  ## Mediator labels
  label_dat <- matrix(bmediatR_object[["ln_post_odds"]][,model_type], ncol = 1)
  colnames(label_dat) <- "post_odds"
  rownames(label_dat) <- rownames(bmediatR_object[["ln_post_odds"]])
  label_dat <- label_dat %>%
    as.data.frame %>%
    tibble::rownames_to_column(med_var) %>%
    dplyr::left_join(med_map_df)
  if (!is.null(label_only_chr)) {
    label_dat <- label_dat %>%
      dplyr::filter(chr == label_only_chr)
  } else {
    label_dat <- label_dat %>%
      dplyr::filter(chr %in% include_chr)
  }
  label_post_odds <- label_dat %>%
    dplyr::select(tidyselect::all_of(med_var), post_odds) %>%
    tibble::column_to_rownames(med_var) %>%
    as.matrix()
  
  
  if (!is.null(label_thresh)) {
    if (label_thresh_greater_than & any(label_post_odds > label_thresh)) {
      labels <- rownames(label_post_odds)[label_post_odds > label_thresh]
    }
    if (!label_thresh_greater_than & any(label_post_odds < label_thresh)) {
      labels <- rownames(label_post_odds)[label_post_odds < label_thresh]
    }
    
    if (!is.null(labels)) {
      label_map_df <- med_map_df %>%
        filter((!!as.symbol(med_var)) %in% labels) 
      
      for (i in 1:nrow(label_map_df)) {
        lab_pos <- xpos[label_map_df[i, med_var]]
        lab_post_odds <- post_odds[label_map_df[i, med_var],]
        
        text(x = lab_pos, y = lab_post_odds, label_map_df$symbol[i], font = 3)
      }
    }
  }
  if (!is.null(outcome_symbol)) {
    rug(x = xpos[med_annot %>% 
                   dplyr::filter(symbol == outcome_symbol) %>% 
                   pull(tidyselect::all_of(med_var))],
        lwd = 3,
        col = "black")
  }
  if (!is.null(qtl_dat)) {
    rug(x = xpos["QTL"],
        lwd = 3,
        col = "red")
  }
}
```

# Mediators of local eQTL {.tabset .tabset-fade .tabset-pills}

## eQTL with best mediator NPC rna

```{r}

peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna > 7.5, match %in% c("npc_rna","shared"), local.npc_rna ==T) -> npc_eqtl_local

eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_local %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 2) %>% 
  filter( target.symbol == mediator.symbol) %>% 
  slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, mediator.symbol , target.chr, target.lod,mediation.lod,  type) %>% 
  filter(type =="npc") %>% 
  mutate_if(is.numeric, round, 2) %>% 
  mutate( target.lod = round(target.lod, 2)) %>% 
  create_dt()

```


## Mediation example

Making a LOD plot for *Gbx2* with mediation results overlaid to examplify QTL mapping and mediaiton analysis. 

```{r gbx2_plot_prep, warning=FALSE, message=FALSE}

gbx2_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Gbx2")
gbx2_scan <- scan1( pheno = exprZ.npc_rna[,gbx2_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)

gbx2_meds <- eqtl_npc_rna_meds %>% 
  filter( target.symbol =="Gbx2", qtl.chr == gbx2_eqtl$peak_chr) %>% mutate(type = "npc") %>% 
  # rbind( eqtl_esc_rna_meds %>% filter( target.symbol =="Gbx2", qtl.chr == gbx2_eqtl$peak_chr) %>% mutate(type = "esc")) %>% 
  mutate( mediator_mid = (mediator.start+mediator.end)/2)


```


```{r gbx2_exp_plot, fig.width=4, fig.height=3, eval= FALSE}

exprZ.npc_rna[,"ENSMUSG00000034486"] %>% 
  as_tibble(rownames = "Gbx2") %>% 
  ggplot( )+
  aes(x = value)+
  geom_histogram(binwidth = 0.1, col = qtl.colors[["npc_rna"]], fill = qtl.colors[["npc_rna"]], alpha = 0.5)+
  theme_pubclean(base_size = 22)+
  xlab("")+
  ylab("")
  

```

```{r gb2_plot, fig.width=12, fig.height=4, warning=FALSE, message=FALSE}

# lod plot + annotate the gene location too
gbx2_scan %>% 
  as.data.frame( ) %>% 
  rename( eqtl = "ENSMUSG00000034486") %>% 
  mutate( marker = dimnames(gbx2_scan)[[1]]) %>% 
  left_join(map_dat2) -> gbx2_scan_df
gbx2_scan_df$cumsum_pos_bp <- gbx2_scan_df$pos_bp + chrom_lens_offset[gbx2_scan_df$chr]

gbx2_scan_df %>% 
  #filter( chr == gbx2_eqtl$peak_chr) %>% 
  ggplot()+
    aes( 
      x= cumsum_pos_bp,
      y = eqtl
      )+
  geom_line( size = 1, col = qtl.colors[["npc_rna"]],alpha = 0.3)+
  theme_pubclean( base_size = 18)+
  xlab("Chr")+
  #xlab(paste0("Chr ",gbx2_eqtl$peak_chr," location (bp)"))+
  ylab( "LOD score")+
  geom_segment( x = gbx2_eqtl$gene_start+chrom_lens_offset[gbx2_eqtl$gene_chr], 
                xend = gbx2_eqtl$gene_end+chrom_lens_offset[gbx2_eqtl$gene_chr], 
                y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= gbx2_eqtl$midpoint+chrom_lens_offset[gbx2_eqtl$gene_chr], y = -0.5, label ="Gbx2", size =6, fontface = 'italic')+
  scale_x_discrete( name = "Chr",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.05)) -> gbx2_lod_plot

gbx2_meds_min <- gbx2_meds %>% 
  slice_min( mediation.lod)
gbx2_lod_plot+
  geom_point( data = gbx2_meds, 
              aes( x = mediator_mid+chrom_lens_offset[gbx2_eqtl$gene_chr],
                   y = mediation.lod,
                   col = type),
              size = 4,
              alpha = 0.9)+
    scale_color_manual( values = c( "npc"=qtl.colors[["npc_rna"]]),
                      labels = c( "npc"="NPC transcript"), name = "Mediator")+
  annotate( "label", 
            y = gbx2_meds_min$mediation.lod+0.1, 
            x =gbx2_meds_min$mediator_mid+chrom_lens_offset[gbx2_eqtl$gene_chr]-1.2e08 ,
            label = gbx2_meds_min$mediator.symbol, 
            size = 6, 
            col = qtl.colors[["npc_rna"]],
            fontface = "italic")

# overlay mediation lods as dots + add text for the best one

```


## eQTL with best mediator ESC rna


```{r}

eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_local %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 2) %>% 
  filter( target.symbol == mediator.symbol) %>% 
  slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol,mediator.symbol, target.chr, target.lod, mediation.lod,  type) %>% 
  filter(type =="esc")%>% 
  mutate_if(is.numeric, round, 2) %>% 
  mutate( target.lod = round(target.lod, 2)) %>% 
  create_dt()

```


## Mediation example

Making a LOD plot for *Fancg* with mediation results overlaid to examplify QTL mapping and mediaiton analysis. 

```{r fancg_plot_prep, warning=FALSE, message=FALSE}

fancg_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Fancg", local.npc_rna == T)
fancg_scan <- scan1( pheno = exprZ.npc_rna[,fancg_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)

fancg_meds <- eqtl_esc_rna_meds %>% 
  filter( target.symbol =="Fancg", qtl.chr == fancg_eqtl$peak_chr) %>% 
  mutate( type = "esc") %>% 
  rbind( eqtl_npc_rna_meds %>% filter(target.symbol =="Fancg", qtl.chr == fancg_eqtl$peak_chr) %>% mutate(type ="npc")) %>% 
  mutate( mediator_mid = (mediator.start+mediator.end)/2)


```

```{r fancg_plot, fig.width=12, fig.height=5, warning=FALSE, message=FALSE}

# lod plot + annotate the gene location too
fancg_scan %>% 
  as.data.frame( ) %>% 
  rename( eqtl = "ENSMUSG00000028453") %>% 
  mutate( marker = dimnames(fancg_scan)[[1]]) %>% 
  left_join(map_dat2) -> fancg_scan_df
fancg_scan_df$cumsum_pos_bp <- fancg_scan_df$pos_bp + chrom_lens_offset[fancg_scan_df$chr]

fancg_scan_df %>% 
  #filter( chr == fancg_eqtl$peak_chr) %>% 
  ggplot()+
    aes( 
      x= cumsum_pos_bp,
      y = eqtl
      )+
  geom_line( size = 1, col = qtl.colors[["npc_rna"]],alpha = 0.3)+
  theme_pubclean( base_size = 18)+
  xlab("Chr")+
  #xlab(paste0("Chr ",fancg_eqtl$peak_chr," location (bp)"))+
  ylab( "LOD score")+
  geom_segment( x = fancg_eqtl$gene_start+chrom_lens_offset[fancg_eqtl$gene_chr], 
                xend = fancg_eqtl$gene_end+chrom_lens_offset[fancg_eqtl$gene_chr], 
                y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= fancg_eqtl$midpoint+chrom_lens_offset[fancg_eqtl$gene_chr], y = -0.5, label ="Fancg", size =6, fontface = 'italic')+
  scale_x_discrete( name = "Chr",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.05)) -> fancg_lod_plot

fancg_meds_min <- fancg_meds %>% 
  slice_min( mediation.lod)
fancg_lod_plot+
  geom_point( data = fancg_meds, 
              aes( x = mediator_mid+chrom_lens_offset[fancg_eqtl$peak_chr],
                   y = mediation.lod,
                   col = type),
              size = 4,
              alpha = 0.8)+
  scale_color_manual( values = c("esc"=qtl.colors[["esc_rna"]], "npc"=qtl.colors[["npc_rna"]]),
                      labels = c("esc"="ESC transcript", "npc"="NPC transcript"), name = "Mediator")+
  annotate( "label", 
            y = fancg_meds_min$mediation.lod+0.1, 
            x =fancg_meds_min$mediator_mid+chrom_lens_offset[fancg_eqtl$peak_chr]-1.3e08 ,
            label = fancg_meds_min$mediator.symbol, 
            size = 6, 
            fontface = "italic",
            col = qtl.colors[["esc_rna"]])

# overlay mediation lods as dots + add text for the best one

```


# Mediators of distant eQTL {.tabset .tabset-fade .tabset-pills}

## eQTL with best mediator NPC rna

```{r}

peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna > 7.5, match %in% c("shared","npc_rna"), local.npc_rna ==F) -> npc_eqtl_dist


eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>% 
  #filter( target.symbol == mediator.symbol) %>% 
  #slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, target.chr, mediator.symbol, mediator.chr, qtl.chr, target.lod,mediation.lod,  type) %>% 
  filter(type =="npc")%>% 
  mutate_if(is.numeric, round, 2) %>% 
  mutate( target.lod = round(target.lod, 2)) %>% 
  create_dt()

```

```{r, eval = FALSE}

npc_eqtl_dist <- peaks.npc_rna.wEffs %>% 
  filter( lod.npc_rna > 7.5, local.npc_rna ==F)
# get distant eQTL with a good mediator ESC or NPC
med_ex <- eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  filter( mediation.lod < 0.4*(target.lod),
          !str_detect(target.symbol, "-ps"),
          !mediator.symbol %in% c("Cwc22","Gm13698","Gm13696","Gm13694") )%>% # get high lod drop examples and filter pseudogenes
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>%  # getting top 2 best mediators to test with bmediatR
  filter(type =="npc") 
    # select( target.symbol, target.chr, mediator.symbol, mediator.chr, qtl.chr, target.lod,mediation.lod,  type) %>% 
  # distinct() %>%  
  # mutate_if(is.numeric, round, 2) %>% 
  # mutate( target.lod = round(target.lod, 2)) %>% 
  # create_dt()

# Let's run bmediatr on all of these! 
#create an empty list to store results
med_ex_results <- c()

#create a for loop to run mediation with `bmediatR`
for(i in 1:nrow(med_ex)){
  gene_expression <- exprZ.npc_rna[,med_ex$target.id[i]]
  marker <- map_dat2 %>%
       mutate(diff = abs(pos_bp - med_ex$qtl.pos[i]*1e06)) %>%
       slice_min(diff)
  mediator <- exprZ.npc_rna[,colnames(exprZ.npc_rna) !=med_ex$target.id[i]]
  genotype <- pull_genoprobpos(
    genoprobs = probs.npc_rna,
    marker = marker$marker
  )
  bmediatr_scan <- bmediatR(
    y = gene_expression,
    M = mediator,
    X = genotype,
    Z = covar.npc_rna
  )
  med_ex_results[[i]] <- bmediatr_scan
}

med_ex_post_odds <- c()
for(i in 1:length(med_ex_results)){
  med_ex_post_odds[[i]] <- (med_ex_results[[i]]$ln_post_odds)[med_ex$mediator.id[i],]
 
}

mediation_results_odds <- do.call(rbind, med_ex_post_odds) %>% 
  as_tibble() %>%
  cbind( med_ex 
         ) %>% 
  pivot_longer(mediation:y_depends_m, names_to = "mediation_type", values_to = "odds") 
  

save( mediation_results_odds, file = here("_data","NPC_dist_eQTL_bmediatr_ex.RData"))
```


## Bmediatr with NPC-NPC mediation examples

```{r, eval = T, fig.height=7, fig.width=10}

load(here("_data","NPC_dist_eQTL_bmediatr_ex.RData"))

mediation_results_odds %>% 
  filter( mediation_type %in% c("colocal","complete","mediation","partial")) %>% 
  ggplot()+
  aes(x = target.symbol,
      y = odds,
      col = mediation_type,
      fill = mediation_type)+
  geom_point()+
  theme_pubclean( base_size = 14)+
  theme(legend.position = "right",
        axis.text.x = element_blank())+
  ylim(-60,60)+
  ylab("Log Posterior Odds")+
  xlab("")+
  ggtitle("NPC mediation examples")

```


```{r}

#load(here("_data","NPC_dist_eQTL_bmediatr_ex.RData"))

mediation_results_odds %>% 
  filter( mediation_type %in% c("mediation","partial","complete"), 
          odds > 10) %>% 
  select( target.symbol, qtl.chr, mediator.symbol, target.lod, mediation.lod, mediation_type, `Ln(odds)`=odds) %>% 
  arrange(desc(`Ln(odds)`)) %>% 
  mutate_if( is.numeric, round, 2) %>% 
  create_dt()

```


### Highlight: Slc38a10 mediates Sfpq, Rdx, D16Ertd472e distant eQTLs on chr 11

All three distant eQTL have negative lod for partial mediation and colocal models, moderate positive lod for complete and high positive lod for mediation suggesting complete mediation. 

```{r}

# overlapping scans for all 4 genes on chr11 + add Slc38a10 gene on the genome
# get scans:
ex_qtls <- peaks.npc_rna.wEffs %>% 
  filter( mgi_symbol %in% c("Slc38a10", "Sfpq","Rdx","D16Ertd472e"),
          peak_chr == 11, 
          lod.npc_rna>7.5)

qtl_scans <- scan1(genoprobs = probs.npc_rna, 
                   pheno = exprZ.npc_rna[,ex_qtls$ensembl_gene_id,drop = F], 
                   kinship = kinship_loco.npc_rna,
                   addcovar = covar.npc_rna)

qtl_scans %>% 
  as_tibble( rownames = "marker") %>% 
  left_join( map_dat2) %>% 
  filter( chr == 11) %>% 
  pivot_longer( cols = 2:5, 
                names_to = "ensembl_gene_id", 
                values_to = "lod") %>% 
  left_join( all.genes %>% select(ensembl_gene_id, mgi_symbol)) %>%  
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod,
      col = mgi_symbol
      )+
    geom_line( size = 1)+
    theme_pubclean( base_size = 18)+
  xlab("Chr 11 location (Mbp)")+
  ylab( "LOD score")+
  labs(col = "Gene name")+
  xlim(50,125)+
  geom_segment( x = ex_qtls$gene_start[2]/1e06, xend = ex_qtls$gene_end[2]/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= ex_qtls$midpoint[2]/1e06, y = -0.8, label ="Slc38a10", size =6, fontface = "italic")-> qtl_ex_lod_plot


# allele effect plot

ex_qtls %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".npc_rna")) %>% 
  pivot_longer( cols = paste0(LETTERS[1:8], ".npc_rna"), 
          names_to = "effect",
          values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = mgi_symbol,
       group = mgi_symbol)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> qtl_ex_haplotype_plot


```



```{r npc_qtl_ex_lod_haps_plot, fig.height=5, fig.width=10}

ggarrange( qtl_ex_lod_plot, qtl_ex_haplotype_plot, widths = c(1, 0.5))

```

```{r npc_eqtl_ex_bmediatr_plots, fig.height=4, fig.width=5}

mediation_results_odds %>% 
  filter(target.symbol %in% c("Sfpq","Rdx","D16Ertd472e")) %>% 
  filter( mediation_type %in% c("colocal","complete","mediation","partial","reactive")) %>% 
  ggplot()+
  aes(x = target.symbol,
      y = odds,
      col = mediation_type,
      fill = mediation_type)+
  geom_point( size = 3)+
  theme_pubclean( base_size = 14)+
  theme(legend.position = "right")+
  ylim(-50,20)+
  ylab("Log Posterior Odds")+
  xlab("")

```

Association mapping for Slc38a10 to identify SNPs:

```{r}

pmap_Mbp <- lapply( pmap, function(x){x/1e06})

slc_qtl <- peaks.npc_rna.wEffs %>% 
  filter( mgi_symbol %in% c("Slc38a10"),
          peak_chr == 11, 
          lod.npc_rna>7.5)

probs <- probs.npc_rna
attr(probs, "is_x_chr") <- NULL
slc_as <- scan1snps(genoprobs = probs, 
                     map = pmap_Mbp, 
                     pheno = exprZ.npc_rna[,slc_qtl$ensembl_gene_id,drop = F],
                     kinship = kinship_loco.npc_rna[[slc_qtl$peak_chr]],
                     addcovar = covar.npc_rna, 
                     chr = as.character(slc_qtl$peak_chr),
                     query_func = query_variants,
                     start = slc_qtl$interp_bp_peak.npc_rna/1e06 - 5,
                     end = slc_qtl$interp_bp_peak.npc_rna/1e06 + 5,
                     keep_all_snps = F
                    )


genes <- query_genes( slc_qtl$peak_chr, 
                      slc_qtl$interp_bp_peak.npc_rna/1e06 - 5, 
                      slc_qtl$interp_bp_peak.npc_rna/1e06 + 5) %>%  
  distinct()


```


```{r slc_as_plot, fig.width=12, fig.height=8}

par(mar=c(4.1, 4.1,2, 1))
plot(slc_as$lod,
     slc_as$snpinfo,
     #drop_hilit = 1,
     genes = genes, cex.lab =1.5)

```


SNPs with matching split in the area:

```{r slc_as_table}

slc_as %>% 
   as.data.frame() %>% 
  filter( ENSMUSG00000061306 > 5) %>% 
  select( LOD = ENSMUSG00000061306,
          snp_id = snpinfo.snp_id,
          pos = snpinfo.pos,
          alleles = snpinfo.alleles,
          consequence = snpinfo.consequence,
          ensembl_gene_id = snpinfo.ensembl_gene,
          `129` = snpinfo.129S1_SvImJ,
          `AJ` = snpinfo.A_J,
          `CAST` = snpinfo.CAST_EiJ,
          `NZO` = snpinfo.NZO_HlLtJ,
          `PWK` = snpinfo.PWK_PhJ,
          `B6` = snpinfo.C57BL_6J,
          `NOD` = snpinfo.NOD_ShiLtJ,
          `WSB` = snpinfo.WSB_EiJ
          ) %>% 
  filter( `129`==`CAST`, `CAST` == `PWK` & `WSB` == `B6`, `B6`==`NZO`, `B6`==`NOD`) %>%
  left_join(., select(all.genes, ensembl_gene_id, mgi_symbol) %>%  distinct() ) %>% 
  mutate_if(is.numeric, round, 1) %>% 
  arrange(desc(LOD)) %>% 
  create_dt()

```

### Higlight: Zfp943 mediating Ppp5c distant eQTL on chr 17



<!-- ## Mediation example: Ezh1 mediated by Pign -->

<!-- I am getting many genes with higher odds that are >10Mb away from the eQTL peak. -->

<!-- ```{r, fig.width=10, fig.height=5, warning=FALSE, message=FALSE} -->

<!-- # run bmediatR -->
<!-- ezh1_eqtl <- peaks.npc_rna.wesc.overlap %>%  -->
<!--   filter( lod.npc_rna >7.5, mgi_symbol =="Ezh1", local.npc_rna == F) -->

<!-- marker <- map_dat2 %>% -->
<!--        mutate(diff = abs(pos_bp - ezh1_eqtl$interp_bp_peak.npc_rna)) %>% -->
<!--        slice_min(diff) -->
<!-- genotype <- pull_genoprobpos( -->
<!--     genoprobs = probs.npc_rna, -->
<!--     marker = marker$marker -->
<!--   ) -->

<!-- ezh1_bmediatr_scan <- bmediatR( -->
<!--     y = exprZ.npc_rna[,ezh1_eqtl$ensembl_gene_id,drop=FALSE], -->
<!--     M = exprZ.npc_rna[,colnames(exprZ.npc_rna) != ezh1_eqtl$ensembl_gene_id], -->
<!--     X = genotype,  -->
<!--     Z = covar.npc_rna -->
<!--   ) -->

<!-- Pign_details <- npc.genes %>% -->
<!--    filter(mgi_symbol=="Pign")  -->

<!-- # exp(ezh1_bmediatr_scan$ln_post_c)[Pign_details$ensembl_gene_id,] -->
<!-- # (ezh1_bmediatr_scan$ln_post_odds)[Pign_details$ensembl_gene_id,] -->

<!-- # plot_posterior_bar(ezh1_bmediatr_scan, med_annot = npc.genes %>% mutate( middle = midpoint ,symbol = mgi_symbol), med_var = "ensembl_gene_id", mediator_id = Pign_details$ensembl_gene_id) -->
<!-- plot_posterior_odds(bmediatR_object = ezh1_bmediatr_scan, model_type = "mediation",med_annot = npc.genes %>% mutate( chr = gene_chr, middle = midpoint ,symbol = mgi_symbol), med_var = "ensembl_gene_id", bgcol = "white", altbgcol = "white", include_chr = 1, label_thresh = 6, label_thresh_greater_than = T) -->


<!-- ``` -->


<!-- ## Mediation example: Pitpnc1 mediating Dtna, Etv5, Sumf1, Srpk2 -->

<!-- The only one that shows odds >0 for mediation is Srpk2, plotted below.  -->

<!-- ```{r, warning=FALSE, message=FALSE} -->

<!-- # run bmediatr -->
<!-- med_ex <- eqtl_npc_rna_meds %>%  -->
<!--   filter( target.symbol %in% c("Dtna","Etv5", "Sumf1", "Srpk2"), qtl.chr ==11) -->

<!-- #create an empty list to store results -->
<!-- med_ex_results <- c() -->

<!-- #create a for loop to run mediation with `bmediatR` -->
<!-- for(i in 1:nrow(med_ex)){ -->
<!--   gene_expression <- exprZ.npc_rna[,med_ex$target.id[i]] -->
<!--   marker <- map_dat2 %>% -->
<!--        mutate(diff = abs(pos_bp - med_ex$qtl.pos[i]*1e06)) %>% -->
<!--        slice_min(diff) -->
<!--   mediator <- exprZ.npc_rna[,colnames(exprZ.npc_rna) !=med_ex$target.id[i]] -->
<!--   genotype <- pull_genoprobpos( -->
<!--     genoprobs = probs.npc_rna, -->
<!--     marker = marker$marker -->
<!--   ) -->
<!--   bmediatr_scan <- bmediatR( -->
<!--     y = gene_expression, -->
<!--     M = mediator, -->
<!--     X = genotype,  -->
<!--     Z = covar.npc_rna -->
<!--   ) -->
<!--   med_ex_results[[i]] <- bmediatr_scan -->
<!-- } -->
<!-- ``` -->

<!-- ```{r, fig.width=10, fig.height=5, warning=FALSE, message=FALSE} -->
<!-- pitpnc1_dets <- all.genes %>%  -->
<!--   filter( mgi_symbol=="Pitpnc1") -->

<!-- #med_ex_results[[4]]$ln_post_odds[pitpnc1_dets$ensembl_gene_id,] -->

<!-- plot_posterior_odds(bmediatR_object = med_ex_results[[4]],  -->
<!--                     model_type = "mediation", -->
<!--                     med_annot = npc.genes %>%  -->
<!--                       mutate( chr = gene_chr, middle = midpoint ,symbol = mgi_symbol),  -->
<!--                     med_var = "ensembl_gene_id",  -->
<!--                     bgcol = "white", altbgcol = "white",  -->
<!--                     include_chr = 11, label_thresh = 10, label_thresh_greater_than = T) -->

<!-- ``` -->

## eQTL with best mediator ESC rna

```{r}

eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>% 
  #filter( target.symbol == mediator.symbol) %>% 
  #slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, target.chr, mediator.symbol, mediator.chr, qtl.chr, target.lod,mediation.lod,  type) %>% 
  filter(type =="esc") %>% 
  mutate_if(is.numeric, round, 2) %>% 
  mutate( target.lod = round(target.lod, 2)) %>% 
  create_dt()

  
```

For any of these NPC eQTL are we getting a best mediator that is not expressed in NPCs but is expressed in ESCs? Yes!

```{r}

eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>% 
  #filter( target.symbol == mediator.symbol) %>% 
  #slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, target.chr, mediator.symbol, mediator.id, mediator.chr, qtl.chr, target.lod,mediation.lod,  type) %>% 
  filter(type =="esc") %>% 
  filter( !mediator.symbol %in% npc.genes$mgi_symbol) %>% 
  mutate_if(is.numeric, round, 2) %>% 
  mutate( target.lod = round(target.lod, 2)) %>% 
  create_dt()


```



Let's make some interactive plots to look at mediation results with large lod drops:

```{r}

esc_med_exs <- eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>% 
  filter(type =="esc",
         !mediator.symbol %in% npc.genes$mgi_symbol, # mediator not in npc gene list
          mediation.lod < 0.4*(target.lod),
          !str_detect(target.symbol, "-ps"),
          mediator.symbol != "Cwc22") # get high lod drop examples and filter pseudogenes



```


## Mediation example: Zfp51

```{r Zfp51_plot_prep, warning=FALSE, message=FALSE}

Zfp51_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.npc_rna >7.5, mgi_symbol =="Zfp51", local.npc_rna == F)
Zfp51_scan <- scan1( pheno = exprZ.npc_rna[,Zfp51_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)

Zfp51_meds <- eqtl_esc_rna_meds %>% 
  filter( target.symbol =="Zfp51", qtl.chr == Zfp51_eqtl$peak_chr) %>% 
  mutate( type = "esc") %>% 
  rbind( eqtl_npc_rna_meds %>% filter(target.symbol =="Zfp51", qtl.chr == Zfp51_eqtl$peak_chr) %>% mutate(type ="npc")) %>% 
  mutate( mediator_mid = (mediator.start+mediator.end)/2)


```

```{r Zfp51_plot, fig.width=12, fig.height=5, warning=FALSE, message=FALSE}

# lod plot + annotate the gene location too
Zfp51_scan %>% 
  as.data.frame( ) %>% 
  rename( eqtl = Zfp51_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(Zfp51_scan)[[1]]) %>% 
  left_join(map_dat2) -> Zfp51_scan_df
Zfp51_scan_df$cumsum_pos_bp <- Zfp51_scan_df$pos_bp + chrom_lens_offset[Zfp51_scan_df$chr]

Zfp51_scan_df %>% 
  #filter( chr == Zfp51_eqtl$peak_chr) %>% 
  ggplot()+
    aes( 
      x= cumsum_pos_bp,
      y = eqtl
      )+
  geom_line( size = 1, col = qtl.colors[["npc_rna"]],alpha = 0.3)+
  theme_pubclean( base_size = 18)+
  xlab("Chr")+
  #xlab(paste0("Chr ",Zfp51_eqtl$peak_chr," location (bp)"))+
  ylab( "LOD score")+
  geom_segment( x = Zfp51_eqtl$gene_start+chrom_lens_offset[Zfp51_eqtl$gene_chr], 
                xend = Zfp51_eqtl$gene_end+chrom_lens_offset[Zfp51_eqtl$gene_chr], 
                y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= Zfp51_eqtl$midpoint+chrom_lens_offset[Zfp51_eqtl$gene_chr], y = -0.5, label ="Zfp51", size =6, fontface = 'italic')+
  scale_x_discrete( name = "Chr",
                    limits = chrom_lens_midpt, 
                    labels = names(chrom_lens), 
                    expand = expansion( mult = 0.05)) -> Zfp51_lod_plot

Zfp51_meds_min <- Zfp51_meds %>% 
  slice_min( mediation.lod)
Zfp51_lod_plot+
  geom_point( data = Zfp51_meds, 
              aes( x = mediator_mid+chrom_lens_offset[Zfp51_eqtl$peak_chr],
                   y = mediation.lod,
                   col = type),
              size = 4,
              alpha = 0.8)+
  scale_color_manual( values = c("esc"=qtl.colors[["esc_rna"]], "npc"=qtl.colors[["npc_rna"]]),
                      labels = c("esc"="ESC transcript", "npc"="NPC transcript"), name = "Mediator")

# overlay mediation lods as dots + add text for the best one

```

<!-- Let's look at bmediatR results for ZFP51: -->

<!-- ```{r, fig.height=4, fig.width=5} -->

<!-- Zfp51_meds %>%  -->
<!--   slice_min( mediation.lod, n= 10) -> Zfp51_meds_min10 -->

<!-- zfp51_i <- which(data$target.symbol =="Zfp51") -->

<!-- (results[[zfp51_i]]$ln_post_odds)[Zfp51_meds_min10$mediator.id, ] %>%  -->
<!--   as_tibble( rownames = "mediator.id") %>%  -->
<!--   left_join( Zfp51_meds_min10) %>%  -->
<!--   pivot_longer(2:9, names_to = "group", values_to = "odds") %>% -->
<!--   filter( group =="mediation") %>%  -->
<!--   mutate( lod_drop =target.lod - mediation.lod) %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = odds, -->
<!--       y = lod_drop)+ -->
<!--   geom_point()+ -->
<!--   geom_text(  aes(label = mediator.symbol), nudge_x = 0.22, nudge_y = 0.04)+ -->
<!--   theme_pubclean()+ -->
<!--   #ylim(6,8)+ -->
<!--   #xlim(0, 5.5)+ -->
<!--   xlab("Log posterior odds for mediation (partial or complete)")+ -->
<!--   ylab("Drop in LOD score with mediation") -->

<!-- ``` -->

<!-- ```{r, fig.height=5, fig.width=10} -->

<!-- plot_posterior_odds(bmediatR_object = results[[zfp51_i]],  -->
<!--                     model_type = "mediation", -->
<!--                     med_annot = npc.genes %>%  -->
<!--                       mutate( chr = gene_chr, middle = midpoint ,symbol = mgi_symbol),  -->
<!--                     med_var = "ensembl_gene_id",  -->
<!--                     bgcol = "white", altbgcol = "white",  -->
<!--                     include_chr = Zfp51_eqtl$peak_chr[1], label_thresh = 0, label_thresh_greater_than = T) -->

<!-- ``` -->

## BmediatR doesn't work with ESC expression

Can bmediatR help us get directionality? We don't get any with high odds! 

```{r}
# #create an empty list to store results
# results <- c()
# 
# #create a for loop to run mediation with `bmediatR`
# for(i in 1:nrow(data)){
#   gene_expression <- exprZ.npc_rna[,data$target.id[i]]
#   marker <- map_dat2 %>%
#        mutate(diff = abs(pos_bp - data$interp_bp_peak.npc_rna[i])) %>%
#        slice_min(diff)
#   mediator <- exprZ.esc_rna[,colnames(exprZ.esc_rna) !=data$target.id[i]]
#   genotype <- pull_genoprobpos(
#     genoprobs = probs.npc_rna,
#     marker = marker$marker
#   )
#   bmediatr_scan <- bmediatR(
#     y = gene_expression,
#     M = mediator,
#     X = genotype, 
#     Z = covar.npc_rna
#   )
#   results[[i]] <- bmediatr_scan
# }
# 
# save(results, file = here("data","bmediatr_results_mNPC_eQTL_wESC.RData"))


load(here("_data","bmediatr_results_mNPC_eQTL_wESC.RData")) #results


data <- eqtl_npc_rna_meds %>% 
  mutate( type = "npc") %>% 
  rbind( eqtl_esc_rna_meds %>% mutate( type ="esc")) %>% 
  inner_join(
    npc_eqtl_dist %>% select( target.id = ensembl_gene_id, qtl.chr = peak_chr, target.lod = lod.npc_rna)
  ) %>% 
  group_by( target.id, target.lod, qtl.chr) %>% 
  slice_min( mediation.lod, n = 1) %>% 
  #filter( target.symbol == mediator.symbol) %>% 
  #slice_min( mediation.lod) %>% 
  arrange( desc(target.lod)) %>% 
  select( target.symbol, target.chr, mediator.symbol, mediator.id, mediator.chr, qtl.chr, target.lod,mediation.lod,  type) %>% 
  filter(type =="esc") %>% 
  filter( !mediator.symbol %in% npc.genes$mgi_symbol)
#find posterior probabilities 
post_results_c <- c()
post_results_odds <- c()

for (i in 1:length(results)) {
  post_results_c[[i]] <- exp(results[[i]]$ln_post_c)[data$mediator.id[i], ]
  post_results_odds[[i]] <- (results[[i]]$ln_post_odds)[data$mediator.id[i], ]
}

```


```{r, fig.width=6, fig.height=8}

mediation_results_odds <- do.call(rbind, post_results_odds) %>%
  as_tibble() %>%
  cbind( data
         ) %>%
  pivot_longer(1:8, names_to = "group", values_to = "odds") 

mediation_results_odds %>%
  filter( !str_detect(group, "y_"),
          !group %in% c("mediation_or_colocal","reactive")) %>% 
  ggplot()+
  aes(x = odds,
      y = target.id,
      col = group,
      fill = group)+
  geom_point()+
  theme_pubclean()+
  theme(legend.position = "right",
        axis.text.y = element_blank())+
  #xlim(0,1)+
  xlab("Log Posterior Odds")+
  ylab("")+
  #scale_x_log10()+
  ggtitle("Mediation with mNPC eQTL wESC Expression")

```

<!-- ```{r, fig.height=8, fig.width=6} -->

<!-- mediation_results <- do.call(rbind, post_results_c) %>% -->
<!--   as_tibble() %>% -->
<!--   dplyr::rename( -->
<!--     ML1 = `0,0,0`, -->
<!--     ML2 = `0,1,0`, -->
<!--     ML3 = `1,0,0`, -->
<!--     complete = `1,1,0`, -->
<!--     ML5 = `0,0,1`, -->
<!--     ML6 = `0,1,1`, -->
<!--     colocal = `1,0,1`, -->
<!--     partial = `1,1,1`, -->
<!--     ML9 = `0,*,0`, -->
<!--     ML10 = `1,*,0`, -->
<!--     ML11 = `0,*,1`, -->
<!--     ML12 = `1,*,1` -->
<!--   ) %>% -->
<!--   cbind( data -->
<!--          ) %>% -->
<!--   pivot_longer(ML1:ML12, names_to = "group", values_to = "prob") %>% -->
<!--   mutate( mediation_type = case_when( -->
<!--     group %in% c("complete","colocal","partial")~group, -->
<!--     group %in% c(paste0("ML",seq(1:12)))~"other non-med" -->
<!--   )) -->

<!-- mediation_results %>% -->
<!--   filter( !mediation_type =="other non-med") %>%  -->
<!--   ggplot()+ -->
<!--   aes(x = prob, -->
<!--       y = target.id, -->
<!--       col = mediation_type, -->
<!--       fill = mediation_type)+ -->
<!--   geom_point()+ -->
<!--   theme_pubclean()+ -->
<!--   theme(legend.position = "right", -->
<!--         axis.text.y = element_blank())+ -->
<!--   xlim(0,1)+ -->
<!--   xlab("Posterior Probabilities")+ -->
<!--   ylab("")+ -->
<!--   #scale_x_log10()+ -->
<!--   ggtitle("Mediation with mNPC eQTL wESC Expression") -->


<!-- ``` -->

