---
title: ""
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: 
      collapsed: false
      smooth_scroll: false
    df_print: paged
    code_folding: hide
---

```{r setup}

options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(progress=FALSE)

```

# DO mNPC eQTL hotspots

```{r transbands_lod6, warning=FALSE, message=FALSE}
# eQTL map for NPC rna
# prep the objects
#map_dat2 <- map_dat2 %>% mutate(pos_bp = as.numeric(pos_bp), pos_cM = as.numeric(pos))
map_dat2$chromF <- factor(map_dat2$chrom, levels = c(as.character(1:19), "X"))
chrom_markers <- select(map_dat2, chromF, n) %>%
  rename(chrom = chromF) %>%
  group_by(chrom) %>%
  summarize(start = min(n), end = max(n)) %>%
  GenomicRanges::GRanges()
windows <- unlist(GenomicRanges::slidingWindows(chrom_markers, width = 50, step = 10))
markers_bynum <- select(map_dat2, chrom, n) %>%
  dplyr::rename(start = n) %>%
  mutate(end = start) %>%
  GenomicRanges::GRanges()

distant_esc_rna <- filter(peaks.esc_rna.wEffs, lod.esc_rna > 7.5, !is.na(local.esc_rna) & !(local.esc_rna)) %>%
  select(peak_chr, interp_bp_peak.esc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.esc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()
distant_esc_rna_lod6 <- filter(peaks.esc_rna.wEffs, lod.esc_rna > 6, !is.na(local.esc_rna) & !(local.esc_rna)) %>%
  select(peak_chr, interp_bp_peak.esc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.esc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()

distant_npc_rna <- filter(peaks.npc_rna.wEffs, lod.npc_rna > 7.5, !is.na(local.npc_rna) & !(local.npc_rna)) %>%
  select(peak_chr, interp_bp_peak.npc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.npc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()
distant_npc_rna_lod6 <- filter(peaks.npc_rna.wEffs, lod.npc_rna > 6, !is.na(local.npc_rna) & !(local.npc_rna)) %>%
  select(peak_chr, interp_bp_peak.npc_rna) %>%
  dplyr::rename(chrom = peak_chr, end = interp_bp_peak.npc_rna) %>%
  mutate(start = end) %>%
  GenomicRanges::GRanges()

markers <- select(map_dat2, chrom, pos_bp) %>%
  dplyr::rename(start = pos_bp) %>%
  mutate(end = start) %>%
  GenomicRanges::GRanges() # length 69,005
esc_hotspot <- GenomicRanges::nearest(distant_esc_rna, markers)
esc_hotspot_lod6 <- GenomicRanges::nearest(distant_esc_rna_lod6, markers)
npc_hotspot <- GenomicRanges::nearest(distant_npc_rna, markers)
npc_hotspot_lod6 <- GenomicRanges::nearest(distant_npc_rna_lod6, markers)
# assert_that(noNA(x), noNA(y))
# assert_that(noNA(x))
windows$distant_esc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[esc_hotspot])
windows$distant_esc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[esc_hotspot_lod6])
windows$distant_npc_rna <- GenomicRanges::countOverlaps(windows, markers_bynum[npc_hotspot])
windows$distant_npc_rna_lod6 <- GenomicRanges::countOverlaps(windows, markers_bynum[npc_hotspot_lod6])
window_counts <- tibble(
  chrom = as.character(GenomicRanges::seqnames(windows)),
  start = GenomicRanges::start(windows),
  end = GenomicRanges::end(windows),
  distant_esc_rna = windows$distant_esc_rna, 
  distant_npc_rna = windows$distant_npc_rna,
  distant_npc_rna_lod6 = windows$distant_npc_rna_lod6
)

# plotting
map_dat2 <- map_dat2 %>% mutate( pos_cM = as.numeric(pos))
mm <- match(window_counts$start, map_dat2$n)
m2 <- match(window_counts$end, map_dat2$n)
window_counts$pos_cM_start <- map_dat2$pos_cM[mm]
window_counts$pos_bp_start <- map_dat2$pos_bp[mm]
window_counts$pos_cM_end <- map_dat2$pos_cM[m2]
window_counts$pos_bp_end <- map_dat2$pos_bp[m2]
window_counts <- window_counts %>%
  mutate(midpoint = (pos_cM_end + pos_cM_start) / 2, 4)

x <- select(window_counts, chrom, starts_with("pos_bp"), starts_with("distant")) %>%
  filter(
    #distant_esc_rna > quantile(distant_esc_rna,0.995) | 
    #distant_npc_rna_lod6 > quantile(distant_npc_rna_lod6, 0.995) | 
           distant_npc_rna > quantile(distant_npc_rna,0.995) )
bands.esc.npc.rna <- x %>%
  rename(start = pos_bp_start, end = pos_bp_end) %>%
  GenomicRanges::GRanges() %>%
  GenomicRanges::reduce()
# reduce collapses overlapping windows into one big window. Works perfectly here.
bands.esc.npc.rna$distant_esc_rna <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_esc_rna)
bands.esc.npc.rna$distant_esc_rna_lod6<- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_esc_rna_lod6)
bands.esc.npc.rna$distant_npc_rna <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_npc_rna)
bands.esc.npc.rna$distant_npc_rna_lod6 <- GenomicRanges::countOverlaps(bands.esc.npc.rna, distant_npc_rna_lod6)


# plot(todrawR+zero+sp, mapMarkers, ylab="Number of distant QTL", col=qtl.colors[2], 
#     ylim=c(-2*sp, 800+2*sp), yaxt='n', bgcolor='gray90', altbgcolor='white')
# plot(zero-sp-todrawA, mapMarkers, add=TRUE, col=qtl.colors[1])
# axis(2, at=seq(0, 800, by=200)+sp, labels=c("400", "200", "0", "200", "400"))
# text(1600, 780, "ESC eQTL LOD>6", cex=1.2, col=qtl.colors[2], adj=c(1, 1))
# text(1600, 30, "NPC eQTL LOD>6", cex=1.2, col=qtl.colors[1], adj=c(1, 0))

```

To identify eQTL trans-bands, aka "hotspots" - regions of the genome that appear to modulate the expression of many genes and likely harbor an important transcriptional regulator - I broke up the genome into smaller, overlapping windows (1cM width, 0.25cM shifts) and counted the number of distant NPC eQTLs that mapped to that window. We defined a hotspot as any window that contained `r quantile((select(window_counts, chrom, starts_with("pos_bp"), starts_with("distant")))$distant_npc_rna,0.995)[[1]]` or more distant eQTL (the top 0.05% value for distant eQTL), and then collapsed overlapping and adjacent windows into one larger window to estimate the critical interval for each hotspot.

```{r}

bands.esc.npc.rna  %>% 
  as_tibble() %>% 
  mutate( chr = seqnames) -> bands.esc.npc.rna
bands.esc.npc.rna %>% 
    filter(distant_npc_rna > 20)  %>% 
  select( Chromosome = seqnames, `Start (bp)`=start, `End (bp)`=end, `# of distant eQTL (LOD >7.5)` = distant_npc_rna, 
          `# of distant eQTL (LOD >6)` = distant_npc_rna_lod6 ) %>% 
  create_dt()

```

## eQTL hotspot on chromosome 1 {.tabset .tabset-fade .tabset-pills}

**SUMMARY:** We identified a `r round((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 1))$width/1e06,1)`Mb wide NPC-specific eQTL trans-band on Chromosome 1 at `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 1))$start,big.mark=",")`-`r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 1))$end,big.mark=",")` bp that contains `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 1))$distant_npc_rna_lod6,big.mark=",")` NPC eQTL at the suggestive threshold of LOD \> 6. Target genes within the trans-band show an over-representation for genes involved in RNA splicing and the spindle assembly checkpoint; most of these genes are downregulated in NPCs relative to ESCs. Inferred founder allele effects of the target NPC eQTL show a 4:4 pattern where the wild-derived strains WSB and PWK group with laboratory strains A/J and NZO, while the wild-derived CAST strain groups with the laboratory strains B6, NOD, and 129. Surprisingly, comparison of mediation scores for the eQTLs (shown as LOD drops) in this NPC-specific trans-band suggest that transcript abundance in the ESCs is actually a better mediator of the eQTLs than the transcript abundance in the NPCs themselves. This suggests temporal regulation where the underlying causal variant directly affects expression of a transcriptional regulator of cell fate in the ESCs, but then its downstream trans regulatory effects are only observed later in the NPCs. *Pign* ESC transcript abundance is identified as the best candidate regulator for this trans-band; it is the best mediator for 157 out of 322 total eQTL, and *Pign* has a local eQTL with a similar founder allele grouping to what is observed for the NPC trans-band. NPC transcript abundance of Pign is also identified as a good candidate mediator, but does not perform as well as ESC transcript of *Pign*.

```{r, eval = FALSE}

bands.esc.npc.rna %>% filter(seqnames==1)

```

```{r chr1_eqtl}

chr1.npc.eQTL1 <- peaks.npc_rna.wesc.overlap %>% 
  filter( peak_chr == 1) %>% 
  select( -peak_cM.npc_rna) %>% 
  left_join(peaks.npc_rna.wEffs) %>% 
  filter( lod.npc_rna > 6 & !local.npc_rna & !is.na(local.npc_rna) &
            interp_bp_peak.npc_rna >= filter( bands.esc.npc.rna,chr==1)$start[1] & 
            interp_bp_peak.npc_rna  <= filter( bands.esc.npc.rna,chr==1)$end[1] ) %>% 
  mutate( lod.esc_rna = ifelse( is.na(lod.esc_rna), 0, lod.esc_rna)) # genes are not in the ESC data.

chr1.esc.eQTL1 <- peaks.esc_rna.wnpc.overlap %>% 
  filter( peak_chr == 1) %>%
  select( -peak_cM.esc_rna) %>% 
  left_join(peaks.esc_rna.wEffs) %>% 
  filter(lod.esc_rna >6 & !local.esc_rna & !is.na(local.esc_rna) & 
           interp_bp_peak.esc_rna >= filter( bands.esc.npc.rna,chr==1)$start[1] &
           interp_bp_peak.esc_rna  <= filter( bands.esc.npc.rna,chr==1)$end[1]) %>% 
  mutate( lod.npc_rna = ifelse( is.na(lod.npc_rna), 0, lod.npc_rna)) # gene is not found in the npc data 


```

### eQTL in the region

```{r chr1_eqtl_plot, fig.height=5, fig.width=5}

chr1.npc.eQTL1 %>% 
  mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match))  %>%
  ggplot()+
  aes(x=lod.esc_rna,y=lod.npc_rna,col=match)+
  geom_point(size=3,alpha=0.5)+
  theme_pubclean(base_size = 16)+
  geom_vline(xintercept = 6,linetype=2)+
  geom_hline(yintercept=6,linetype=2)+
  scale_color_manual(values= c("NPC eQTL"=qtl.colors[["npc_rna"]],
                               "ESC eQTL"=qtl.colors[["esc_rna"]],
                               "shared"=qtl.colors[["shared"]]))+
  xlab("LOD score from ESC eQTL scan")+
  ylab("LOD score from NPC eQTL scan")+
  xlim(0,15)+
  ylim(0,15) -> p

p+
  geom_point(
    data = (chr1.esc.eQTL1 %>% 
              mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match))
            ),
    size = 3,
    alpha = 0.5
    )
  

```

```{r}

chr1.npc.eQTL1 %>% 
  mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match)) %>% 
  select( ensembl_gene_id, mgi_symbol, peak_chr, gene_chr, lod.npc_rna, lod.esc_rna, interp_bp_peak.npc_rna, interp_bp_peak.esc_rna) %>% 
  mutate_if( is.numeric, round, 2) %>% 
  create_dt()

```

### ORA with NPC eQTL in the region

Note that I am running ORA for the genes with unique NPC eQTL using a custom background of the full NPC gene list instead of genes with distant eQTL.

```{r chr1_ora}

genes.w.eqtl <- peaks.npc_rna.wEffs %>% 
  filter(lod.npc_rna > 7.5) %>% 
  select(ensembl_gene_id, local.npc_rna) %>% 
  distinct() %>% 
  left_join(npc.genes) %>% 
  filter( ensembl_gene_id %in% shared.genes)


g.chr1 <- gost(query =  unique(chr1.npc.eQTL1$mgi_symbol),
                        organism = "mmusculus", 
                        domain_scope = "custom", 
                        custom_bg = npc.genes$mgi_symbol, #unique((genes.w.eqtl %>% filter(!local.npc_rna))$mgi_symbol)
                        correction_method = "fdr",
                        evcodes = TRUE)
g.chr1$result <- g.chr1$result %>% filter(term_size < 500)

```

```{r chr1_ora_plot}

gostplot(g.chr1, capped = F)

```

```{r chr1_ora_table}

g.chr1$result %>% 
  select(source, term_name, intersection_size, term_size , FDR = p_value, intersection) %>%
  filter( FDR < 0.01) %>%
  mutate_if(is.numeric, formatC, digits =2, format ="fg") %>%
  create_dt()
  
```


```{r, fig.height=8, fig.width=12}

# mRNA processing
# DNA repair
# chromatin organization

# cell division

# Mitotic Spindle Checkpoint
# chromosome segregation
# sister chromatid segregation
# nuclear chromosome segregation
# mitotic sister chromatid segregation
# Mitotic Prometaphase
# Cell Cycle Checkpoints
# Amplification of signal from unattached kinetochores via a MAD2 inhibitory signal
# Amplification of signal from the kinetochores
# Resolution of Sister Chromatid Cohesion
# Separation of Sister Chromatids
# M Phase
# Mitotic Anaphase

# qwraps2::lazyload_cache_dir("01d_esc_npc_deseq_cache/html/")

# any category with spindle, segregation
chr1_spindle_genes <- g.chr1$result %>% 
  filter( p_value <0.01) %>% 
  filter( term_name %in% c("mRNA processing",
                           "DNA repair",
                           "chromatin organization",
                           "chromosome segregation",
                           "cell division",
                           "Mitotic Spindle Checkpoint")) %>% 
  select(term_name, intersection) %>%
  separate_rows(intersection, sep = ",") %>%
  rename( mgi_symbol = intersection) %>%
  left_join( npc_esc_wilcox_df %>%
               select( ensembl_gene_id, log2foldchange, p) %>%
               left_join( all.genes %>%
                            select(mgi_symbol, ensembl_gene_id))) %>%
  select( mgi_symbol, term_name, log2foldchange)


g.chr1.all <- graph_from_data_frame(chr1_spindle_genes, directed=FALSE)
V(g.chr1.all)$lfc <- c((chr1_spindle_genes %>% 
  select(mgi_symbol, log2foldchange) %>% 
  distinct())$log2foldchange, rep(-10, length(unique(chr1_spindle_genes$term_name)))) 

p <- ggraph(g.chr1.all)+
  geom_edge_link0(col="gray", alpha=0.1)+
  geom_node_point(aes(col=lfc),size=5)+
  geom_node_text(aes(label = name), repel = TRUE )+
  scale_color_viridis( limits = c(-4,4))+
  theme_classic(base_size=10)+
  theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank())+
  labs(col="Log Fold Change in NPCs")

# ggsave(p, filename = here("figures","chr1_network_plot.svg"), width = 12, height = 8, units = "in", dpi = 300)

p


```

### Expression change in overrepresented pathways

Looking at expression change in genes identified in relation to spindle assembly checkpoint and related pathways. 

```{r deg_ora_pathways, fig.height=10, fig.width=16, eval = T}

# qwraps2::lazyload_cache_dir("01d_esc_npc_deseq_cache/html/")

# any category with spindle, segregation
chr1_spindle_genes <- g.chr1$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "(?i)spindle") |
            str_detect(term_name, "(?i)segregation")|
            str_detect(term_name, "(?i)chromatid") |
            str_detect(term_name, "(?i)kinetochore") |
            str_detect(intersection,"Mad2l1")) %>% 
  select(term_name, intersection) %>%
  separate_rows(intersection, sep = ",") %>%
  rename( mgi_symbol = intersection) %>%
  left_join( npc_esc_wilcox_df %>%
               select( ensembl_gene_id, log2foldchange, p) %>%
               left_join( all.genes %>%
                            select(mgi_symbol, ensembl_gene_id))) %>%
  select( mgi_symbol, term_name, log2foldchange)


g.chr1.all <- graph_from_data_frame(chr1_spindle_genes, directed=FALSE)
V(g.chr1.all)$lfc <- c((chr1_spindle_genes %>% 
  select(mgi_symbol, log2foldchange) %>% 
  distinct())$log2foldchange, rep(-10, length(unique(chr1_spindle_genes$term_name)))) 

ggraph(g.chr1.all)+
  geom_edge_link0(col="gray", alpha=0.1)+
  geom_node_point(aes(col=lfc),size=5)+
  geom_node_text(aes(label = name), repel = TRUE )+
  scale_color_viridis( limits = c(-4,4))+
  theme_classic(base_size=10)+
  theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank())+
  labs(col="Log Fold Change in NPCs")

```

### NPC eQTL allele effects

```{r npc_allele_effects_chr1_sig, fig.width=6, fig.height=4}

haps <- c("A","B","C","D","E","F","G","H")

chr1.npc.eQTL1.mat <- chr1.npc.eQTL1 %>% 
  filter(lod.npc_rna > 7.5) %>% 
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol")) %>% 
  distinct() %>%
  column_to_rownames("mgi_symbol") %>%
  as.matrix() %>% 
  t()

annotation <- chr1.npc.eQTL1 %>% 
  dplyr::mutate(match = ifelse(match =="npc_rna","NPC_eQTL",match)) %>%
  filter(mgi_symbol %in% colnames(chr1.npc.eQTL1.mat)) %>%
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol","match")) %>% 
  distinct() %>% 
  column_to_rownames("mgi_symbol") %>% 
  select(match) 

annotation_row <- data.frame(strain=c("AJ","B6","129","NOD","NZO","CAST","PWK","WSB"))
rownames(annotation_row) <- rownames(chr1.npc.eQTL1.mat)  

annot.colors <- list(strain = founder_colors,
                     match=c( NPC_eQTL=qtl.colors[["npc_rna"]], shared=qtl.colors[["shared"]]))

# svg(filename = here("figures","chr1_hotspot_allele_eff.svg"),width = 800, height = 600)
pheatmap(chr1.npc.eQTL1.mat,
                                   cluster_rows=T,show_rownames=FALSE,
                                   cluster_cols=T, show_colnames = FALSE, 
                                   clustering_method = "complete", 
                                   scale="none",
                                   clustering_distance_cols="correlation",
                                   clustering_distance_rows="correlation",
                                   #main = "Founder allele effects of chr 1 NPC eQTL hotspot genes",
                                   annotation_col= annotation, 
                                   annotation_row = annotation_row, 
                                   annotation_colors = annot.colors, 
                                   cutree_rows = 2)


# dev.off()

```

```{r npc_allele_effects_chr1, fig.width=8, fig.height=6 }

haps <- c("A","B","C","D","E","F","G","H")

chr1.npc.eQTL1.mat <- chr1.npc.eQTL1 %>% 
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol")) %>% 
  distinct() %>%
  column_to_rownames("mgi_symbol") %>%
  as.matrix() %>% 
  t()

annotation <- chr1.npc.eQTL1 %>% 
  dplyr::mutate(match = ifelse(match =="npc_rna","NPC_eQTL",match)) %>%
  filter(mgi_symbol %in% colnames(chr1.npc.eQTL1.mat)) %>%
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol","match")) %>% 
  distinct() %>% 
  column_to_rownames("mgi_symbol") %>% 
  select(match) 

annotation_row <- data.frame(strain=c("AJ","B6","129","NOD","NZO","CAST","PWK","WSB"))
rownames(annotation_row) <- rownames(chr1.npc.eQTL1.mat)  

annot.colors <- list(strain = founder_colors,
                     match=c( NPC_eQTL=qtl.colors[["npc_rna"]], shared=qtl.colors[["shared"]]))

# svg(filename = here("figures","chr1_hotspot_allele_eff.svg"),width = 800, height = 600)
pheatmap(chr1.npc.eQTL1.mat,
                                   cluster_rows=T,show_rownames=FALSE,
                                   cluster_cols=T, show_colnames = FALSE, 
                                   clustering_method = "complete", 
                                   scale="none",
                                   clustering_distance_cols="correlation",
                                   clustering_distance_rows="correlation",
                                   #main = "Founder allele effects of chr 1 NPC eQTL hotspot genes",
                                   annotation_col= annotation, 
                                   annotation_row = annotation_row, 
                                   annotation_colors = annot.colors, 
                                   cutree_rows = 2)


# dev.off()

```


### Mediation of target eQTL

```{r chr1_npc_eqtl_run_mediation, cache = TRUE, eval = FALSE}

chr1_npc_eqtl_npc_meds <- do_global_mediation(gene.name = chr1.npc.eQTL1$mgi_symbol[1],
                    p.chr = chr1.npc.eQTL1$peak_chr[1],
                    peak.info = chr1.npc.eQTL1,
                    probs = shared.probs.npc_rna,#probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],#exprZ.npc_rna,
                    expr.mediator = exprZ.npc_rna[shared.samples,,drop=FALSE],#exprZ.npc_rna,
                    covar= covar.npc_rna[shared.samples,,drop=FALSE],#covar.npc_rna,
                    med.type = "n-n",
                    genes = npc.genes)

for(i in 2:nrow(chr1.npc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr1.npc.eQTL1$mgi_symbol[i],
                    p.chr = chr1.npc.eQTL1$peak_chr[i],
                    peak.info = chr1.npc.eQTL1,
                    probs = shared.probs.npc_rna,#probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],#exprZ.npc_rna,
                    expr.mediator = exprZ.npc_rna[shared.samples,,drop=FALSE],#exprZ.npc_rna,
                    covar= covar.npc_rna[shared.samples,,drop=FALSE],#covar.npc_rna,
                    med.type = "n-n",
                    genes = npc.genes)
  chr1_npc_eqtl_npc_meds <- rbind(chr1_npc_eqtl_npc_meds, meds)
}

chr1_npc_eqtl_esc_meds <- do_global_mediation(gene.name = chr1.npc.eQTL1$mgi_symbol[1],
                    p.chr = chr1.npc.eQTL1$peak_chr[1],
                    peak.info = chr1.npc.eQTL1,
                    probs = shared.probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.esc_rna[shared.samples,,drop=F],
                    covar= covar.npc_rna[shared.samples,,drop=F],
                    med.type = "e-n",
                    genes = npc.genes)

for(i in 2:nrow(chr1.npc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr1.npc.eQTL1$mgi_symbol[i],
                    p.chr = chr1.npc.eQTL1$peak_chr[i],
                    peak.info = chr1.npc.eQTL1,
                    probs = shared.probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.esc_rna[shared.samples,,drop=F],
                    covar= covar.npc_rna[shared.samples,,drop=F],
                    med.type = "e-n",
                    genes = npc.genes)
  chr1_npc_eqtl_esc_meds <- rbind(chr1_npc_eqtl_esc_meds, meds)
}

```

```{r chr1_meds_manual_run, eval = FALSE}

# I reran the mediation locally to see if the results I get are the same and they are identical. So no worries on the mediation results! Mainly, filtering based on z-threshold removes a lot of mediators that are not 'good enough'. 

chr1_npc_eqtl_npc_meds %>% 
  mutate( type = "npc") %>% 
  rbind(chr1_npc_eqtl_esc_meds %>% 
          mutate(type ="esc")) %>% 
    left_join( chr1.npc.eQTL1 %>% 
                 select( target = mgi_symbol, 
                         peak_chr, 
                         target_lod = lod.npc_rna,
                         target_pos = interp_bp_peak.npc_rna,
                         target_chr = gene_chr)) %>%
  select(
    target,
    peak_chr,
    target_lod,
    target_pos,
    target_chr,
    mediator_id = ensembl_gene_id, 
    mediator,
    mediator_chr = gene_chr, 
    mediator_midpoint = midpoint,
    mediation_lod = LOD,
    type
  ) %>% 
  group_by( target, peak_chr, target_lod, target_chr,type) %>% 
  mutate( scaled_LOD = scale(mediation_lod)) %>% 
  filter( abs(target_pos-mediator_midpoint) <= 10e6 &   # filter within 10Mb
          peak_chr == mediator_chr,
          scaled_LOD < -4 # z threshold -4
          )   -> chr1_npc_eqtl_meds
  
chr1_npc_eqtl_meds %>% 
  mutate(mediation_lod = ifelse(target == mediator, NA, mediation_lod)) %>%
  mutate(lod_drop = target_lod - mediation_lod) %>%
  group_by(target) %>%
  arrange(mediation_lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr1.meds.ranked

chr1.meds.ranked.sum <- chr1.meds.ranked %>%
  filter(rank %in% c( 1)) %>% 
  group_by(mediator,type) %>%
  summarize(n = length(target), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator = str_c(toupper(type)," , " ,mediator))

chr1.meds.ranked.sum

```

```{r chr1_meds, results='hide'}

eqtl_esc_rna_meds %>%
  inner_join(select(chr1.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr1.esc.meds

eqtl_npc_rna_meds %>% 
  inner_join(select(chr1.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr1.npc.meds

chr1.esc.meds %>%
  mutate( type="esc") %>% 
  rbind( chr1.npc.meds %>%  mutate( type ="npc")) %>% 
  mutate(mediation.lod = ifelse(target.symbol == mediator.symbol, NA, mediation.lod)) %>%
  mutate(lod_drop = target.lod - mediation.lod) %>%
  group_by(target.symbol) %>%
  arrange(mediation.lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr1.meds.ranked

chr1.meds.ranked.sum <- chr1.meds.ranked %>%
  filter(rank %in% c( 1)) %>% 
  group_by(mediator.symbol,type) %>%
  summarize(n = length(target.symbol), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol))

chr1.meds.ranked.sum

```

```{r chr1_meds_plot, fig.height=5, fig.width=14}

results <- chr1.meds.ranked %>%
  ungroup() %>%
  select(mediator.symbol, target.symbol, mediation.lod, target.lod, lod_drop, type) %>%
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol)) %>% 
  select(-mediation.lod, -target.lod,- type) %>%
  filter((mediator.symbol %in% str_c(chr1.meds.ranked.sum$mediator.symbol[1:5]))) %>%
  mutate(lod_drop = ifelse(lod_drop < 0, 0, lod_drop), lod_drop = ifelse(lod_drop > 6, 6, lod_drop)) %>%
  rename(target = target.symbol, LOD_diff = lod_drop)

p <- ggplot(results, aes(y = mediator.symbol, x = target)) +
  geom_point(aes(color = LOD_diff, size = exp(LOD_diff) / 30), alpha = 0.6) +
  scale_color_gradientn(
    colors = c("white", "firebrick3", "navy"),
    values = scales::rescale(c(0, 3, 6)),
    name = "LOD\ndifference", limits = c(0, 6)
  ) +
  scale_size(breaks = 0:6, labels = as.character(0:6), range = c(0, 8)) +
  guides(size = "none") +
  theme_pubclean(base_size = 18) +
  theme(
    axis.text.y = element_text(size = 14, hjust = 1),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 0),
    axis.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 16),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank()
  ) +
  xlab("Target pQTL")+
  ylab("Mediator")
p

# ggsave(p, filename = here("figures","chr1_hotspot_mediation.svg"), width = 14, height = 5, units = "in", dpi = 300)

```


Which eQTL are best mediated by *Pign* ESC expression?

```{r}

chr1.meds.ranked %>% 
  filter( rank == 1 & mediator.symbol =="Pign" & type =="esc") %>% 
  select( 
    target.id, target.symbol, target.chr, target.lod, mediation.lod ) %>% 
  mutate_if(is.numeric, round, 2) %>% 
  create_dt()


```




### Closer look at *Pign*

```{r pign_eqtl}

pign_eqtl <- peaks.npc_rna.wesc.overlap %>% 
  filter( lod.esc_rna >7.5, mgi_symbol =="Pign")

# LOD plot
pign_esc_scan <- scan1( pheno = exprZ.esc_rna[,pign_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.esc_rna, 
                    kinship = kinship_loco.esc_rna,
                    addcovar = covar.esc_rna)

pign_npc_scan <- scan1( pheno = exprZ.npc_rna[,pign_eqtl$ensembl_gene_id,drop=FALSE], 
                    genoprobs = probs.npc_rna, 
                    kinship = kinship_loco.npc_rna,
                    addcovar = covar.npc_rna)


pign_esc_scan %>% 
  as.data.frame( ) %>% 
  rename( esc_rna = pign_eqtl$ensembl_gene_id) %>% 
  mutate( marker = dimnames(pign_esc_scan)[[1]]) %>% 
  left_join(map_dat2) %>% 
  cbind(
    pign_npc_scan %>% as.data.frame() %>% rename( npc_rna = pign_eqtl$ensembl_gene_id)
  ) -> pign_qtl_scans

pign_qtl_scans %>% 
  filter( chr == pign_eqtl$peak_chr) %>% 
  pivot_longer( cols = c("npc_rna","esc_rna"), names_to = "qtl_type", values_to = "lod") %>% 
  mutate( qtl_type = factor( qtl_type, levels = c("npc_rna","esc_rna"))) %>% 
  ggplot()+
    aes( 
      x= pos_bp/1e06,
      y = lod,
      col = qtl_type
      )+
    geom_rect(  xmin =filter( bands.esc.npc.rna,chr==1)$start[2]/1e06, 
              xmax = filter( bands.esc.npc.rna,chr==1)$end[2]/1e06, 
              ymin = 0, 
              ymax = 25, 
              fill = "gray", 
             inherit.aes = FALSE, 
             alpha = 0.1, 
             show.legend = FALSE)+
    geom_line( size = 1.5)+
    theme_pubclean( base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                       labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  xlab(paste0("Chr ",pign_eqtl$peak_chr," location (Mbp)"))+
  ylab( "LOD score")+
  labs(col = "QTL type")+
  geom_segment( x = pign_eqtl$gene_start/1e06, xend = pign_eqtl$gene_end/1e06 , y = 0, yend = 1, col = "black", size = 2) +
  annotate( "text", x= pign_eqtl$midpoint/1e06, y = -0.8, label ="Pign", size =6, fontface = "italic")+
  annotate( "text", x= 12+filter( bands.esc.npc.rna,chr==1)$end[2]/1e06, y = 23, label ="eQTL \n Hotspot", size =5)-> pign_lod_plot



# Effects plot
pign_esc_eff <- peaks.esc_rna.wEffs %>% 
  inner_join( pign_eqtl %>% select(ensembl_gene_id, peak_chr, lod.esc_rna))

pign_npc_eff <- peaks.npc_rna.wEffs %>% 
  inner_join( pign_eqtl %>% select(ensembl_gene_id, peak_chr, lod.npc_rna))
pign_effs <- pign_esc_eff %>% 
  select( mgi_symbol, paste0(LETTERS[1:8], ".esc_rna")) %>% 
  cbind( pign_npc_eff %>% select(paste0(LETTERS[1:8], ".npc_rna")) ) 

pign_effs %>% 
  pivot_longer( cols = c( paste0(LETTERS[1:8], ".npc_rna"),
          paste0(LETTERS[1:8], ".esc_rna")), 
          names_to = c("effect"),
          values_to = "value") %>% 
  separate(effect, sep ="[.]", into = c("effect","type")) %>% 
  mutate( effect = case_when( effect == "A" ~ "AJ",
                              effect == "B" ~ "B6",
                              effect == "C" ~ "129",
                              effect == "D" ~ "NOD",
                              effect == "E" ~ "NZO",
                              effect == "F" ~ "CAST",
                              effect == "G" ~ "PWK",
                              effect == "H" ~ "WSB")) %>% 
  mutate( type = factor(type, levels =c("npc_rna","esc_rna"))) %>% 
  ggplot()+
  aes( x = effect,
       y = value, 
       col = type,
       group = type)+
  geom_point(size = 4, show.legend = FALSE)+
  geom_line(show.legend = T, size = 1.2)+
  theme_pubclean(base_size = 18)+
  scale_color_manual( values = c(qtl.colors[["npc_rna"]],qtl.colors[["esc_rna"]]), 
                      labels = c("npc_rna"="NPC","esc_rna"="ESC"))+
  ylab("Haplotype effects")+
  xlab("")+
  #ylim(-2,1.1)+
  geom_hline( yintercept = 0)+
  theme(axis.line.x = element_blank(),
        axis.title = element_text(size = 18))+
  labs(col ="QTL type")+
  coord_flip( clip ="off")+
  theme(legend.position = "none") -> pign_haplotype_plot

```

```{r pign_lod_haps_plot, fig.height=5, fig.width=10}

pign_plot<- ggarrange( pign_lod_plot, pign_haplotype_plot, widths = c(1, 0.5))
pign_plot

# ggsave(pign_plot, filename = here("figures","pign_plot.svg"), width = 12, height = 5, dpi = 300)

```

```{r pign_as}

pmap_Mbp <- lapply( pmap, function(x){x/1e06})

probs <- probs.esc_rna
attr(probs, "is_x_chr") <- NULL
pign_as <- scan1snps(genoprobs = probs, 
                     map = pmap_Mbp, 
                     pheno = exprZ.esc_rna[,pign_eqtl$ensembl_gene_id,drop = F],
                     kinship = kinship_loco.esc_rna[[pign_eqtl$peak_chr]],
                     addcovar = covar.esc_rna, 
                     chr = as.character(pign_eqtl$peak_chr),
                     query_func = query_variants,
                     start = pign_eqtl$interp_bp_peak.esc_rna/1e06 - 5,
                     end = pign_eqtl$interp_bp_peak.esc_rna/1e06 + 5,
                     keep_all_snps = F

                                          )
attr(probs.npc_rna, "is_x_chr") <- NULL
pign_as_npc <- scan1snps(genoprobs = probs.npc_rna, 
                     map = pmap_Mbp, 
                     pheno = exprZ.npc_rna[,pign_eqtl$ensembl_gene_id,drop = F],
                     kinship = kinship_loco.npc_rna[[pign_eqtl$peak_chr]],
                     addcovar = covar.npc_rna, 
                     chr = as.character(pign_eqtl$peak_chr),
                     query_func = query_variants,
                     start = pign_eqtl$interp_bp_peak.npc_rna/1e06 - 5,
                     end = pign_eqtl$interp_bp_peak.npc_rna/1e06 + 5,
                     keep_all_snps = F
                     )

genes <- query_genes( pign_eqtl$peak_chr, 
                      pign_eqtl$interp_bp_peak.esc_rna/1e06 - 5, 
                      pign_eqtl$interp_bp_peak.esc_rna/1e06 + 5) %>%  
  distinct()

```

```{r pign_as_plot, fig.width=12, fig.height=8}

par(mar=c(4.1, 4.1,2, 1))
plot(pign_as$lod,
     pign_as$snpinfo,
     #drop_hilit = 1,
     genes = genes, cex.lab =1.5)

```

SNPs with matching 4:4 in the area.

```{r pign_as_table}

pign_as %>% 
   as.data.frame() %>% 
  filter( ENSMUSG00000056536 > 5) %>% 
  select( LOD = ENSMUSG00000056536,
          snp_id = snpinfo.snp_id,
          pos = snpinfo.pos,
          alleles = snpinfo.alleles,
          consequence = snpinfo.consequence,
          ensembl_gene_id = snpinfo.ensembl_gene,
          `129` = snpinfo.129S1_SvImJ,
          `AJ` = snpinfo.A_J,
          `CAST` = snpinfo.CAST_EiJ,
          `NZO` = snpinfo.NZO_HlLtJ,
          `PWK` = snpinfo.PWK_PhJ,
          `B6` = snpinfo.C57BL_6J,
          `NOD` = snpinfo.NOD_ShiLtJ,
          `WSB` = snpinfo.WSB_EiJ
          ) %>% 
  filter( `129`==`B6`, B6==CAST , CAST==NOD &
            AJ == NZO, NZO ==WSB, PWK ==WSB) %>%
  left_join(., select(all.genes, ensembl_gene_id, mgi_symbol) %>%  distinct() ) %>% 
  mutate_if(is.numeric, round, 1) %>% 
  create_dt()

```

Comparing ESC vs NPC expression of **Pign**:

```{r}

pign_expr <- tibble( esc = expr.esc_rna[shared.samples,"ENSMUSG00000056536"],
        npc = expr.npc_rna[shared.samples, "ENSMUSG00000056536"],
        esc_rankz = exprZ.esc_rna[shared.samples, "ENSMUSG00000056536"],
        npc_rankz = exprZ.npc_rna[shared.samples, "ENSMUSG00000056536"],
        sample = shared.samples)

ggscatter( pign_expr,
           x = "esc_rankz",
           y = "npc_rankz",
           size = 3, 
           alpha = 0.6,
           add = "reg.line", # Add regression line
           conf.int = TRUE, # Add confidence interval
           add.params = list(color = "blue", fill = "lightgray"), show.legend.text = FALSE
) +
  stat_cor(method = "spearman") + # Add correlation coefficient
  xlab("ESC") +
  ylab("NPC") +
  theme_pubclean(base_size = 14)

```


### bmediatR results for Pign

**Pign** NPC expression:

```{r}
load(here("_data","bmediatr_results_Pign_300.RData")) # results

qtl_peaks <- peaks.npc_rna.wEffs %>%
  filter(peak_chr == 1,
          between(interp_bp_peak.npc_rna,104395228,112253474),
          local.npc_rna == FALSE,
          lod.npc_rna >= 6)
 
#find posterior probabilities 
post_results_prob <- c()
post_results_odds <- c()

Pign_details <- npc.genes %>%
   filter(mgi_symbol=="Pign") 

for (i in 1:length(results)) {
  post_results_prob[[i]] <- exp(results[[i]]$ln_post_c)[Pign_details$ensembl_gene_id, ]
  post_results_odds[[i]] <- (results[[i]]$ln_post_odds)[Pign_details$ensembl_gene_id, ]
  
}

mediation_results_odds <- do.call(rbind, post_results_odds) %>%
  as_tibble() %>%
  cbind( qtl_peaks
         ) %>%
  pivot_longer(1:8, names_to = "group", values_to = "odds") 

```

Odds:
```{r, fig.width=15, fig.height=5}

mediation_results_odds %>%
  filter( !str_detect(group, "y_"),
          !group %in% c("mediation_or_colocal","reactive")) %>% 
  ggplot()+
  aes(y = odds,
      x = ensembl_gene_id,
      col = group,
      fill = group)+
  geom_point()+
  theme_pubclean()+
  theme(legend.position = "top",
        axis.text.x = element_blank())+
  #xlim(0,1)+
  ylab("Log Posterior Odds")+
  xlab("")+
  #scale_x_log10()+
  ggtitle("Mediation with mNPC eQTL with Pign NPC Expression")

```

<!-- ```{r, eval = F} -->

<!-- mediation_results_odds %>% -->
<!--   filter( odds > 20 & group =="mediation") -->

<!-- ``` -->

<!-- <!-- Example: **Coa5** is the one with the highest odds of causal mediation. --> -->

<!-- ```{r, fig.width=10, fig.height=5, warning=FALSE, message=FALSE, eval = F} -->

<!-- # run bmediatR -->
<!-- Coa5_eqtl <- peaks.npc_rna.wesc.overlap %>%  -->
<!--   filter( lod.npc_rna >6, mgi_symbol =="Coa5", local.npc_rna == F) -->

<!-- marker <- map_dat2 %>% -->
<!--        mutate(diff = abs(pos_bp - Coa5_eqtl$interp_bp_peak.npc_rna)) %>% -->
<!--        slice_min(diff) -->
<!-- genotype <- pull_genoprobpos( -->
<!--     genoprobs = probs.npc_rna, -->
<!--     marker = marker$marker -->
<!--   ) -->

<!-- Coa5_bmediatr_scan <- bmediatR( -->
<!--     y = exprZ.npc_rna[,Coa5_eqtl$ensembl_gene_id,drop=FALSE], -->
<!--     M = exprZ.npc_rna[,colnames(exprZ.npc_rna) != Coa5_eqtl$ensembl_gene_id], -->
<!--     X = genotype,  -->
<!--     Z = covar.npc_rna -->
<!--   ) -->

<!-- Pign_details <- npc.genes %>% -->
<!--    filter(mgi_symbol=="Pign")  -->

<!-- # exp(Coa5_bmediatr_scan$ln_post_c)[Pign_details$ensembl_gene_id,] -->
<!-- # (Coa5_bmediatr_scan$ln_post_odds)[Pign_details$ensembl_gene_id,] -->

<!-- # plot_posterior_bar(Coa5_bmediatr_scan, med_annot = npc.genes %>% mutate( middle = midpoint ,symbol = mgi_symbol), med_var = "ensembl_gene_id", mediator_id = Pign_details$ensembl_gene_id) -->
<!-- # plot_posterior_odds(bmediatR_object = Coa5_bmediatr_scan, model_type = "mediation",med_annot = npc.genes %>% mutate( chr = gene_chr, middle = midpoint ,symbol = mgi_symbol), med_var = "ensembl_gene_id", bgcol = "white", altbgcol = "white", include_chr = 1, label_thresh = 20, label_thresh_greater_than = T, main = "Coa5") -->


<!-- ``` -->


**Pign** ESC expression:

```{r}
load(here("_data","bmediatr_shared_results_Pign_300.RData")) # shared_results

#find posterior probabilities
shared_post_results_prob <- c()
shared_post_results_odds <- c()

Pign_details <- npc.genes %>%
   filter(mgi_symbol=="Pign")

for (i in 1:length(shared_results)) {
  shared_post_results_prob[[i]] <- exp(shared_results[[i]]$ln_post_c)[Pign_details$ensembl_gene_id, ]
  shared_post_results_odds[[i]] <- (shared_results[[i]]$ln_post_odds)[Pign_details$ensembl_gene_id, ]

}

shared_mediation_results_odds <- do.call(rbind, shared_post_results_odds) %>%
  as_tibble() %>%
  cbind( qtl_peaks
         ) %>%
  pivot_longer(1:8, names_to = "group", values_to = "odds")


```

Odds:
```{r, fig.width=15, fig.height=5}

shared_mediation_results_odds %>%
  filter( !str_detect(group, "y_"),
          !group %in% c("mediation_or_colocal","reactive")) %>%
  ggplot()+
  aes(y = odds,
      x = ensembl_gene_id,
      col = group,
      fill = group)+
  geom_point()+
  theme_pubclean()+
  theme(legend.position = "top",
        axis.text.x = element_blank())+
  #xlim(0,1)+
  ylab("Log Posterior Odds")+
  xlab("")+
  #scale_x_log10()+
  ggtitle("Mediation with mNPC eQTL with Pign ESC Expression")

```


### Overlap between chr1 target genes and other sets overrepresented with SAC

I have found SAC genes overrepresented in:

  -   PC1 drivers
  -   Genes with low variation
  -   Genes with significant sex effects
  -   Chr1 hotspot target genes 
  
Let's look at the overlap between those categories. Are we always picking up the same set of genes? Note that, SAC also shows differences in activity between sexes and cell types according to GSVA scores. 

```{r}

chr1_spindle_genes <- g.chr1$result %>% 
  filter( p_value <0.01) %>% 
  filter( str_detect( term_name, "(?i)spindle") |
            str_detect(term_name, "(?i)segregation")|
            str_detect(term_name, "(?i)chromatid") |
            str_detect(term_name, "(?i)kinetochore") |
            str_detect(intersection,"Mad2l1")) %>% 
  select(term_name, mgi_symbol= intersection) %>%
  separate_rows(mgi_symbol, sep = ",") %>% 
  left_join(npc.genes)

# from 01b notebook:
# spindle_genes -- pc1 drivers ORA results
# spindle_genes_sex_eff -- sex eff genes ORA results
# spindle_genes_low_var -- low varitation genes ORA results

spindle_genes %>% 
  mutate( type = "PC1 driver") %>% 
  rbind(
    spindle_genes_sex_eff %>% 
      mutate( type = "Shows significant sex effect")
  ) %>% 
  rbind(
    spindle_genes_low_var %>% 
      mutate( type = "Shows low variation") 
  ) %>% 
  rbind(
    chr1_spindle_genes %>% 
      mutate(type = "Chr1 hotspot target")
  ) %>% 
  select(mgi_symbol,type) %>% 
  distinct() %>% 
  group_by(mgi_symbol) %>% 
  mutate( type = paste0(collapse = ", ", type), n = n()) %>% 
  distinct() %>% 
  create_dt()


  

```



<!-- ### Overlap between chr1 target genes and GPI-anchored proteins -->

<!-- I got a list of proteins that are known to have GPI-anchors from Uniprot and I am looking at their overlap to the targets of chr 1 eQTL hotspot: Nothing! There are no genes with a distant eQTL within the hotspot or genes located within the hotspot. -->

<!-- Next: Can I make any functional connections between the GPI anchored proteins and the chr1 hotspot targets? I manually compared the high confidence physical interaction network with and without the GPI anchored proteins to identify any unique connections that show up when the GPI proteins are included using STRING database. Here are the results: -->

<!-- -   Hmgb1, a target of the chr1 eQTL hotspot, interacts with Cd24a, a protein that has a GPI anchor. No physical interaction, STRING bases its high confidence on co-mentioning in pubmed abstracts in mouse and other organisms. -->

<!-- -   Ptprg, a target of the chr1 eQTL hotspot, interacts with Cntn3, a protein with a GPI anchor. Homologs are reported to be co-expressed and physically interact in other organisms in addition to these genes being co-mentioned in abstracts in mouse and other organisms. -->

<!-- -   Hdac2,a target of the chr1 eQTL hotspot, interacts with Mbd2 and Mbd3, proteins with GPI anchors that also interact with each other. There is experimental evidence for physical interaction between Hdac2 and Mbd3 whereas putative homologs are co-expressed, physically interact in other organisms and are co-mentioned in abstracts in mouse and other organisms. -->

<!-- -   Neo1,a target of the chr1 eQTL hotspot, interacts with Rgma, Rgmb and Hfe2, proteins with GPI anchors that are also predicted to interact with each other. There isn't any evidence for physical interactions in mice but putative homologs are co-expressed, physically interact in other organisms and are co-mentioned in abstracts in mouse and other organisms. -->

<!-- ```{r gpi_genes} -->

<!-- gpi_genes <- read_tsv(here("_data","GPI_anchored_proteins_mouse.tsv")) %>%  -->
<!--   select( "Gene names  (primary )", "Gene names  (synonym )" ) %>%  -->
<!--   separate_rows( "Gene names  (synonym )" , sep = " ") %>%  -->
<!--   pivot_longer( 1:2, names_to = "temp", values_to = "mgi_symbol") %>%  -->
<!--   select( -temp) %>%  -->
<!--   filter( !is.na(mgi_symbol)) %>% -->
<!--   distinct() %>%  -->
<!--   left_join(all.genes) %>%  -->
<!--   filter( !is.na(ensembl_gene_id)) -->

<!-- ``` -->

## eQTL hotspot on chromosome 10 {.tabset .tabset-fade .tabset-pills}

**SUMMARY:** We identified a `r round((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 10))$width/1e06,1)`Mb wide NPC-specific eQTL trans-band on Chromosome 10 at `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 10))$start,big.mark=",")`-`r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 10))$end,big.mark=",")` bp that contains `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 10))$distant_npc_rna_lod6,big.mark=",")` NPC eQTL at the suggestive threshold of LOD \> 6. Targets of the hotspot are overrepresented for genes involved in protein processing in the endoplasmic reticulum. There is a 4:4 split in founder haplotype effects for the target distant eQTL where 

```{r, eval = FALSE}

bands.esc.npc.rna %>% filter(seqnames==10)

```

```{r chr10_eqtl}


chr10.npc.eQTL1 <- peaks.npc_rna.wesc.overlap %>% 
  filter( peak_chr == 10) %>% 
  select( -peak_cM.npc_rna) %>% 
  left_join(peaks.npc_rna.wEffs) %>% 
  filter( lod.npc_rna > 6 & !local.npc_rna & !is.na(local.npc_rna) &
            interp_bp_peak.npc_rna >= filter( bands.esc.npc.rna,chr==10)$start[1] & 
            interp_bp_peak.npc_rna  <= filter( bands.esc.npc.rna,chr==10)$end[1] ) %>% 
  mutate( lod.esc_rna = ifelse( is.na(lod.esc_rna), 0, lod.esc_rna)) # gene is not found in the esc data

chr10.esc.eQTL1 <- peaks.esc_rna.wnpc.overlap %>% 
  filter( peak_chr == 10) %>%
  select(-peak_cM.esc_rna) %>% 
  left_join( peaks.esc_rna.wEffs) %>% 
  filter(lod.esc_rna >6 & !local.esc_rna & !is.na(local.esc_rna) & 
           interp_bp_peak.esc_rna >= filter( bands.esc.npc.rna,chr==10)$start[1] &
           interp_bp_peak.esc_rna  <= filter( bands.esc.npc.rna,chr==10)$end[1]) %>% 
  mutate( lod.npc_rna = ifelse( is.na(lod.npc_rna), 0, lod.npc_rna)) # gene is not found in the npc data  


```

### eQTL in the region

```{r chr10_eqtl_plot, fig.height=5, fig.width=5}

chr10.npc.eQTL1 %>% 
  mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match))  %>%
  ggplot()+
  aes(x=lod.esc_rna,y=lod.npc_rna,col=match)+
  geom_point(size=3,alpha=0.5)+
  theme_pubclean(base_size = 16)+
  geom_vline(xintercept = 6,linetype=2)+
  geom_hline(yintercept=6,linetype=2)+
  scale_color_manual(values= c("NPC eQTL"=qtl.colors[["npc_rna"]],
                               "ESC eQTL"=qtl.colors[["esc_rna"]],
                               "shared"=qtl.colors[["shared"]]))+
  xlab("LOD score from ESC eQTL scan")+
  ylab("LOD score from NPC eQTL scan")+
  xlim(0,10)+
  ylim(0,10) -> p 

p +
  geom_point(
    data = (chr10.esc.eQTL1 %>% 
              mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match)) ),
    size = 3, 
    alpha = 0.5
  )

```

### ORA with NPC eQTL in the region

```{r chr10_ora}

g.chr10 <- gost(query =chr10.npc.eQTL1$mgi_symbol,
                        organism = "mmusculus", 
                        domain_scope = "custom", 
                        custom_bg =npc.genes$mgi_symbol, #unique((genes.w.eqtl %>% filter(!local.npc_rna))$mgi_symbol)
                        correction_method = "fdr",
                        evcodes = TRUE)
g.chr10$result <- g.chr10$result %>% filter(term_size < 500)

```

```{r chr10_ora_plot}

gostplot(g.chr10)

```

```{r chr10_ora_table}

g.chr10$result %>% 
  select(source, term_name, intersection_size, term_size , FDR = p_value) %>%
  filter( FDR < 0.01) %>%
  mutate_if(is.numeric, formatC, digits =2, format ="fg") %>%
  create_dt()
  
```

### NPC eQTL allele effects

```{r npc_allele_effects_chr10}

haps <- c("A","B","C","D","E","F","G","H")

chr10.npc.eQTL1.mat <- chr10.npc.eQTL1 %>% 
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol")) %>% 
  distinct() %>%
  column_to_rownames("mgi_symbol") %>%
  as.matrix() %>% 
  t()

annotation <- chr10.npc.eQTL1 %>% 
  dplyr::mutate(match = ifelse(match =="npc_rna","NPC_eQTL",match)) %>%
  filter(mgi_symbol %in% colnames(chr10.npc.eQTL1.mat)) %>%
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol","match")) %>% 
  distinct() %>% 
  column_to_rownames("mgi_symbol") %>% 
  select(match) 

annotation_row <- data.frame(strain=c("AJ","B6","129","NOD","NZO","CAST","PWK","WSB"))
rownames(annotation_row) <- rownames(chr10.npc.eQTL1.mat)  

annot.colors <- list(strain = founder_colors,
                     match=c( NPC_eQTL=qtl.colors[["npc_rna"]], shared=qtl.colors[["shared"]]))

pheatmap(chr10.npc.eQTL1.mat,
                                   cluster_rows=T,show_rownames=FALSE,
                                   cluster_cols=T, show_colnames = FALSE, 
                                   clustering_method = "complete", 
                                   scale="none",
                                   clustering_distance_cols="correlation",
                                   clustering_distance_rows="correlation",
                                   #main = "Founder allele effects of chr 1 NPC eQTL hotspot genes",
                                   annotation_col= annotation, 
                                   annotation_row = annotation_row, 
                                   annotation_colors = annot.colors, 
                                   cutree_rows = 2)

```

### Mediation of target eQTL

```{r chr10_meds, results='hide'}

eqtl_esc_rna_meds %>%
  inner_join(select(chr10.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr10.esc.meds

eqtl_npc_rna_meds %>% 
  inner_join(select(chr10.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr10.npc.meds

chr10.esc.meds %>%
  mutate( type="esc") %>% 
  rbind( chr10.npc.meds %>%  mutate( type ="npc")) %>% 
  mutate(mediation.lod = ifelse(target.symbol == mediator.symbol, NA, mediation.lod)) %>%
  mutate(lod_drop = target.lod - mediation.lod) %>%
  group_by(target.symbol) %>%
  arrange(mediation.lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr10.meds.ranked

chr10.meds.ranked.sum <- chr10.meds.ranked %>%
  filter(rank %in% c( 1)) %>% #filter(mediator.symbol =="Lifr") %>%  select(lod_drop)
  group_by(mediator.symbol,type) %>%
  summarize(n = length(target.symbol), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol))

chr10.meds.ranked.sum

```

```{r chr10_meds_plot, fig.height=5, fig.width=14}

results <- chr10.meds.ranked %>%
  ungroup() %>%
  select(mediator.symbol, target.symbol, mediation.lod, target.lod, lod_drop, type) %>%
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol)) %>% 
  select(-mediation.lod, -target.lod,- type) %>%
  filter((mediator.symbol %in% str_c(chr10.meds.ranked.sum$mediator.symbol[1:5]))) %>%
  mutate(lod_drop = ifelse(lod_drop < 0, 0, lod_drop), lod_drop = ifelse(lod_drop > 6, 6, lod_drop)) %>%
  rename(target = target.symbol, LOD_diff = lod_drop)

p <- ggplot(results, aes(y = mediator.symbol, x = target)) +
  geom_point(aes(color = LOD_diff, size = exp(LOD_diff) / 30), alpha = 0.6) +
  scale_color_gradientn(
    colors = c("white", "firebrick3", "navy"),
    values = scales::rescale(c(0, 3, 6)),
    name = "LOD\ndifference", limits = c(0, 6)
  ) +
  scale_size(breaks = 0:6, labels = as.character(0:6), range = c(0, 8)) +
  guides(size = "none") +
  theme_pubclean(base_size = 18) +
  theme(
    axis.text.y = element_text(size = 14, hjust = 1),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 0),
    axis.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 16),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank()
  ) +
  xlab("Target pQTL")+
  ylab("Mediator")
p

```

## eQTL hotspot on chromosome 11 {.tabset .tabset-fade .tabset-pills}

**SUMMARY:** We identified a `r round((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 11))$width/1e06,1)`Mb wide NPC-specific eQTL trans-band on Chromosome 11 at `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 11))$start,big.mark=",")`-`r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 11))$end,big.mark=",")` bp that contains `r formatC((bands.esc.npc.rna %>% filter(distant_npc_rna > 20, seqnames == 11))$distant_npc_rna_lod6,big.mark=",")` NPC eQTL at the suggestive threshold of LOD \> 6.

```{r, eval = FALSE}

bands.esc.npc.rna %>% filter(seqnames==11)

```

```{r chr11_eqtl}

chr11.npc.eQTL1 <- peaks.npc_rna.wesc.overlap %>% 
  filter( peak_chr == 11) %>% 
  select( -peak_cM.npc_rna) %>% 
  left_join( peaks.npc_rna.wEffs) %>% 
  filter( lod.npc_rna > 6 & !local.npc_rna & !is.na(local.npc_rna) &
            interp_bp_peak.npc_rna > filter( bands.esc.npc.rna,chr==11)$start[1] & 
            interp_bp_peak.npc_rna  < filter( bands.esc.npc.rna,chr==11)$end[1] ) %>% 
  mutate( lod.esc_rna = ifelse( is.na(lod.esc_rna), 0, lod.esc_rna)) # gene not found in the esc data

chr11.esc.eQTL1 <-  peaks.esc_rna.wnpc.overlap %>% 
  filter( peak_chr == 11) %>%
  select(-peak_cM.esc_rna) %>% 
  left_join(peaks.esc_rna.wEffs) %>% 
  filter(lod.esc_rna >6 & !local.esc_rna & !is.na(local.esc_rna) & 
           interp_bp_peak.esc_rna > filter( bands.esc.npc.rna,chr==11)$start[1] &
           interp_bp_peak.esc_rna  < filter( bands.esc.npc.rna,chr==11)$end[1]) %>% 
   mutate( lod.npc_rna = ifelse( is.na(lod.npc_rna), 0, lod.npc_rna)) # gene is not found in the npc data


```

### eQTL in the region

```{r chr11_eqtl_plot, fig.height=5, fig.width=5}

chr11.npc.eQTL1 %>% 
  mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match))  %>%
  ggplot()+
  aes(x=lod.esc_rna,y=lod.npc_rna,col=match)+
  geom_point(size=3,alpha=0.5)+
  theme_pubclean(base_size = 16)+
  geom_vline(xintercept = 6,linetype=2)+
  geom_hline(yintercept=6,linetype=2)+
  scale_color_manual(values= c("NPC eQTL"=qtl.colors[["npc_rna"]],
                               "ESC eQTL"=qtl.colors[["esc_rna"]],
                               "shared"=qtl.colors[["shared"]]))+
  xlab("LOD score from ESC eQTL scan")+
  ylab("LOD score from NPC eQTL scan")+
  xlim(0,10)+
  ylim(0,10) -> p

p+
  geom_point(
    data = (chr11.esc.eQTL1 %>% 
               mutate(match =ifelse(match=="npc_rna","NPC eQTL",match), match=ifelse(match=="esc_rna","ESC eQTL",match))),
    size = 3, 
    alpha = 0.5
  )

```

### ORA with NPC eQTL in the region

```{r chr11_ora}

g.chr11 <- gost(query =chr11.npc.eQTL1$mgi_symbol,
                        organism = "mmusculus", 
                        domain_scope = "custom", 
                        custom_bg =npc.genes$mgi_symbol, #unique((genes.w.eqtl %>% filter(!local.npc_rna))$mgi_symbol)
                        correction_method = "fdr",
                        evcodes = TRUE)
g.chr11$result <- g.chr11$result %>% filter(term_size < 500)

```

```{r chr11_ora_plot}

gostplot(g.chr11, capped = FALSE)

```

```{r chr11_ora_table}

g.chr11$result %>% 
  select(source, term_name, intersection_size, term_size , FDR = p_value) %>%
  filter( FDR < 0.01) %>%
  mutate_if(is.numeric, formatC, digits =2, format ="fg") %>%
  create_dt()
  
```

### NPC eQTL allele effects

```{r npc_allele_effects_chr11}

haps <- c("A","B","C","D","E","F","G","H")

chr11.npc.eQTL1.mat <- chr11.npc.eQTL1 %>% 
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol")) %>% 
  distinct() %>%
  column_to_rownames("mgi_symbol") %>%
  as.matrix() %>% 
  t()

annotation <- chr11.npc.eQTL1 %>% 
  dplyr::mutate(match = ifelse(match =="npc_rna","NPC_eQTL",match)) %>%
  filter(mgi_symbol %in% colnames(chr11.npc.eQTL1.mat)) %>%
  filter(!is.na(A.npc_rna)) %>% 
  select(c(paste0(haps,".npc_rna"),"mgi_symbol","match")) %>% 
  distinct() %>% 
  column_to_rownames("mgi_symbol") %>% 
  select(match) 

annotation_row <- data.frame(strain=c("AJ","B6","129","NOD","NZO","CAST","PWK","WSB"))
rownames(annotation_row) <- rownames(chr11.npc.eQTL1.mat)  

annot.colors <- list(strain = founder_colors,
                     match=c( NPC_eQTL=qtl.colors[["npc_rna"]], shared=qtl.colors[["shared"]]))

pheatmap(chr11.npc.eQTL1.mat,
                                   cluster_rows=T,show_rownames=FALSE,
                                   cluster_cols=T, show_colnames = FALSE, 
                                   clustering_method = "complete", 
                                   scale="none",
                                   clustering_distance_cols="correlation",
                                   clustering_distance_rows="correlation",
                                   #main = "Founder allele effects of chr 1 NPC eQTL hotspot genes",
                                   annotation_col= annotation, 
                                   annotation_row = annotation_row, 
                                   annotation_colors = annot.colors, 
                                   cutree_rows = 2)

```

### ESC eQTL allele effects

```{r esc_allele_effects_chr11}

haps <- c("A","B","C","D","E","F","G","H")

chr11.esc.eQTL1.mat <- chr11.esc.eQTL1 %>% 
  filter(!is.na(A.esc_rna)) %>% 
  select(c(paste0(haps,".esc_rna"),"mgi_symbol")) %>% 
  distinct() %>%
  column_to_rownames("mgi_symbol") %>%
  as.matrix() %>% 
  t()

annotation <- chr11.esc.eQTL1 %>% 
  dplyr::mutate(match = ifelse(match =="esc_rna","ESC_eQTL",match)) %>%
  filter(mgi_symbol %in% colnames(chr11.esc.eQTL1.mat)) %>%
  filter(!is.na(A.esc_rna)) %>% 
  select(c(paste0(haps,".esc_rna"),"mgi_symbol","match")) %>% 
  distinct() %>% 
  column_to_rownames("mgi_symbol") %>% 
  select(match) 

annotation_row <- data.frame(strain=c("AJ","B6","129","NOD","NZO","CAST","PWK","WSB"))
rownames(annotation_row) <- rownames(chr11.esc.eQTL1.mat)  

annot.colors <- list(strain = founder_colors,
                     match=c( ESC_eQTL=qtl.colors[["esc_rna"]], shared=qtl.colors[["shared"]]))

pheatmap(chr11.esc.eQTL1.mat,
                                   cluster_rows=T,show_rownames=FALSE,
                                   cluster_cols=T, show_colnames = FALSE, 
                                   clustering_method = "complete", 
                                   scale="none",
                                   clustering_distance_cols="correlation",
                                   clustering_distance_rows="correlation",
                                   annotation_col= annotation, 
                                   annotation_row = annotation_row, 
                                   annotation_colors = annot.colors, 
                                   cutree_rows = 2)

```

### Mediation of target NPC eQTL

```{r chr11_meds, results='hide'}

eqtl_esc_rna_meds %>%
  inner_join(select(chr11.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr11.esc.meds

eqtl_npc_rna_meds %>% 
  inner_join(select(chr11.npc.eQTL1, "target.id" = ensembl_gene_id, "qtl.chr" = peak_chr)) -> chr11.npc.meds

chr11.esc.meds %>%
  mutate( type="esc") %>% 
  rbind( chr11.npc.meds %>%  mutate( type ="npc")) %>% 
  mutate(mediation.lod = ifelse(target.symbol == mediator.symbol, NA, mediation.lod)) %>%
  mutate(lod_drop = target.lod - mediation.lod) %>%
  group_by(target.symbol) %>%
  arrange(mediation.lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr11.meds.ranked

chr11.meds.ranked.sum <- chr11.meds.ranked %>%
  filter(rank %in% c( 1)) %>% #filter(mediator.symbol =="Lifr") %>%  select(lod_drop)
  group_by(mediator.symbol,type) %>%
  summarize(n = length(target.symbol), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol))

chr11.meds.ranked.sum

```

```{r chr11_npc_eqtl_run_mediation, cache = TRUE, eval = FALSE}

chr11_npc_eqtl_npc_meds <- do_global_mediation(gene.name = chr11.npc.eQTL1$mgi_symbol[1],
                    p.chr = chr11.npc.eQTL1$peak_chr[1],
                    peak.info = chr11.npc.eQTL1,
                    probs = probs.npc_rna,
                    expr.target = exprZ.npc_rna,
                    expr.mediator = exprZ.npc_rna,
                    covar= covar.npc_rna,
                    med.type = "n-n",
                    genes = npc.genes)

for(i in 2:nrow(chr11.npc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr11.npc.eQTL1$mgi_symbol[i],
                    p.chr = chr11.npc.eQTL1$peak_chr[i],
                    peak.info = chr11.npc.eQTL1,
                    probs = probs.npc_rna,
                    expr.target = exprZ.npc_rna,
                    expr.mediator = exprZ.npc_rna,
                    covar= covar.npc_rna,
                    med.type = "n-n",
                    genes = npc.genes)
  chr11_npc_eqtl_npc_meds <- rbind(chr11_npc_eqtl_npc_meds, meds)
}

chr11_npc_eqtl_esc_meds <- do_global_mediation(gene.name = chr11.npc.eQTL1$mgi_symbol[1],
                    p.chr = chr11.npc.eQTL1$peak_chr[1],
                    peak.info = chr11.npc.eQTL1,
                    probs = shared.probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.esc_rna[shared.samples,,drop=F],
                    covar= covar.npc_rna[shared.samples,,drop=F],
                    med.type = "e-n",
                    genes = npc.genes)

for(i in 2:nrow(chr11.npc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr11.npc.eQTL1$mgi_symbol[i],
                    p.chr = chr11.npc.eQTL1$peak_chr[i],
                    peak.info = chr11.npc.eQTL1,
                    probs = shared.probs.npc_rna,
                    expr.target = exprZ.npc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.esc_rna[shared.samples,,drop=F],
                    covar= covar.npc_rna[shared.samples,,drop=F],
                    med.type = "e-n",
                    genes = npc.genes)
  chr11_npc_eqtl_esc_meds <- rbind(chr11_npc_eqtl_esc_meds, meds)
}

```

```{r, eval = FALSE}

# I reran the mediation locally to see if the results I get are the same and they are identical. So no worries on the mediation results! Mainly, filtering based on z-threshold removes a lot of mediators that are not 'good enough'. 

chr11_npc_eqtl_npc_meds %>% 
  mutate( type = "npc") %>% 
  rbind(chr11_npc_eqtl_esc_meds %>% 
          mutate(type ="esc")) %>% 
    left_join( chr11.npc.eQTL1 %>% 
                 select( target = mgi_symbol, 
                         peak_chr, 
                         target_lod = lod.npc_rna,
                         target_pos = interp_bp_peak.npc_rna,
                         target_chr = gene_chr)) %>%
  select(
    target,
    peak_chr,
    target_lod,
    target_pos,
    target_chr,
    mediator_id = ensembl_gene_id, 
    mediator,
    mediator_chr = gene_chr, 
    mediator_midpoint = midpoint,
    mediation_lod = LOD,
    type
  ) %>% 
  group_by( target, peak_chr, target_lod, target_chr,type) %>% 
  mutate( scaled_LOD = scale(mediation_lod)) %>% 
  filter( abs(target_pos-mediator_midpoint) <= 10e6 &   # filter within 10Mb
          peak_chr == mediator_chr,
          scaled_LOD < -4 # z threshold -4
          )   -> chr11_npc_eqtl_meds
  
chr11_npc_eqtl_meds %>% 
  mutate(mediation_lod = ifelse(target == mediator, NA, mediation_lod)) %>%
  mutate(lod_drop = target_lod - mediation_lod) %>%
  group_by(target) %>%
  arrange(mediation_lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr11.meds.ranked

chr11.meds.ranked.sum <- chr11.meds.ranked %>%
  filter(rank %in% c( 1)) %>% 
  group_by(mediator,type) %>%
  summarize(n = length(target), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator = str_c(toupper(type)," , " ,mediator))

chr11.meds.ranked.sum

```

```{r chr11_meds_plot, fig.height=5, fig.width=14}

results <- chr11.meds.ranked %>%
  ungroup() %>%
  select(mediator.symbol, target.symbol, mediation.lod, target.lod, lod_drop, type) %>%
  mutate(mediator.symbol = str_c(toupper(type)," , " ,mediator.symbol)) %>% 
  select(-mediation.lod, -target.lod,- type) %>%
  filter((mediator.symbol %in% str_c(chr11.meds.ranked.sum$mediator.symbol[1:5]))) %>%
  mutate(lod_drop = ifelse(lod_drop < 0, 0, lod_drop), lod_drop = ifelse(lod_drop > 6, 6, lod_drop)) %>%
  rename(target = target.symbol, LOD_diff = lod_drop)

p <- ggplot(results, aes(y = mediator.symbol, x = target)) +
  geom_point(aes(color = LOD_diff, size = exp(LOD_diff) / 30), alpha = 0.6) +
  scale_color_gradientn(
    colors = c("white", "firebrick3", "navy"),
    values = scales::rescale(c(0, 3, 6)),
    name = "LOD\ndifference", limits = c(0, 6)
  ) +
  scale_size(breaks = 0:6, labels = as.character(0:6), range = c(0, 8)) +
  guides(size = "none") +
  theme_pubclean(base_size = 18) +
  theme(
    axis.text.y = element_text(size = 14, hjust = 1),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 0),
    axis.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 16),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank()
  ) +
  xlab("Target pQTL")+
  ylab("Mediator")
p

```

### Mediation of target ESC eQTL

```{r chr11_esc_eqtl_run_mediation, cache = TRUE}

chr11_esc_eqtl_esc_meds <- do_global_mediation(gene.name = chr11.esc.eQTL1$mgi_symbol[1],
                    p.chr = chr11.esc.eQTL1$peak_chr[1],
                    peak.info = chr11.esc.eQTL1,
                    probs = probs.esc_rna,
                    expr.target = exprZ.esc_rna,
                    expr.mediator = exprZ.esc_rna,
                    covar= covar.esc_rna,
                    med.type = "e-e",
                    genes = esc.genes)

for(i in 2:nrow(chr11.esc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr11.esc.eQTL1$mgi_symbol[i],
                    p.chr = chr11.esc.eQTL1$peak_chr[i],
                    peak.info = chr11.esc.eQTL1,
                    probs = probs.esc_rna,
                    expr.target = exprZ.esc_rna,
                    expr.mediator = exprZ.esc_rna,
                    covar= covar.esc_rna,
                    med.type = "e-e",
                    genes = esc.genes)
  chr11_esc_eqtl_esc_meds <- rbind(chr11_esc_eqtl_esc_meds, meds)
}

chr11_esc_eqtl_npc_meds <- do_global_mediation(gene.name = chr11.esc.eQTL1$mgi_symbol[1],
                    p.chr = chr11.esc.eQTL1$peak_chr[1],
                    peak.info = chr11.esc.eQTL1,
                    probs = shared.probs.esc_rna,
                    expr.target = exprZ.esc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.npc_rna[shared.samples,,drop=F],
                    covar= covar.esc_rna[shared.samples,,drop=F],
                    med.type = "n-e",
                    genes = esc.genes)

for(i in 2:nrow(chr11.esc.eQTL1)){
  meds <- do_global_mediation(gene.name = chr11.esc.eQTL1$mgi_symbol[i],
                    p.chr = chr11.esc.eQTL1$peak_chr[i],
                    peak.info = chr11.esc.eQTL1,
                    probs = shared.probs.esc_rna,
                    expr.target = exprZ.esc_rna[shared.samples,,drop=FALSE],
                    expr.mediator = exprZ.npc_rna[shared.samples,,drop=F],
                    covar= covar.esc_rna[shared.samples,,drop=F],
                    med.type = "n-e",
                    genes = esc.genes)
  chr11_esc_eqtl_npc_meds <- rbind(chr11_esc_eqtl_npc_meds, meds)
}

```

```{r, eval = TRUE}

chr11_esc_eqtl_npc_meds %>% 
  mutate( type = "npc") %>% 
  rbind(chr11_esc_eqtl_esc_meds %>% 
          mutate(type ="esc")) %>% 
    left_join( chr11.esc.eQTL1 %>% 
                 select( target = mgi_symbol, 
                         peak_chr, 
                         target_lod = lod.npc_rna,
                         target_pos = interp_bp_peak.npc_rna,
                         target_chr = gene_chr)) %>%
  select(
    target,
    peak_chr,
    target_lod,
    target_pos,
    target_chr,
    mediator_id = ensembl_gene_id, 
    mediator,
    mediator_chr = gene_chr, 
    mediator_midpoint = midpoint,
    mediation_lod = LOD,
    type
  ) %>% 
  group_by( target, peak_chr, target_lod, target_chr,type) %>% 
  mutate( scaled_LOD = scale(mediation_lod)) %>% 
  filter( abs(target_pos-mediator_midpoint) <= 10e6 &   # filter within 10Mb
          peak_chr == mediator_chr,
          scaled_LOD < -4 # z threshold -4
          )   -> chr11_esc_eqtl_meds
  
chr11_esc_eqtl_meds %>% 
  mutate(mediation_lod = ifelse(target == mediator, NA, mediation_lod)) %>%
  mutate(lod_drop = target_lod - mediation_lod) %>%
  filter(lod_drop >0) %>% 
  group_by(target) %>%
  arrange(mediation_lod) %>%
  mutate(rank = rep(seq(1:n()))) -> chr11.meds.ranked

chr11.meds.ranked.sum <- chr11.meds.ranked %>%
  filter(rank %in% c( 1)) %>% 
  group_by(mediator,type) %>%
  summarize(n = length(target), min_drop = min(lod_drop, na.rm = T), max_drop = max(lod_drop, na.rm = T), med_drop = median(lod_drop, na.rm = T)) %>%
  arrange(desc(n)) %>% 
  mutate(mediator = str_c(toupper(type)," , " ,mediator))

chr11.meds.ranked.sum

```

```{r chr11_meds_plot2, fig.height=5, fig.width=14}

results <- chr11.meds.ranked %>%
  ungroup() %>%
  select(mediator, target, mediation_lod, target_lod, lod_drop, type) %>%
  mutate(mediator = str_c(toupper(type)," , " ,mediator)) %>% 
  select(-mediation_lod, -target_lod,-type) %>%
  filter((mediator %in% str_c(chr11.meds.ranked.sum$mediator[1:5]))) %>%
  mutate(lod_drop = ifelse(lod_drop < 0, 0, lod_drop), lod_drop = ifelse(lod_drop > 6, 6, lod_drop)) %>%
  rename( LOD_diff = lod_drop)

p <- ggplot(results, aes(y = mediator, x = target)) +
  geom_point(aes(color = LOD_diff, size = exp(LOD_diff) / 30), alpha = 0.6) +
  scale_color_gradientn(
    colors = c("white", "firebrick3", "navy"),
    values = scales::rescale(c(0, 3, 6)),
    name = "LOD\ndifference", limits = c(0, 6)
  ) +
  scale_size(breaks = 0:6, labels = as.character(0:6), range = c(0, 8)) +
  guides(size = "none") +
  theme_pubclean(base_size = 18) +
  theme(
    axis.text.y = element_text(size = 14, hjust = 1),
    axis.ticks = element_blank(),
    axis.text.x = element_text(size = 0),
    axis.title = element_text(size = 14),
    legend.text = element_text(size = 14),
    legend.title = element_text(size = 16),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank()
  ) +
  xlab("Target pQTL")+
  ylab("Mediator")
p

```
